{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports & Constants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from rdflib import Graph, Namespace\n",
    "from collections import defaultdict\n",
    "from tqdm.notebook import tqdm\n",
    "import json\n",
    "\n",
    "VIOLATION_LIST_FILE = \"./json/violation_list.json\" # path to store the violation list\n",
    "STUDY_CSV_FILE = \"./csv/study.csv\" # path to store the tabluraized study csv file\n",
    "STUDY_TTL_PATH = './ttl/study.ttl' # contains the actual study data\n",
    "OMICS_MODEL_TTL_PATH = './ttl/omics_model.ttl' # contains the ontology\n",
    "OMICS_MODEL_UNION_VIOLATION_EXEMPLAR_TTL_PATH = './ttl/omics_model_union_violation_exemplar.ttl' # contains the union of the ontology and the violation exemplars\n",
    "VIOLATION_REPORT_TTL_PATH = './ttl/violation_report.ttl' # contains the shacl violation test case result\n",
    "EXEMPLAR_EDGE_COUNT_DICT_PATH = './json/exemplar_edge_count_dict.json' # contains the nested dict of exemplar edge counts\n",
    "FOCUS_NODE_EXEMPLAR_DICT_PATH = './json/focus_node_exemplar_dict.json' # contains the dict of exemplars per focus node\n",
    "EXEMPLAR_FOCUS_NODE_DICT_PATH = './json/exemplar_focus_node_dict.json' # contains the dict of focus nodes per exemplar"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read & Parse Graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading study graph...\n",
      "Reading violations graph...\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Graph identifier=Nfae2ee43a79b431fa55aa74ce80085d5 (<class 'rdflib.graph.Graph'>)>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Reading study graph...\")\n",
    "study_g = Graph()\n",
    "study_g.parse(STUDY_TTL_PATH, format=\"ttl\")\n",
    "\n",
    "print(\"Reading violations graph...\")\n",
    "violations_g = Graph()\n",
    "violations_g.parse(VIOLATION_REPORT_TTL_PATH, format=\"ttl\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Give Enumarting Names to Unnamed NodeShapes\n",
    "example: [] a sh:NodeShape -> NodeShape1 a sh:NodeShape\n",
    "etc\n",
    "this could be done with rdflib but then when storing the graph to a .ttl file it would automatically mess with the namespaces and remove unused ones\n",
    "so until these names are provided by BI, we decided to do simple string replacement"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"color:orange\">Careful: If the omics_model.ttl contains [] anywhere else, e.g., in a label, it will also be replaced there!</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "# Read the input file\n",
    "with open(OMICS_MODEL_TTL_PATH, 'r') as file:\n",
    "    data = file.read()\n",
    "\n",
    "# Define backup file path\n",
    "backup_file_path = OMICS_MODEL_TTL_PATH + \".bak\"\n",
    "\n",
    "# Ensure the output directory exists\n",
    "os.makedirs(os.path.dirname(OMICS_MODEL_TTL_PATH), exist_ok=True)\n",
    "\n",
    "# store backup file\n",
    "with open(backup_file_path, 'w') as file:\n",
    "    file.write(data)\n",
    "\n",
    "# Initialize the counter\n",
    "counter = 1\n",
    "\n",
    "# Prepend '@prefix cns: <http://custom_node_shapes/> .'\n",
    "data = \"@prefix cns: <http://customnamespace.com/> .\\n\" + data\n",
    "\n",
    "# Replace occurrences of [] with NodeShape{counter}\n",
    "while \"[]\" in data:\n",
    "    data = data.replace(\"[]\", f\"cns:NodeShape{counter}\", 1)\n",
    "    counter += 1\n",
    "\n",
    "# Write the updated string to the new file\n",
    "with open(OMICS_MODEL_TTL_PATH, 'w') as file:\n",
    "    file.write(data)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Differentiate NaN Cases - Predicate cannot Exist on Class vs. Predicate is Missing vs. Predicta Exists Even Though It Cannot\n",
    "A node can have an edge if\n",
    "\n",
    "s1 a sh:NodeShape\n",
    "\n",
    "s1 sh:targetClass node\n",
    "\n",
    "s1 sh:property s2\n",
    "\n",
    "s2 a sh:PropertyShape\n",
    "\n",
    "s2 sh:Path edge\n",
    "\n",
    "OR \n",
    "\n",
    "(\n",
    "node rdfs:subClassOf s3\n",
    "AND\n",
    "s3 can have the edge\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Roots: ['omics:Thing', 'omics:KeyValueHolder', 'omics:ReferenceData']\n",
      "Edges: ['omics:hasCellType', 'omics:hasSourceTissue', 'omics:hasSourceCellType']\n",
      "Parents: ['omics:Specimen']\n",
      "Children: ['omics:PDXSpecimen', 'omics:PrimaryCellSpecimen']\n"
     ]
    }
   ],
   "source": [
    "from rdflib import Graph\n",
    "from rdflib.namespace import Namespace\n",
    "import networkx as nx\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# read ontology into a graph\n",
    "ontology_g = Graph()\n",
    "ontology_g.parse(OMICS_MODEL_TTL_PATH, format=\"ttl\")\n",
    "\n",
    "# take namespace prefix for sh and omics from the graph\n",
    "sh = Namespace(\"http://www.w3.org/ns/shacl#\")\n",
    "omics = Namespace(\"http://data.boehringer.com/ontology/omics#\")\n",
    "owl = Namespace(\"http://www.w3.org/2002/07/owl#\")\n",
    "\n",
    "# define prefixes and corresponding namespaces\n",
    "prefixes = {\n",
    "    \"sh\": \"http://www.w3.org/ns/shacl#\",\n",
    "    \"omics\": \"http://data.boehringer.com/ontology/omics#\",\n",
    "    \"owl\": \"http://www.w3.org/2002/07/owl#\"\n",
    "}\n",
    "\n",
    "def convert_to_prefixed(graph, uri):\n",
    "    \"\"\"Convert a URI to its prefixed form using a rdflib graph's namespace manager.\"\"\"\n",
    "    prefix, namespace, name = graph.namespace_manager.compute_qname(uri)\n",
    "    return f\"{prefix}:{name}\"\n",
    "\n",
    "def select_direct_possible_edges(node):\n",
    "    select_statement = \"\"\"\n",
    "    SELECT ?edge\n",
    "    WHERE {\n",
    "        ?ns a sh:NodeShape .\n",
    "        ?ns sh:targetClass omics:%s .\n",
    "        ?ns sh:property ?ps .\n",
    "        ?ps a sh:PropertyShape .\n",
    "        ?ps sh:path ?edge .\n",
    "    }\n",
    "    \"\"\" % node\n",
    "    return [str(result[0].n3(ontology_g.namespace_manager)) for result in ontology_g.query(select_statement)] # type: ignore\n",
    "\n",
    "def select_direct_parents(node):\n",
    "    select_statement = \"\"\"\n",
    "    SELECT ?parent\n",
    "    WHERE {\n",
    "        omics:%s rdfs:subClassOf ?parent .\n",
    "    }\n",
    "    \"\"\" % node\n",
    "    return [str(result[0].n3(ontology_g.namespace_manager)) for result in ontology_g.query(select_statement)] # type: ignore\n",
    "\n",
    "def select_direct_children(node):\n",
    "    select_statement = \"\"\"\n",
    "    SELECT ?child\n",
    "    WHERE {\n",
    "        ?child rdfs:subClassOf omics:%s .\n",
    "    }\n",
    "    \"\"\" % node\n",
    "    return [str(result[0].n3(ontology_g.namespace_manager)) for result in ontology_g.query(select_statement)] # type: ignore\n",
    "\n",
    "def select_roots():\n",
    "    select_statement = \"\"\"\n",
    "    SELECT ?root\n",
    "    WHERE {\n",
    "        ?root rdfs:subClassOf owl:Thing .\n",
    "    }\n",
    "    \"\"\"\n",
    "    return [str(result[0].n3(ontology_g.namespace_manager)) for result in ontology_g.query(select_statement)] # type: ignore\n",
    "\n",
    "def select_root_classes():\n",
    "    select_statement = \"\"\"\n",
    "    SELECT ?class\n",
    "    WHERE {\n",
    "        ?class a owl:Class .\n",
    "        FILTER NOT EXISTS { ?class rdfs:subClassOf ?parent . }\n",
    "    }\n",
    "    \"\"\"\n",
    "    return [str(result[0].n3(ontology_g.namespace_manager)) for result in ontology_g.query(select_statement)] # type: ignore\n",
    "\n",
    "\n",
    "# use the functions\n",
    "roots = select_root_classes()\n",
    "edges = select_direct_possible_edges('PrimaryCellSpecimen')\n",
    "parents = select_direct_parents('BiopsySpecimen')\n",
    "children = select_direct_children('BiopsySpecimen')\n",
    "\n",
    "print('Roots:', roots)\n",
    "print('Edges:', edges)\n",
    "print('Parents:', parents)\n",
    "print('Children:', children)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "class OntologyTree:\n",
    "    def __init__(self):\n",
    "        self.tree = {}\n",
    "\n",
    "    def add_node(self, node, edges, children):\n",
    "        if node not in self.tree:\n",
    "            self.tree[node] = {'edges': edges, 'children': children}\n",
    "\n",
    "    def add_child(self, parent, child):\n",
    "        if parent in self.tree:\n",
    "            self.tree[parent]['children'].append(child)\n",
    "        else:\n",
    "            self.tree[parent] = {'edges': [], 'children': [child]}\n",
    "\n",
    "def aggregate_edges(node, tree):\n",
    "    edges = set(tree[node]['edges'])\n",
    "    queue = [parent_node for parent_node in tree if node in tree[parent_node]['children']]\n",
    "    while queue:\n",
    "        parent_node = queue.pop(0)\n",
    "        edges |= set(tree[parent_node]['edges'])\n",
    "        queue += [grand_parent for grand_parent in tree if parent_node in tree[grand_parent]['children']]\n",
    "    return edges\n",
    "\n",
    "def create_aggregated_edges_dict(tree):\n",
    "    agg_edges_dict = {}\n",
    "    for node in tree.keys():\n",
    "        agg_edges_dict[node] = aggregate_edges(node, tree)\n",
    "    return agg_edges_dict\n",
    "\n",
    "def convert_to_full_uri(graph, abbreviated_uri):\n",
    "    \"\"\"Converts a prefixed URI to a full URI using a rdflib graph's namespace manager.\"\"\"\n",
    "    prefix, uri = abbreviated_uri.split(':')\n",
    "    namespace = graph.namespace_manager.store.namespace(prefix)\n",
    "    return namespace + uri\n",
    "\n",
    "def remove_prefix(uri):\n",
    "    if ':' in uri:\n",
    "        return uri.split(':', 1)[1]\n",
    "    return uri\n",
    "\n",
    "def build_tree(root_nodes, ontology_tree):\n",
    "    queue = root_nodes\n",
    "    while queue:\n",
    "        current_node = queue.pop(0)\n",
    "        unprefixed_node = remove_prefix(current_node)\n",
    "        edges = select_direct_possible_edges(unprefixed_node)\n",
    "        children = select_direct_children(unprefixed_node)\n",
    "\n",
    "        ontology_tree.add_node(current_node, edges, children)\n",
    "\n",
    "        for child in children:\n",
    "            queue.append(child)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "roots = select_root_classes()\n",
    "ontology_tree = OntologyTree()\n",
    "build_tree(roots, ontology_tree)\n",
    "agg_edges_dict = create_aggregated_edges_dict(ontology_tree.tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "http://data.boehringer.com/ontology/omics/hasCellType\n"
     ]
    }
   ],
   "source": [
    "full_uri = convert_to_full_uri(ontology_g, \"omics:hasCellType\")\n",
    "print(full_uri)  # Outputs: http://[blabla...]/omics:hasCellType\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def select_all_focus_nodes_and_classes(graph):\n",
    "    select_statement = \"\"\"\n",
    "    SELECT ?node ?class\n",
    "    WHERE {\n",
    "        ?node a ?class .\n",
    "    }\n",
    "    \"\"\"\n",
    "    results = graph.query(select_statement)\n",
    "    focus_nodes_classes_dict = {str(result[0]): str(result[1]) for result in results}\n",
    "    return focus_nodes_classes_dict\n",
    "\n",
    "\n",
    "focus_nodes_class_dict = select_all_focus_nodes_and_classes(study_g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "http://data.boehringer.com/ontology/omics/TranscriptOmicsSample\n",
      "omics:TranscriptOmicsSample\n",
      "{'omics:hasSequencingProtocol', 'omics:sampleOrder', 'omics:isPartOfSampleSet', 'omics:hasExclusionReason', 'omics:comment', 'omics:sourceId', 'omics:isMeasuredOnPlatform', 'omics:hasKeyValue', 'omics:hasStrandedness', 'omics:id', 'omics:isDerivedFromSpecimen', 'omics:isPartOfStudy', 'omics:hasRIN', 'omics:isPreparedByLibrary', 'omics:processingDate', 'omics:isExcluded', 'omics:isProcessedByOrganisation', 'omics:name'}\n"
     ]
    }
   ],
   "source": [
    "x = focus_nodes_class_dict['http://data.boehringer.com/uuid/TranscriptOmicsSample/sample-EX51-EX51_545']\n",
    "print(x) # this returns http://data.boehringer.com/ontology/omics/TranscriptOmicsSample\n",
    "print(convert_to_prefixed(ontology_g, x)) # should return omics:TranscriptOmicsSample\n",
    "print(agg_edges_dict[convert_to_prefixed(ontology_g, x)])\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tabularize & Join Study + Violations Graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "study graph triples: 97794\n",
      "violations graph triples: 286592\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing Graph Once: 100%|██████████| 286592/286592 [00:01<00:00, 147201.27it/s]\n",
      "Going Over Violation Instances: 100%|██████████| 17911/17911 [00:00<00:00, 900046.47it/s]\n",
      "Processing study: 100%|██████████| 97794/97794 [00:01<00:00, 75484.80it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count_skipped = 19186\n",
      "count_not_skipped = 78608\n",
      "count_allowed = 50295\n",
      "count_not_allowed_exists = 28313\n",
      "count_not_allowed_empty = 0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>focus_node</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/afb51f95-5b17-45a4-b62d-c58f4998f930</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/0c2600b5-30b2-40c7-917a-b402e73f55ae</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/248e9792-8549-4e0d-8665-06c42aff8ffc</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/9589ebf0-6f4f-4b80-9a5a-75aa25b5715f</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/38c0a1c7-1c92-489a-a22e-87f5335ccb1a</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/af998a22-f11b-43ea-b0ac-44e728baeafb</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-hasSequencingProtocol</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-hasStrandedness</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-isMeasuredOnPlatform</th>\n",
       "      <th>...</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/keywords</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/isSequencedForIndication</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/sourceStudyId</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/externalStudyId</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/studyId</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/hasStudyType</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/technicalStudyTitle</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/hasContactScientist</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/hasContactCompBio</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/sequencingDate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/Donor/9abcf968-791a-4520-b650-2439b9237515</th>\n",
       "      <td>http://data.boehringer.com/uuid/Donor/9abcf968...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/Donor/4b2b6b89-7e27-4475-a0e6-e6638541de43</th>\n",
       "      <td>http://data.boehringer.com/uuid/Donor/4b2b6b89...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/PrimaryCellSpecimen/3c0bedfd-345a-4b5a-8528-abe0516a4232</th>\n",
       "      <td>http://data.boehringer.com/uuid/PrimaryCellSpe...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/PrimaryCellSpecimen/f5f9d112-0e74-4363-9b71-351a541172d9</th>\n",
       "      <td>http://data.boehringer.com/uuid/PrimaryCellSpe...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/TranscriptOmicsSample/sample-EX51-EX51_104</th>\n",
       "      <td>http://data.boehringer.com/uuid/TranscriptOmic...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                           focus_node  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...  http://data.boehringer.com/uuid/Donor/9abcf968...   \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...  http://data.boehringer.com/uuid/Donor/4b2b6b89...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  http://data.boehringer.com/uuid/PrimaryCellSpe...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  http://data.boehringer.com/uuid/PrimaryCellSpe...   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...  http://data.boehringer.com/uuid/TranscriptOmic...   \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/afb51f95-5b17-45a4-b62d-c58f4998f930  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                0.0                                \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/0c2600b5-30b2-40c7-917a-b402e73f55ae  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                0.0                                \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/248e9792-8549-4e0d-8665-06c42aff8ffc  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                0.0                                \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/9589ebf0-6f4f-4b80-9a5a-75aa25b5715f  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                0.0                                \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/38c0a1c7-1c92-489a-a22e-87f5335ccb1a  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                0.0                                \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/af998a22-f11b-43ea-b0ac-44e728baeafb  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                0.0                                \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-hasSequencingProtocol  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                                0.0                                            \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                                0.0                                            \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                            \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                            \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                1.0                                            \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-hasStrandedness  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                                0.0                                      \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                                0.0                                      \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                      \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                      \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                1.0                                      \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-isMeasuredOnPlatform  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                                0.0                                           \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                                0.0                                           \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                           \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                           \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                1.0                                           \n",
       "\n",
       "                                                    ...  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...  ...   \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...  ...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  ...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  ...   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...  ...   \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/keywords  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                     EdgeNotPresent    \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                     EdgeNotPresent    \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent    \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent    \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent    \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/isSequencedForIndication  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                     EdgeNotPresent                    \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                     EdgeNotPresent                    \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent                    \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent                    \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent                    \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/sourceStudyId  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                     EdgeNotPresent         \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                     EdgeNotPresent         \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent         \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent         \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent         \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/externalStudyId  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                     EdgeNotPresent           \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                     EdgeNotPresent           \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent           \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent           \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent           \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/studyId  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                     EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                     EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent   \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/hasStudyType  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                     EdgeNotPresent        \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                     EdgeNotPresent        \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent        \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent        \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent        \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/technicalStudyTitle  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                     EdgeNotPresent               \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                     EdgeNotPresent               \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent               \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent               \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent               \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/hasContactScientist  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                     EdgeNotPresent               \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                     EdgeNotPresent               \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent               \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent               \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent               \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/hasContactCompBio  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                     EdgeNotPresent             \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                     EdgeNotPresent             \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent             \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent             \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent             \n",
       "\n",
       "                                                   http://data.boehringer.com/ontology/omics/sequencingDate  \n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                     EdgeNotPresent        \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                     EdgeNotPresent        \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent        \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent        \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent        \n",
       "\n",
       "[5 rows x 64 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from rdflib import Graph, Namespace\n",
    "from rdflib.namespace import NamespaceManager\n",
    "from collections import defaultdict\n",
    "import json\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "SH = Namespace(\"http://www.w3.org/ns/shacl#\")\n",
    "\n",
    "def count_violations(violations: Graph):\n",
    "    d_focus_node_d_source_shape_counts = defaultdict(lambda: defaultdict(int))\n",
    "    d_violation_focus_node = defaultdict((str))\n",
    "    d_violation_source_shape = defaultdict((str))\n",
    "    violation_list = []\n",
    "\n",
    "    for s, p, o in tqdm(violations, desc=\"Processing Graph Once\"):\n",
    "        if p == SH.sourceShape:\n",
    "            d_violation_source_shape[str(s)] = str(o)\n",
    "            violation_list += [str(o)]\n",
    "        elif p == SH.focusNode:\n",
    "            d_violation_focus_node[str(s)] = str(o)\n",
    "\n",
    "    # for key in violation_focus_node dict\n",
    "    for key in tqdm(d_violation_focus_node, desc=\"Going Over Violation Instances\"):\n",
    "        # init with 1 or add 1 if already exists\n",
    "        d_focus_node_d_source_shape_counts[d_violation_focus_node[key]][d_violation_source_shape[key]] += 1\n",
    "\n",
    "    violation_list = list(set(violation_list))\n",
    "    return d_focus_node_d_source_shape_counts, violation_list\n",
    "\n",
    "\n",
    "def save_violation_list(violation_list):\n",
    "    with open(VIOLATION_LIST_FILE, 'w') as f:\n",
    "        json.dump(violation_list, f)  \n",
    "\n",
    "\n",
    "def create_study_dataframe(study: Graph, d_focus_node_d_source_shape_counts, focus_nodes_class_dict, agg_edges_dict):\n",
    "    # for convenience\n",
    "    pre = lambda x: convert_to_prefixed(ontology_g, str(x))\n",
    "\n",
    "    # create an empty study_df, fill the violation counts\n",
    "    study_df = pd.DataFrame(d_focus_node_d_source_shape_counts).T.fillna(0)\n",
    "\n",
    "    count_not_skipped, count_skipped, count_allowed, count_not_allowed_exists, count_not_allowed_empty = 0, 0, 0, 0, 0\n",
    "\n",
    "    # iterate all study triples\n",
    "    for s, p, o in tqdm(study, desc=\"Processing study\"):\n",
    "        # if s is not in d_focus_node_d_source_shape_counts, continue as it is not a focus node\n",
    "        if str(s) not in d_focus_node_d_source_shape_counts:\n",
    "            count_skipped += 1\n",
    "            continue\n",
    "        count_not_skipped += 1\n",
    "\n",
    "        # if p is not in study_df.columns, add it and set it to EdgeNotPresent for all rows because so far there has not been a value for it for anything, and future values will overwrite it anyway\n",
    "        if str(p) not in study_df.columns:\n",
    "            study_df[str(p)] = \"EdgeNotPresent\"\n",
    "\n",
    "        # if object is empty throw exception and print details\n",
    "        if str(o) == \"\":\n",
    "            print(\"empty object for \" + str(s) + \" \" + str(p))\n",
    "            raise Exception(\"empty object for \" + str(s) + \" \" + str(p))\n",
    "\n",
    "        # get class of focus node\n",
    "        focus_node_class = focus_nodes_class_dict.get(str(s), None)\n",
    "        assert focus_node_class is not None, f\"The focus_node_class for {str(s)} was None, but expected a value.\"\n",
    "        # get allowed edges\n",
    "        focus_node_edges = agg_edges_dict.get(convert_to_prefixed(ontology_g, focus_node_class), set())\n",
    "        assert len(focus_node_edges) > 0\n",
    "\n",
    "        # if edge is allowed (and object is present otherwise we wouldn't get here), set value study_df.at[str(s), str(p)] = str(o)\n",
    "        if pre(p) in focus_node_edges:\n",
    "            study_df.at[str(s), str(p)] = str(o)\n",
    "            count_allowed += 1\n",
    "        # if edge is not allowed, and there is no object, set value study_df.at[str(s), str(p)] = \"EdgeNotPresent\"\n",
    "        elif pre(p) not in focus_node_edges and str(o) == \"\":\n",
    "            study_df.at[str(s), str(p)] = \"EdgeNotPresent\"\n",
    "            count_not_allowed_empty += 1\n",
    "        # if edge is not allowed, and there is an object, set value study_df.at[str(s), str(p)] = \"InvalidEdgeWUnexpectedEdgeWithDataithData: str(o)\"\n",
    "        elif pre(p) not in focus_node_edges and str(o) != \"\":\n",
    "            # study_df.at[str(s), str(p)] = \"UnexpectedEdgeWithData: \" + str(o)\n",
    "            study_df.at[str(s), str(p)] = str(o)\n",
    "            count_not_allowed_exists += 1\n",
    "        else:\n",
    "            raise Exception(\"should not happen: \" + str(s) + \" \" + str(p) + \" \" + str(o))\n",
    "\n",
    "    print(f\"count_skipped = {count_skipped}\")\n",
    "    print(f\"count_not_skipped = {count_not_skipped}\")\n",
    "    print(f\"count_allowed = {count_allowed}\")\n",
    "    print(f\"count_not_allowed_exists = {count_not_allowed_exists}\")\n",
    "    print(f\"count_not_allowed_empty = {count_not_allowed_empty}\")\n",
    "\n",
    "    assert study_df.isin(['']).sum().sum() == 0\n",
    "    assert study_df.isnull().sum().sum() == 0, f\"study_df.isnull().sum().sum() = {study_df.isnull().sum().sum()}\"\n",
    "\n",
    "    # add focus_node column, set it to the study_df.index, and move it to the very left of the dataframe\n",
    "    study_df['focus_node'] = study_df.index\n",
    "    cols = study_df.columns.tolist()\n",
    "    cols = cols[-1:] + cols[:-1]  # move the last column to the first\n",
    "    study_df = study_df[cols]\n",
    "\n",
    "    return study_df\n",
    "\n",
    "\n",
    "def save_study_dataframe(study_df):\n",
    "    study_df.to_csv(STUDY_CSV_FILE)\n",
    "\n",
    "\n",
    "def tabularize_graphs(study: Graph, violations: Graph):\n",
    "    print('study graph triples:', len(study))  # type: ignore\n",
    "    print('violations graph triples:', len(violations))  # type: ignore\n",
    "    d_focus_node_d_source_shape_counts, violation_list = count_violations(violations)\n",
    "    save_violation_list(violation_list)\n",
    "    study_df = create_study_dataframe(study, d_focus_node_d_source_shape_counts, focus_nodes_class_dict, agg_edges_dict)\n",
    "    save_study_dataframe(study_df)\n",
    "    return d_focus_node_d_source_shape_counts\n",
    "\n",
    "\n",
    "d_focus_node_d_source_shape_counts = tabularize_graphs(study_g, violations_g)\n",
    "\n",
    "study_df = pd.read_csv(STUDY_CSV_FILE, index_col=0)\n",
    "study_df.head()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Unit Tests for Violation Counting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing Graph Once: 100%|██████████| 6/6 [00:00<?, ?it/s]\n",
      "Going Over Violation Instances: 100%|██████████| 3/3 [00:00<?, ?it/s]\n",
      "Processing Graph Once: 100%|██████████| 8/8 [00:00<00:00, 7902.60it/s]\n",
      "Going Over Violation Instances: 100%|██████████| 4/4 [00:00<?, ?it/s]\n",
      ".\n",
      "----------------------------------------------------------------------\n",
      "Ran 2 tests in 0.010s\n",
      "\n",
      "OK\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<unittest.runner.TextTestResult run=2 errors=0 failures=0>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import unittest\n",
    "from rdflib import Graph, Literal, Namespace, BNode, URIRef\n",
    "\n",
    "SH = Namespace(\"http://www.w3.org/ns/shacl#\")\n",
    "\n",
    "class TestCountViolations(unittest.TestCase):\n",
    "    def setUp(self):\n",
    "        self.violations = Graph()\n",
    "\n",
    "    def test_count_violations_node_with_multiple_and_node_with_double(self):\n",
    "        self.focusNode1 = BNode()\n",
    "        self.focusNode2 = BNode()\n",
    "        self.sourceShape1 = Literal('shape1')\n",
    "        self.sourceShape2 = Literal('shape2')\n",
    "        self.violation1 = BNode()\n",
    "        self.violation2 = BNode()\n",
    "        self.violation3 = BNode()\n",
    "        self.violation4 = BNode()\n",
    "\n",
    "        # Add triples to the violations graph\n",
    "        self.violations.add((self.violation1, SH.focusNode, self.focusNode1))\n",
    "        self.violations.add((self.violation1, SH.sourceShape, self.sourceShape1))\n",
    "\n",
    "        self.violations.add((self.violation2, SH.focusNode, self.focusNode2))\n",
    "        self.violations.add((self.violation2, SH.sourceShape, self.sourceShape2))\n",
    "\n",
    "        self.violations.add((self.violation3, SH.focusNode, self.focusNode1))\n",
    "        self.violations.add((self.violation3, SH.sourceShape, self.sourceShape1))\n",
    "\n",
    "        self.violations.add((self.violation4, SH.focusNode, self.focusNode2))\n",
    "        self.violations.add((self.violation4, SH.sourceShape, self.sourceShape1))\n",
    "        violation_counts, violation_list = count_violations(self.violations)\n",
    "        expected_counts = {\n",
    "            str(self.focusNode1): {str(self.sourceShape1): 2},\n",
    "            str(self.focusNode2): {str(self.sourceShape2): 1, str(self.sourceShape1): 1}\n",
    "        }\n",
    "        expected_list = [str(self.sourceShape1), str(self.sourceShape2)] \n",
    "        \n",
    "        self.assertEqual(violation_counts, expected_counts)\n",
    "        self.assertEqual(set(violation_list), set(expected_list))  # Use set to ignore order\n",
    "\n",
    "    def test_count_violations_basic(self):\n",
    "        self.focusNode1 = BNode()\n",
    "        self.focusNode2 = BNode()\n",
    "        self.focusNode3 = BNode()\n",
    "        self.sourceShape1 = Literal('shape1')\n",
    "        self.sourceShape2 = Literal('shape2')\n",
    "        self.sourceShape3 = Literal('shape3')\n",
    "        self.violation1 = BNode()\n",
    "        self.violation2 = BNode()\n",
    "        self.violation3 = BNode()\n",
    "\n",
    "        # Add triples to the violations graph\n",
    "        self.violations.add((self.violation1, SH.focusNode, self.focusNode1))\n",
    "        self.violations.add((self.violation1, SH.sourceShape, self.sourceShape1))\n",
    "\n",
    "        self.violations.add((self.violation2, SH.focusNode, self.focusNode2))\n",
    "        self.violations.add((self.violation2, SH.sourceShape, self.sourceShape2))\n",
    "\n",
    "        self.violations.add((self.violation3, SH.focusNode, self.focusNode3))\n",
    "        self.violations.add((self.violation3, SH.sourceShape, self.sourceShape3))\n",
    "\n",
    "        violation_counts, violation_list = count_violations(self.violations)\n",
    "        expected_counts = {\n",
    "            str(self.focusNode1): {str(self.sourceShape1): 1},\n",
    "            str(self.focusNode2): {str(self.sourceShape2): 1},\n",
    "            str(self.focusNode3): {str(self.sourceShape3): 1}\n",
    "        }\n",
    "        expected_list = [str(self.sourceShape1), str(self.sourceShape2), str(self.sourceShape3)] \n",
    "        \n",
    "        self.assertEqual(violation_counts, expected_counts)\n",
    "        self.assertEqual(set(violation_list), set(expected_list))  # Use set to ignore order\n",
    "\n",
    "suite = unittest.TestLoader().loadTestsFromTestCase(TestCountViolations)\n",
    "unittest.TextTestRunner().run(suite)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Abbreviation using QNames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from rdflib import Graph\n",
    "from rdflib.namespace import split_uri\n",
    "\n",
    "def namespace_in_nsm(nsm, namespace):\n",
    "    \"\"\"\n",
    "    Checks if a given namespace is in the provided NamespaceManager.\n",
    "\n",
    "    Parameters:\n",
    "    nsm (NamespaceManager): the NamespaceManager object to check\n",
    "    namespace (str): the namespace to look for in the NamespaceManager\n",
    "\n",
    "    Returns:\n",
    "    bool: True if the namespace is in the NamespaceManager, False otherwise\n",
    "    \"\"\"\n",
    "    for prefix, ns in nsm.namespaces():\n",
    "        if str(ns) == str(namespace):\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "def get_qname(nsm, uri):\n",
    "    \"\"\"\n",
    "    Returns a QName for a given URI if its namespace is in the provided NamespaceManager.\n",
    "    If the namespace isn't found in the NamespaceManager, it returns the original URI.\n",
    "\n",
    "    Parameters:\n",
    "    nsm (NamespaceManager): the NamespaceManager object to use\n",
    "    uri (str): the URI to transform into a QName\n",
    "\n",
    "    Returns:\n",
    "    str: a QName if the namespace of the URI is in the NamespaceManager, the original URI otherwise\n",
    "    \"\"\"\n",
    "    namespace, name = split_uri(uri)\n",
    "    if namespace_in_nsm(nsm, namespace):\n",
    "        try:\n",
    "            prefix, _, _ = nsm.compute_qname(uri)\n",
    "            return f\"{prefix}:{name}\"\n",
    "        except Exception as e:\n",
    "            return uri\n",
    "    else:\n",
    "        return uri\n",
    "\n",
    "def abbreviate_using_namespaces(study_graph: Graph, violations_graph: Graph):\n",
    "    \"\"\"\n",
    "    Replaces URIs in a DataFrame and a list of violations with QNames if their namespaces\n",
    "    are found in two provided RDF graphs. Saves the updated DataFrame and the list back to disk.\n",
    "\n",
    "    Parameters:\n",
    "    study_graph (Graph): the first RDF graph to use for namespaces\n",
    "    violations_graph (Graph): the second RDF graph to use for namespaces\n",
    "    \"\"\"\n",
    "\n",
    "    # load violation_list\n",
    "    with open(VIOLATION_LIST_FILE, 'r') as f:\n",
    "        violation_list = json.load(f)\n",
    "\n",
    "    # read study_df tabularized data\n",
    "    study_df = pd.read_csv(STUDY_CSV_FILE, index_col=0)\n",
    "\n",
    "    # Create a new graph to combine namespaces\n",
    "    combined_graph = Graph()\n",
    "\n",
    "    # Get namespaces from both graphs and bind them to the combined graph\n",
    "    for prefix, ns_uri in study_graph.namespace_manager.namespaces():\n",
    "        combined_graph.namespace_manager.bind(prefix, ns_uri)\n",
    "\n",
    "    for prefix, ns_uri in violations_graph.namespace_manager.namespaces():\n",
    "        combined_graph.namespace_manager.bind(prefix, ns_uri)\n",
    "\n",
    "    # Now use the NamespaceManager of the combined graph\n",
    "    nsm_combined = NamespaceManager(combined_graph)\n",
    "\n",
    "    # Change column names\n",
    "    for col in study_df.columns:\n",
    "        try:\n",
    "            study_df.rename(columns={col: get_qname(nsm_combined, col)}, inplace=True)\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "    # Change indices and cell values\n",
    "    for col in study_df.columns:\n",
    "        for idx in study_df.index:\n",
    "            try:\n",
    "                study_df.loc[idx, col] = get_qname(nsm_combined, study_df.loc[idx, col])\n",
    "            except:\n",
    "                pass\n",
    "\n",
    "    # change index\n",
    "    for idx in study_df.index:\n",
    "        try:\n",
    "            study_df.rename(index={idx: get_qname(nsm_combined, str(idx))}, inplace=True)\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "    # change violation_list\n",
    "    for i in range(len(violation_list)):\n",
    "        try:\n",
    "            violation_list[i] = get_qname(nsm_combined, violation_list[i])\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "    # store tabular graph as csv\n",
    "    study_df.to_csv(STUDY_CSV_FILE)    \n",
    "\n",
    "    # store violation_list as json\n",
    "    with open(VIOLATION_LIST_FILE, 'w') as f:\n",
    "        json.dump(violation_list, f)\n",
    "\n",
    "abbreviate_using_namespaces(study_g, violations_g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "unique violations (18):\n",
      "\n",
      "omics:0c2600b5-30b2-40c7-917a-b402e73f55ae\n",
      "omics:TranscriptOmicsSampleShape-isMeasuredOnPlatform\n",
      "omics:TranscriptOmicsSampleShape-hasSequencingProtocol\n",
      "omics:OmicsStudyShape-hasContactScientist\n",
      "omics:OmicsStudyShape-isSequencedForTherapeuticArea\n",
      "omics:OmicsStudyShape-hasStudyType\n",
      "omics:afb51f95-5b17-45a4-b62d-c58f4998f930\n",
      "omics:38c0a1c7-1c92-489a-a22e-87f5335ccb1a\n",
      "omics:af998a22-f11b-43ea-b0ac-44e728baeafb\n",
      "omics:OmicsStudyShape-hasContactCompBio\n",
      "omics:OmicsStudyShape-hasResponsibleSite\n",
      "omics:TranscriptOmicsSampleShape-isPreparedByLibrary\n",
      "omics:ece15faa-30b3-471c-ae51-2d6d1f80e3a9\n",
      "omics:OmicsStudyShape-isSequencedForIndication\n",
      "omics:OmicsStudyShape-isPartOfProject\n",
      "omics:TranscriptOmicsSampleShape-hasStrandedness\n",
      "omics:248e9792-8549-4e0d-8665-06c42aff8ffc\n",
      "omics:9589ebf0-6f4f-4b80-9a5a-75aa25b5715f\n"
     ]
    }
   ],
   "source": [
    "with open(VIOLATION_LIST_FILE, 'r') as f:\n",
    "    violation_list = json.load(f)\n",
    "\n",
    "print('unique violations ('+str(len(violation_list))+'):\\n')\n",
    "print('\\n'.join([str(v) for v in violation_list]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>focus_node</th>\n",
       "      <th>omics:afb51f95-5b17-45a4-b62d-c58f4998f930</th>\n",
       "      <th>omics:0c2600b5-30b2-40c7-917a-b402e73f55ae</th>\n",
       "      <th>omics:248e9792-8549-4e0d-8665-06c42aff8ffc</th>\n",
       "      <th>omics:9589ebf0-6f4f-4b80-9a5a-75aa25b5715f</th>\n",
       "      <th>omics:38c0a1c7-1c92-489a-a22e-87f5335ccb1a</th>\n",
       "      <th>omics:af998a22-f11b-43ea-b0ac-44e728baeafb</th>\n",
       "      <th>omics:TranscriptOmicsSampleShape-hasSequencingProtocol</th>\n",
       "      <th>omics:TranscriptOmicsSampleShape-hasStrandedness</th>\n",
       "      <th>omics:TranscriptOmicsSampleShape-isMeasuredOnPlatform</th>\n",
       "      <th>...</th>\n",
       "      <th>omics:keywords</th>\n",
       "      <th>omics:isSequencedForIndication</th>\n",
       "      <th>omics:sourceStudyId</th>\n",
       "      <th>omics:externalStudyId</th>\n",
       "      <th>omics:studyId</th>\n",
       "      <th>omics:hasStudyType</th>\n",
       "      <th>omics:technicalStudyTitle</th>\n",
       "      <th>omics:hasContactScientist</th>\n",
       "      <th>omics:hasContactCompBio</th>\n",
       "      <th>omics:sequencingDate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/Donor/9abcf968-791a-4520-b650-2439b9237515</th>\n",
       "      <td>http://data.boehringer.com/uuid/Donor/9abcf968...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/Donor/4b2b6b89-7e27-4475-a0e6-e6638541de43</th>\n",
       "      <td>http://data.boehringer.com/uuid/Donor/4b2b6b89...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/PrimaryCellSpecimen/3c0bedfd-345a-4b5a-8528-abe0516a4232</th>\n",
       "      <td>http://data.boehringer.com/uuid/PrimaryCellSpe...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/PrimaryCellSpecimen/f5f9d112-0e74-4363-9b71-351a541172d9</th>\n",
       "      <td>http://data.boehringer.com/uuid/PrimaryCellSpe...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/TranscriptOmicsSample/sample-EX51-EX51_104</th>\n",
       "      <td>http://data.boehringer.com/uuid/TranscriptOmic...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                           focus_node  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...  http://data.boehringer.com/uuid/Donor/9abcf968...   \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...  http://data.boehringer.com/uuid/Donor/4b2b6b89...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  http://data.boehringer.com/uuid/PrimaryCellSpe...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  http://data.boehringer.com/uuid/PrimaryCellSpe...   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...  http://data.boehringer.com/uuid/TranscriptOmic...   \n",
       "\n",
       "                                                    omics:afb51f95-5b17-45a4-b62d-c58f4998f930  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                         1.0   \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                         1.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         0.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         0.0   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                         0.0   \n",
       "\n",
       "                                                    omics:0c2600b5-30b2-40c7-917a-b402e73f55ae  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                         0.0   \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                         0.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                         0.0   \n",
       "\n",
       "                                                    omics:248e9792-8549-4e0d-8665-06c42aff8ffc  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                         0.0   \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                         0.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                         0.0   \n",
       "\n",
       "                                                    omics:9589ebf0-6f4f-4b80-9a5a-75aa25b5715f  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                         0.0   \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                         0.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                         0.0   \n",
       "\n",
       "                                                    omics:38c0a1c7-1c92-489a-a22e-87f5335ccb1a  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                         0.0   \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                         0.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                         0.0   \n",
       "\n",
       "                                                    omics:af998a22-f11b-43ea-b0ac-44e728baeafb  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                         0.0   \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                         0.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         0.0   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                         0.0   \n",
       "\n",
       "                                                    omics:TranscriptOmicsSampleShape-hasSequencingProtocol  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                                0.0        \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                                0.0        \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0        \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0        \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                1.0        \n",
       "\n",
       "                                                    omics:TranscriptOmicsSampleShape-hasStrandedness  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                               0.0   \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                               0.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                               0.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                               0.0   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                               1.0   \n",
       "\n",
       "                                                    omics:TranscriptOmicsSampleShape-isMeasuredOnPlatform  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                                                0.0       \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                                                0.0       \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0       \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0       \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                1.0       \n",
       "\n",
       "                                                    ...  omics:keywords  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...  ...  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...  ...  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  ...  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  ...  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...  ...  EdgeNotPresent   \n",
       "\n",
       "                                                    omics:isSequencedForIndication  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...                  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...                  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                  EdgeNotPresent   \n",
       "\n",
       "                                                    omics:sourceStudyId  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...       EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...       EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...       EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...       EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...       EdgeNotPresent   \n",
       "\n",
       "                                                    omics:externalStudyId  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...         EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...         EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...         EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...         EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...         EdgeNotPresent   \n",
       "\n",
       "                                                     omics:studyId  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...  EdgeNotPresent   \n",
       "\n",
       "                                                    omics:hasStudyType  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...      EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...      EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...      EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...      EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...      EdgeNotPresent   \n",
       "\n",
       "                                                    omics:technicalStudyTitle  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...             EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...             EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...             EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...             EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...             EdgeNotPresent   \n",
       "\n",
       "                                                    omics:hasContactScientist  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...             EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...             EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...             EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...             EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...             EdgeNotPresent   \n",
       "\n",
       "                                                    omics:hasContactCompBio  \\\n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...           EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...           EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...           EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...           EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...           EdgeNotPresent   \n",
       "\n",
       "                                                   omics:sequencingDate  \n",
       "http://data.boehringer.com/uuid/Donor/9abcf968-...       EdgeNotPresent  \n",
       "http://data.boehringer.com/uuid/Donor/4b2b6b89-...       EdgeNotPresent  \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...       EdgeNotPresent  \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...       EdgeNotPresent  \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...       EdgeNotPresent  \n",
       "\n",
       "[5 rows x 64 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study_df = pd.read_csv(STUDY_CSV_FILE, index_col=0)\n",
    "study_df.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compute Violation Embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating UMAP embedding... (this may take some time)\n",
      "UMAP(n_neighbors=72, random_state=0, verbose=True)\n",
      "Thu Aug 24 10:14:00 2023 Construct fuzzy simplicial set\n",
      "Thu Aug 24 10:14:00 2023 Finding Nearest Neighbors\n",
      "Thu Aug 24 10:14:00 2023 Building RP forest with 9 trees\n",
      "Thu Aug 24 10:14:00 2023 NN descent for 12 iterations\n",
      "\t 1  /  12\n",
      "\tStopping threshold met -- exiting after 1 iterations\n",
      "Thu Aug 24 10:14:20 2023 Finished Nearest Neighbor Search\n",
      "Thu Aug 24 10:14:22 2023 Construct embedding\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e7c90107ac0540d6a9f1b7856896b866",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Epochs completed:   0%|            0/500 [00:00]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Thu Aug 24 10:15:26 2023 Finished embedding\n",
      "UMAP embedding created and saved to ./csv/study.csv\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAi0AAAGdCAYAAADey0OaAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/NK7nSAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAmS0lEQVR4nO3de3hU9b3v8U8SciWZwUAupAQIgkDkJgghKtZLJFzqKQXdaLcQLMcqBjwY1JpWoLVAKDxHW1sVj7sCpz2IdXdjN4oojRK0BrBgdotcBBpMbDIJSpmBCLnNOn+wnRIJkITMrPkx79fzzPPImsla38UA83bNmjVhlmVZAgAACHLhdg8AAADQFkQLAAAwAtECAACMQLQAAAAjEC0AAMAIRAsAADAC0QIAAIxAtAAAACN0sXuAS+X1elVVVaWEhASFhYXZPQ4AAGgDy7J04sQJpaWlKTy8bcdQjI+Wqqoqpaen2z0GAADogMrKSvXq1atNjzU+WhISEiSd2WmHw2HzNAAAoC08Ho/S09N9r+NtYXy0fPWWkMPhIFoAADBMe07t4ERcAABgBKIFAAAYgWgBAABGIFoAAIARiBYAAGAEogUAABiBaAEAAEYgWgAAgBGMv7gcAJis2WtpZ/kx1Z44reSEGI3JSFREON+jBrSGaAGAS9DQ5NVL7/1N//HRZ/Kcbla/HnG6/8YrdcNVSReMj2avpbznN+n9ypbLEyKklXeP1IQhPf08OWCeMMuyLLuHuBQej0dOp1Nut5vL+AMIqKJNe/XCtvLz3v/0ncP1nVG91NDk1fPvHNKq9w7pVGPb/slddQ/hgstbR16/iRYA6ICLBUtnOLxsEm8V4bLVkddvTsQFgHZqaPLqxff8GyyS9KPXdvt9G4BJiBYAaKfflB6RNwDHqNfvdKk5EBsCDEG0AEA7fXrsy4Bta2f5sYBtCwh2RAsAtFOfxLiAbav2xOmAbQsIdkQLALTTjOy+AdtWjDgRF/gK0QIA7RTVJVzf7BWYfz4XbfxrQLYDmIBoAYAOWDt3YkC2U1PXFJDtACYgWgCggw4vm2T3CEBIIVoAoIMC8cmeft1j/b4NwBRECwB0UCA+2fPvc27w+zYAUxAtANBByQkxfl1/UnyUEuOj/LoNwCRECwB00JiMRCVE+GfdSfFR+vCJ2/yzcsBQRAsAdFBEeJhW3j2y09f7i6lDCRagFUQLAFyCCUN6atU9nRsu37o2vVPXB1wuiBYAuEQThvTU4WWTtHJKZovl4ZJu6H+FZoxte4SsumekIsK5Ci7Qmi7+XPnzzz+v559/XkeOHJEkXX311Vq0aJEmTjxzUabTp09rwYIFWr9+verr65Wbm6vnnntOKSkp/hwLADpdRHiY7hyboTvHZrR6//X9k/TQ+jI1NHnPu45V94zUhCE9/TUiYLwwy7L89r3nGzduVEREhAYMGCDLsrR27VqtXLlSH330ka6++mrNmTNHb7zxhtasWSOn06m5c+cqPDxcf/rTn9q8DY/HI6fTKbfbLYfD4a9dAYBL1uy19P6Bo3rmnU+0q9It6czRmKemDdPto3pxhAUhpSOv336NltYkJiZq5cqVuuOOO5SUlKR169bpjjvukCTt379fgwcPVmlpqcaOHdum9REtAACYpyOv3wE7p6W5uVnr169XXV2dsrOztWvXLjU2NionJ8f3mEGDBql3794qLS0973rq6+vl8Xha3AAAwOXP79Hy17/+VfHx8YqOjtYDDzygDRs2KDMzUy6XS1FRUerWrVuLx6ekpMjlcp13fUVFRXI6nb5bejpn2QMAEAr8Hi0DBw5UWVmZduzYoTlz5igvL0979+7t8PoKCwvldrt9t8rKyk6cFgAABCu/fnpIkqKiotS/f39J0qhRo/Thhx/qF7/4haZPn66GhgYdP368xdGWmpoapaamnnd90dHRio6O9vfYAAAgyAT8Oi1er1f19fUaNWqUIiMjVVxc7LvvwIEDqqioUHZ2dqDHAgAAQc6vR1oKCws1ceJE9e7dWydOnNC6deu0detWvfXWW3I6nZo9e7YKCgqUmJgoh8OhefPmKTs7u82fHAIAAKHDr9FSW1urmTNnqrq6Wk6nU8OGDdNbb72l2247850aTz/9tMLDwzVt2rQWF5cDAAD4uoBfp6WzcZ0WAADME9TXaQEAALgURAsAADAC0QIAAIxAtAAAACMQLQAAwAhECwAAMALRAgAAjEC0AAAAIxAtAADACEQLAAAwAtECAACMQLQAAAAjEC0AAMAIRAsAADAC0QIAAIxAtAAAACMQLQAAwAhECwAAMALRAgAAjEC0AAAAIxAtAADACEQLAAAwAtECAACMQLQAAAAjEC0AAMAIRAsAADAC0QIAAIxAtAAAACMQLQAAwAhECwAAMALRAgAAjEC0AAAAIxAtAADACEQLAAAwAtECAACMQLQAAAAjEC0AAMAIRAsAADAC0QIAAIxAtAAAACMQLQAAwAhECwAAMALRAgAAjEC0AAAAIxAtAADACEQLAAAwAtECAACMQLQAAAAjEC0AAMAIRAsAADAC0QIAAIxAtAAAACMQLQAAwAhECwAAMALRAgAAjEC0AAAAIxAtAADACEQLAAAwAtECAACMQLQAAAAjEC0AAMAIfo2WoqIijR49WgkJCUpOTtaUKVN04MCBFo85ffq08vPz1b17d8XHx2vatGmqqanx51gAAMBAfo2WkpIS5efna/v27dqyZYsaGxs1fvx41dXV+R7z8MMPa+PGjXr11VdVUlKiqqoqTZ061Z9jAQAAA4VZlmUFamNHjx5VcnKySkpKdOONN8rtdispKUnr1q3THXfcIUnav3+/Bg8erNLSUo0dO/ai6/R4PHI6nXK73XI4HP7eBQAA0Ak68vod0HNa3G63JCkxMVGStGvXLjU2NionJ8f3mEGDBql3794qLS1tdR319fXyeDwtbgAA4PIXsGjxer2aP3++rr/+eg0ZMkSS5HK5FBUVpW7durV4bEpKilwuV6vrKSoqktPp9N3S09P9PToAAAgCAYuW/Px87dmzR+vXr7+k9RQWFsrtdvtulZWVnTQhAAAIZl0CsZG5c+fq9ddf17Zt29SrVy/f8tTUVDU0NOj48eMtjrbU1NQoNTW11XVFR0crOjra3yMDAIAg49cjLZZlae7cudqwYYPeeecdZWRktLh/1KhRioyMVHFxsW/ZgQMHVFFRoezsbH+OBgAADOPXIy35+flat26d/vCHPyghIcF3norT6VRsbKycTqdmz56tgoICJSYmyuFwaN68ecrOzm7TJ4cAAEDo8OtHnsPCwlpdvnr1as2aNUvSmYvLLViwQC+//LLq6+uVm5ur55577rxvD30dH3kGAMA8HXn9Duh1WvyBaAEAwDxBf50WAACAjiJaAACAEYgWAABgBKIFAAAYgWgBAABGIFoAAIARiBYAAGAEogUAABiBaAEAAEYgWgAAgBGIFgAAYASiBQAAGIFoAQAARiBaAACAEYgWAABgBKIFAAAYgWgBAABGIFoAAIARiBYAAGAEogUAABiBaAEAAEYgWgAAgBGIFgAAYASiBQAAGIFoAQAARiBaAACAEYgWAABgBKIFAAAYgWgBAABGIFoAAIARiBYAAGAEogUAABiBaAEAAEYgWgAAgBGIFgAAYIQudg9gsmavpe1/+0Klh7+QZCm7Xw+NvbK7IsLD7B4NAIDLDtHSQZv+UqUH133UYtmv3j0sSXruu9do0rC0Fvc1ey29s8eluet2q/6s5etmZem6QT38PS4AAMYLsyzLsnuIS+HxeOR0OuV2u+VwOAKyzaJNe/XCtvILPub+GzNUOClTkrR5T7Xm/Ha3LvQbfWT55E6cEACA4NaR12/OaWmnTX+pvmiwSNIL28q16S/V2rynWg9cJFgkqe/jb3TOgAAAXKY40nIeDU1erf2gXKWHjupj10lFhodrUGq8/rj/aKdt4+smD3bo2bxxfls/AADBoiOv35zT0orzvf3z2fFTft3uG/s8errJq6guHAADAODreHX8mracr+JPvyk9Ytu2AQAIZkTLWRqavLYGiyR9euxLW7cPAECwIlrO8n9KDtk9gurqm+weAQCAoMQ5LWdZEwRvzfx+99+VFB+tcQOSuFAdAABnIVrOUt8UHB+kWrXtb1q17W/qFttFy6cN04QhPe0eCQAA2/H20FkGJHW1e4QWjp9q0gO/3a3Ne6rtHgUAANsRLWdZfW+W3SO06qGXP1KzNziOAgEAYBei5SzOuMig/A1paLb0/kH/XdQOAAATBONrtG2OnWyQ1+4hzuPF9/5m9wgAANiKaDnLv7zwgd0jnJf7VKPdIwAAYCui5SzBfGG3Yb2cdo8AAICtiJazBPM1UZ6YfLXdIwAAYCui5SyTh6bYPUKrbstMVmxUhN1jAABgK6LlLEumDLd7hHPclpmsF2eOtnsMAABsxxVxzxIbFaHbMpO1ZW+t3aMoUtJfnpzAERYAAP4bR1q+5sWZo3VbZrJt2+/RNVKfLJmog8snEywAAJyFaGmFnW/HzB7XT1FdeFoAAPg6Xh3PY1BSjC3bnX1DP1u2CwBAsCNazuOVOTcGfJv335jBURYAAM6DV8jzcMZFqk/32E5b333jMvQvY1LPe//9N2aocFJmp20PAIDLjV+jZdu2bbr99tuVlpamsLAwvfbaay3utyxLixYtUs+ePRUbG6ucnBwdPHjQnyO1S8mjt1wwXLqESavuGalhvRwXXM9z3x2pH03O1Iqpo3R42ST95ntj9J0RaRqfmaIfTRqkT5ZMJFgAALgIv37kua6uTsOHD9f3vvc9TZ069Zz7V6xYoWeeeUZr165VRkaGFi5cqNzcXO3du1cxMfacU/J1JY/eIveXjcpbvV0HXCckhWn4Nxyac9MA3XBVkiLCwzRhSE+dPN2kB//fhyo9fExeS+oeF6ml3x6qW4aktrjSbkR4mMZdlaRxVyXZt1MAABgozLIsKyAbCgvThg0bNGXKFElnjrKkpaVpwYIFeuSRRyRJbrdbKSkpWrNmje666642rdfj8cjpdMrtdsvhuPARDwAAEBw68vpt2zkt5eXlcrlcysnJ8S1zOp3KyspSaWnpeX+uvr5eHo+nxQ0AAFz+bIsWl8slSUpJafl9PykpKb77WlNUVCSn0+m7paen+3VOAAAQHIz79FBhYaHcbrfvVllZafdIAAAgAGyLltTUMx//rampabG8pqbGd19roqOj5XA4WtwAAMDlz7ZoycjIUGpqqoqLi33LPB6PduzYoezsbLvGAgAAQcqvH3k+efKkDh065Pt1eXm5ysrKlJiYqN69e2v+/PlasmSJBgwY4PvIc1pamu8TRgAA4PyavZZ2lh+Ty31Kx+oalBgfrVRHjMZkJLa43Mblwq/R8uc//1k333yz79cFBQWSpLy8PK1Zs0aPPfaY6urq9P3vf1/Hjx/XDTfcoM2bNwfNNVoAALDT34+d0sRnSnTidLO6hEvfvylD/+uWQYrqEq7/+LBCBb//a6s/l5IQrduG9NBvS//eYvnd2T205PYxxgZNwK7T4i9cpwUAcDka8MNNavS2/hIdGxmuU43eDq971T0jNWFIzw7/fGfoyOu3X4+0AACA9utX+IbO0yuSdEnBIkkP/Ha3nvvuSF3RNUrVx0/po8p/yJKU0b2rZmT3Ddov7yVaAAAIIoX/XnbBYOksD67b3erypZv26b5xwfklvkQLAABB4OTpJs18sUS7/37a1jm8lvTCtnJJCrpwIVoAALDZ//jVe/rLZ8H1tTQvvleuBeMHBdVbRcEzCQAAISgYg0U6c8Tlpff+ZvcYLXCkBQCAAHN/2ahZq3dob5Vb9c12T3N+y986oF2V/9CLM0fbPYokogUAgIAa97M/qvIf9XaP0WZb9tbqvv/7YVCEC9ECAECADF745iV/XNkOW/bW6lRDs2KjImydg3NaAAAIgLyXdhgZLF8Z/+Rmu0cgWgAA8LdTDc0q+eRzu8e4JJVNZ77ryE5ECwAAfrZs0167R+gUO8uP2bp9ogUAAD878sWXdo/QKWpP2HvhO6IFAAA/69s9zu4ROkVyQoyt2ydaAADwsx8G2eXwO+ruF7fbun2iBQAAP4uNitBtmcl2j9Ep+j7+hm3bJloAAAiAF2eOvmzC5Xfv23N5f6IFAIAAeXHmaK3Ns//Kspfqsdf32bJdogUAgAC6YWCSutp7YVljES0AAARQRHiYRveKtHsMIxEtAAAE2NZPG+0e4ZKs+NZgW7ZLtAAAgHb5lxv62bJdogUAALTZkeWTbds20QIAQIA9N3W43SO0m1P2BoskdbF16wAAhKBJY3pJ//Ffdo/RqmG9HFr3P7P18CsfqeIfp9T7ilg9Pf0axcfYnwxhlmXZ+z3Tl8jj8cjpdMrtdsvhcNg9DgAAbWbn1WVb84vpI/Tta74RkG115PXb/mwCACBEHVk+WZt2fqYHv3bUZcLAeG0+cNLv23/wm1cqPDxM2Vd219h+3RURHub3bV4KogUAABtNGtNLR8b0Omf55j3Vemj9R2po8s8bIvffmKHHJg7yy7r9hRNxAQAIQhOG9NS+Jydqbd5ojegZ51seGS79cf43dWT5ZH2yZKLm3XyluoT/876HvtlPnyyZqF/dNeK8677/xgwVGvjN05zTAgDAZarZa+mDg5/r9x99pi8bmjW6b6LyruurqC72H7PgnBYAAOATER6mcQOTNG5gkt2jdAr7UwsAAKANiBYAAGAEogUAABiBaAEAAEYgWgAAgBGIFgAAYASiBQAAGIFoAQAARiBaAACAEYgWAABgBKIFAAAYgWgBAABGIFoAAIARiBYAAGAEogUAABiBaAEAAEYgWgAAgBGIFgAAYASiBQAAGIFoAQAARiBaAACAEYgWAABgBKIFAAAYgWgBAABGIFoAAIARiBYAAGAEogUAABiBaAEAAEYgWgAAgBGIFgAAYASiBQAAGCEoouXZZ59V3759FRMTo6ysLO3cudPukQAAQJCxPVpeeeUVFRQUaPHixdq9e7eGDx+u3Nxc1dbW2j0aAAAIIrZHy1NPPaX77rtP9957rzIzM7Vq1SrFxcXppZdesns0AAAQRGyNloaGBu3atUs5OTm+ZeHh4crJyVFpaWmrP1NfXy+Px9PiBgAALn+2Rsvnn3+u5uZmpaSktFiekpIil8vV6s8UFRXJ6XT6bunp6YEYFQAA2Mz2t4faq7CwUG6323errKy0eyQAABAAXezceI8ePRQREaGampoWy2tqapSamtrqz0RHRys6OjoQ4wEAgCBi65GWqKgojRo1SsXFxb5lXq9XxcXFys7OtnEyAAAQbGw90iJJBQUFysvL07XXXqsxY8bo5z//uerq6nTvvffaPRoAAAgitkfL9OnTdfToUS1atEgul0sjRozQ5s2bzzk5FwAAhLYwy7Isu4e4FB6PR06nU263Ww6Hw+5xAABAG3Tk9du4Tw8BAIDQRLQAAAAjEC0AAMAIRAsAADAC0QIAAIxAtAAAACMQLQAAwAhECwAAMALRAgAAjEC0AAAAIxAtAADACEQLAAAwAtECAACMQLQAAAAjEC0AAMAIRAsAADAC0QIAAIxAtAAAACMQLQAAwAhECwAAMALRAgAAjEC0AAAAIxAtAADACEQLAAAwAtECAACMQLQAAAAjEC0AAMAIRAsAADAC0QIAAIxAtAAAACMQLQAAwAhECwAAMALRAgAAjEC0AAAAIxAtAADACEQLAAAwAtECAACMQLQAAAAjEC0AAMAIRAsAADAC0QIAAIxAtAAAACMQLQAAwAhECwAAMALRAgAAjEC0AAAAIxAtAADACEQLAAAwAtECAACMQLQAAAAjEC0AAMAIRAsAADAC0QIAAIxAtAAAACMQLQAAwAhECwAAMALRAgAAjEC0AAAAIxAtAADACEQLAAAwAtECAACM4LdoWbp0qa677jrFxcWpW7durT6moqJCkydPVlxcnJKTk/Xoo4+qqanJXyMBAACDdfHXihsaGnTnnXcqOztbv/71r8+5v7m5WZMnT1Zqaqo++OADVVdXa+bMmYqMjNSyZcv8NRYAADBUmGVZlj83sGbNGs2fP1/Hjx9vsfzNN9/Ut771LVVVVSklJUWStGrVKv3gBz/Q0aNHFRUV1ab1ezweOZ1Oud1uORyOzh4fAAD4QUdev207p6W0tFRDhw71BYsk5ebmyuPx6OOPPz7vz9XX18vj8bS4AQCAy59t0eJyuVoEiyTfr10u13l/rqioSE6n03dLT0/365wAACA4tCtaHn/8cYWFhV3wtn//fn/NKkkqLCyU2+323SorK/26PQAAEBzadSLuggULNGvWrAs+pl+/fm1aV2pqqnbu3NliWU1Nje++84mOjlZ0dHSbtgGgY5q9lore+i/9W8nfWyz/39+5WlNG91FEeJhNkwEIZe2KlqSkJCUlJXXKhrOzs7V06VLV1tYqOTlZkrRlyxY5HA5lZmZ2yjYAtN/mPdV64Le7W71vwYaPtWDDx/rl3dfo9uFp7VpvQ5NXa/5Urrf3nnn797bMVN17fYaiunC5KABt47ePPFdUVOjYsWOqqKhQc3OzysrKJEn9+/dXfHy8xo8fr8zMTM2YMUMrVqyQy+XSE088ofz8fI6kADa5ULCcbd7LH+k//+vvenHmaDV7LT31dpme3VrV4jHTxybrp98apagu4Xpy48d66U9HWtz/50+Pq+jN/br/xgwVTuJ/VABcnN8+8jxr1iytXbv2nOXvvvuubrrpJknSp59+qjlz5mjr1q3q2rWr8vLytHz5cnXp0vaW4iPPQOdo9lq68oeb2vUzOYOT9Md9Ry/4GEdMF3lOX/iikYQLEHo68vrt9+u0+BvRAnSO0sNf6O4Xt9u2/U+WTOStIiCEGHWdFgDBpfbEaVu3//BvS23dPoDgR7QA0FFPvR5bX2brDG/sP66+j7+ht/9cdfEHAwhJvD0EhLhhP37rouec2OHI8sl2jwDAj3h7CEC7BGuwSFLfx9+wewQAQYZoAULUUU990AbLV5ZsuvjHrwGEDqIFCFHf/lWJ3SNc1L9tq9b3f/2OGpq8do8CIAgQLUCIqvY02j1Cm7x98JSueuJNFW3aa/coAGxGtAAwwgvbygkXIMQRLUCI6uWMsnuEdnthWzlvFQEhjGgBQlCz19Kxkw12j9Ehvyk9YvcIAGxCtAAhaGf5MdU12z1Fx3x67Eu7RwBgE6IFCEF2X7L/UvRJjLN7BAA2IVqAEJScEGP3CB02I7uv3SMAsAnRAoQg9zEzj7TcNy6Db4IGQhh/+4EQ9MDvy+weod2uSo7XjyZn2j0GABsRLQCMMPbK7naPAMBmRAsAI3ACLgCiBQhBsXYP0AGcgAuAaAFC0LuP32r3CO3yzd5dOAEXANEChKLUbjGKjTTnr//aB3PtHgFAEDDnXy0AnWrfTycaES5Hlk+2ewQAQSL4/8UC4Df7fjpR2x+/Vd3jIu0epVVzb+5v9wgAggjRAoS41G4x2rVovCLDw+we5RzZfMwZwFmIFgCSJMuy7B7hHGP7ES0A/oloASBJ6tM9uD4IveqekYoIwqM/AOxDtACQJL36wA0B2c6eH+cqMS7igo9Zdc9ITRjSMyDzADBHF7sHABAcEuOjlBQfpaMnGy762PtvzFDhpH9+D1DRpr16YVv5RX9uWC+H4mO6aPeiCTp2skE3Ldkiz1n359+SroKcoRxhAdCqMCsY38huB4/HI6fTKbfbLYfDYfc4gPFGL9lywXB57rsjNWnYuUdBLhYuw3o59J9zx3XKjADM15HXb6IFwDmOnWzQbUVb9EXzP5cVjs/Q/7xp8AWPgjQ0efWD3+3Qhr8c8y0b3E16dX6u4mM4sAvgn4gWogUAACN05PWbE3EBAIARiBYAAGAEogUAABiBaAEAAEYgWgAAgBGIFgAAYASiBQAAGIFoAQAARiBaAACAEYy/rvZXF/T1eDwXeSQAAAgWX71ut+fC/MZHy4kTJyRJ6enpNk8CAADa68SJE3I6nW16rPHfPeT1elVVVaWEhASFhZn9dfYej0fp6emqrKwMue9RYt9Db99Ddb8l9j0U9z1U91s6/75blqUTJ04oLS1N4eFtO1vF+CMt4eHh6tWrl91jdCqHwxFyf6i/wr6H3r6H6n5L7Hso7nuo7rfU+r639QjLVzgRFwAAGIFoAQAARiBagkh0dLQWL16s6Ohou0cJOPY99PY9VPdbYt9Dcd9Ddb+lzt1340/EBQAAoYEjLQAAwAhECwAAMALRAgAAjEC0AAAAIxAtQWLp0qW67rrrFBcXp27durX6mLCwsHNu69evD+ygftCWfa+oqNDkyZMVFxen5ORkPfroo2pqagrsoAHQt2/fc57j5cuX2z2WXzz77LPq27evYmJilJWVpZ07d9o9kt/9+Mc/Puf5HTRokN1j+cW2bdt0++23Ky0tTWFhYXrttdda3G9ZlhYtWqSePXsqNjZWOTk5OnjwoD3DdqKL7fesWbPO+TMwYcIEe4btREVFRRo9erQSEhKUnJysKVOm6MCBAy0ec/r0aeXn56t79+6Kj4/XtGnTVFNT067tEC1BoqGhQXfeeafmzJlzwcetXr1a1dXVvtuUKVMCM6AfXWzfm5ubNXnyZDU0NOiDDz7Q2rVrtWbNGi1atCjAkwbGk08+2eI5njdvnt0jdbpXXnlFBQUFWrx4sXbv3q3hw4crNzdXtbW1do/md1dffXWL5/f999+3eyS/qKur0/Dhw/Xss8+2ev+KFSv0zDPPaNWqVdqxY4e6du2q3NxcnT59OsCTdq6L7bckTZgwocWfgZdffjmAE/pHSUmJ8vPztX37dm3ZskWNjY0aP3686urqfI95+OGHtXHjRr366qsqKSlRVVWVpk6d2r4NWQgqq1evtpxOZ6v3SbI2bNgQ0HkC6Xz7vmnTJis8PNxyuVy+Zc8//7zlcDis+vr6AE7of3369LGefvppu8fwuzFjxlj5+fm+Xzc3N1tpaWlWUVGRjVP53+LFi63hw4fbPUbAff3fLq/Xa6WmplorV670LTt+/LgVHR1tvfzyyzZM6B+t/Zudl5dnffvb37ZlnkCqra21JFklJSWWZZ15fiMjI61XX33V95h9+/ZZkqzS0tI2r5cjLYbJz89Xjx49NGbMGL300kvt+kpvU5WWlmro0KFKSUnxLcvNzZXH49HHH39s42T+sXz5cnXv3l3XXHONVq5cedm9DdbQ0KBdu3YpJyfHtyw8PFw5OTkqLS21cbLAOHjwoNLS0tSvXz/967/+qyoqKuweKeDKy8vlcrla/BlwOp3KysoKiT8DW7duVXJysgYOHKg5c+boiy++sHukTud2uyVJiYmJkqRdu3apsbGxxXM+aNAg9e7du13PufFfmBhKnnzySd1yyy2Ki4vT22+/rQcffFAnT57UQw89ZPdofuVyuVoEiyTfr10ulx0j+c1DDz2kkSNHKjExUR988IEKCwtVXV2tp556yu7ROs3nn3+u5ubmVp/T/fv32zRVYGRlZWnNmjUaOHCgqqur9ZOf/ETjxo3Tnj17lJCQYPd4AfPV39vW/gxcbn+nv27ChAmaOnWqMjIydPjwYf3whz/UxIkTVVpaqoiICLvH6xRer1fz58/X9ddfryFDhkg685xHRUWdc95ie59zosWPHn/8cf3sZz+74GP27dvX5hPxFi5c6Pvva665RnV1dVq5cmVQRktn77vJ2vN7UVBQ4Fs2bNgwRUVF6f7771dRUVFIXv77cjNx4kTffw8bNkxZWVnq06ePfve732n27Nk2ToZAueuuu3z/PXToUA0bNkxXXnmltm7dqltvvdXGyTpPfn6+9uzZ45fztYgWP1qwYIFmzZp1wcf069evw+vPysrST3/6U9XX1wfdC1pn7ntqauo5nyz56ozz1NTUDs0XSJfye5GVlaWmpiYdOXJEAwcO9MN0gdejRw9FRESc86mBmpoaI57PztStWzddddVVOnTokN2jBNRXz3NNTY169uzpW15TU6MRI0bYNJU9+vXrpx49eujQoUOXRbTMnTtXr7/+urZt26ZevXr5lqempqqhoUHHjx9vcbSlvX/viRY/SkpKUlJSkt/WX1ZWpiuuuCLogkXq3H3Pzs7W0qVLVVtbq+TkZEnSli1b5HA4lJmZ2Snb8KdL+b0oKytTeHi4b78vB1FRURo1apSKi4t9n37zer0qLi7W3Llz7R0uwE6ePKnDhw9rxowZdo8SUBkZGUpNTVVxcbEvUjwej3bs2HHRT1Bebj777DN98cUXLeLNRJZlad68edqwYYO2bt2qjIyMFvePGjVKkZGRKi4u1rRp0yRJBw4cUEVFhbKzs9u8HaIlSFRUVOjYsWOqqKhQc3OzysrKJEn9+/dXfHy8Nm7cqJqaGo0dO1YxMTHasmWLli1bpkceecTewTvBxfZ9/PjxyszM1IwZM7RixQq5XC498cQTys/PD8pg66jS0lLt2LFDN998sxISElRaWqqHH35Y99xzj6644gq7x+tUBQUFysvL07XXXqsxY8bo5z//uerq6nTvvffaPZpfPfLII7r99tvVp08fVVVVafHixYqIiNDdd99t92id7uTJky2OIJWXl6usrEyJiYnq3bu35s+fryVLlmjAgAHKyMjQwoULlZaWZvxlHC6034mJifrJT36iadOmKTU1VYcPH9Zjjz2m/v37Kzc318apL11+fr7WrVunP/zhD0pISPCdp+J0OhUbGyun06nZs2eroKBAiYmJcjgcmjdvnrKzszV27Ni2b6iTP+WEDsrLy7MknXN79913LcuyrDfffNMaMWKEFR8fb3Xt2tUaPny4tWrVKqu5udnewTvBxfbdsizryJEj1sSJE63Y2FirR48e1oIFC6zGxkb7hvaDXbt2WVlZWZbT6bRiYmKswYMHW8uWLbNOnz5t92h+8ctf/tLq3bu3FRUVZY0ZM8bavn273SP53fTp062ePXtaUVFR1je+8Q1r+vTp1qFDh+weyy/efffdVv9e5+XlWZZ15mPPCxcutFJSUqzo6Gjr1ltvtQ4cOGDv0J3gQvv95ZdfWuPHj7eSkpKsyMhIq0+fPtZ9993X4nIOpmptnyVZq1ev9j3m1KlT1oMPPmhdccUVVlxcnPWd73zHqq6ubtd2wv57YwAAAEGN67QAAAAjEC0AAMAIRAsAADAC0QIAAIxAtAAAACMQLQAAwAhECwAAMALRAgAAjEC0AAAAIxAtAADACEQLAAAwAtECAACM8P8BNV/xFoCp5RIAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "import umap\n",
    "import pandas as pd\n",
    "import json\n",
    "import numpy as np\n",
    "\n",
    "from scipy.sparse import coo_matrix\n",
    "\n",
    "def create_embedding():\n",
    "    # read study.csv\n",
    "    study_df = pd.read_csv(STUDY_CSV_FILE, index_col=0)\n",
    "    \n",
    "    # read violation_list.json\n",
    "    with open(VIOLATION_LIST_FILE, 'r') as f:\n",
    "        violation_columns = json.load(f)\n",
    "    \n",
    "    view_df = study_df[violation_columns]  # view of study dataframe with only violation columns\n",
    "\n",
    "    print(\"Creating UMAP embedding... (this may take some time)\")\n",
    "    reducer = umap.UMAP(n_neighbors=int(np.sqrt(len(view_df))), min_dist=0.1, n_components=2, random_state=0, verbose=True)\n",
    "    result = reducer.fit_transform(view_df)  # 2D projection\n",
    "\n",
    "    # Check if the result is a tuple and unpack accordingly\n",
    "    if isinstance(result, tuple):\n",
    "        embedding, *_ = result\n",
    "    else:\n",
    "        embedding = result\n",
    "\n",
    "    # Convert to dense numpy array if it's a sparse matrix\n",
    "    if isinstance(embedding, coo_matrix):\n",
    "        embedding = embedding.toarray()\n",
    "\n",
    "    # Assuming that the embedding is now a NumPy array, based on the docstring\n",
    "    study_df[\"x\"] = embedding[:, 0]\n",
    "    study_df[\"y\"] = embedding[:, 1]\n",
    "\n",
    "    study_df.to_csv(STUDY_CSV_FILE)\n",
    "    print(\"UMAP embedding created and saved to\", STUDY_CSV_FILE)\n",
    "\n",
    "\n",
    "def plot_embedding():\n",
    "    study_df = pd.read_csv(STUDY_CSV_FILE, index_col=0)\n",
    "\n",
    "    plt.scatter(study_df[\"x\"], study_df[\"y\"])\n",
    "    plt.show()\n",
    "\n",
    "create_embedding()\n",
    "plot_embedding()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Replace URIs with Label Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "from rdflib import URIRef\n",
    "\n",
    "# list of label predicates in ascending order of priority\n",
    "label_predicates = [\n",
    "    # URIRef('http://data.boehringer.com/ontology/omics/hasValue'),\n",
    "    # URIRef('http://data.boehringer.com/ontology/omics/name'),\n",
    "    URIRef('http://www.w3.org/2004/02/skos/core#prefLabel')\n",
    "]\n",
    "\n",
    "# read tabularized graph .csv as a dataframe, and violations list .json as a list\n",
    "study_df = pd.read_csv(STUDY_CSV_FILE, index_col=0)\n",
    "\n",
    "with open(VIOLATION_LIST_FILE, 'r') as f:\n",
    "    violation_columns = json.load(f)\n",
    "\n",
    "# Create a new graph to combine namespaces\n",
    "combined_graph = Graph()\n",
    "\n",
    "# Get namespaces from both graphs and bind them to the combined graph\n",
    "for prefix, ns_uri in study_g.namespace_manager.namespaces():\n",
    "    combined_graph.namespace_manager.bind(prefix, ns_uri)\n",
    "\n",
    "for prefix, ns_uri in violations_g.namespace_manager.namespaces():\n",
    "    combined_graph.namespace_manager.bind(prefix, ns_uri)\n",
    "\n",
    "# Now use the NamespaceManager of the combined graph\n",
    "nsm_combined = NamespaceManager(combined_graph)\n",
    "\n",
    "# create a dictionary of {s: o} pairs for translating source shapes to their labels\n",
    "label_dict = {}\n",
    "for label_predicate in label_predicates:\n",
    "    temp_dict = {str(s): str(o) for s, p, o in study_g.triples((None, label_predicate, None))}\n",
    "    # replace the temp_dict keys with their corresponding QNames\n",
    "    temp_dict = {get_qname(nsm_combined, k): v for k, v in temp_dict.items()}\n",
    "    # Update label_dict with temp_dict, overwriting existing keys\n",
    "    label_dict.update(temp_dict)\n",
    "\n",
    "# replace all column names, indices, and cell values in the dataframe with their corresponding labels\n",
    "study_df.columns = [label_dict.get(col, col) for col in study_df.columns]\n",
    "study_df.index = pd.Index([label_dict.get(idx, idx) for idx in study_df.index])\n",
    "for col in study_df.columns:\n",
    "    study_df[col] = study_df[col].apply(lambda x: label_dict.get(x, x))\n",
    "\n",
    "# replace all dictionary keys and values with their corresponding labels\n",
    "violation_columns = [label_dict.get(x, x) for x in violation_columns]\n",
    "\n",
    "# store dataframe back to .csv and violations list back to .json\n",
    "study_df.to_csv(STUDY_CSV_FILE)\n",
    "with open(VIOLATION_LIST_FILE, 'w') as f:\n",
    "    json.dump(violation_columns, f)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create Union of Ontology and Violation Report Exemplars"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rdflib import URIRef, RDFS\n",
    "\n",
    "def replace_uris_with_labels(data, label_dict, data_type=\"graph\"):\n",
    "    if data_type == \"graph\":\n",
    "        # Iterate through each triple\n",
    "        for s, p, o in list(data.triples((None, None, None))):\n",
    "            new_s = URIRef(label_dict.get(str(s), s))\n",
    "            new_o = URIRef(label_dict.get(str(o), o))\n",
    "            \n",
    "            if new_s != s or new_o != o:\n",
    "                data.remove((s, p, o))\n",
    "                data.add((new_s, p, new_o))\n",
    "                \n",
    "    elif data_type == \"dict\":\n",
    "        new_data = {}\n",
    "        for key, values in data.items():\n",
    "            new_key = label_dict.get(key, key)\n",
    "            \n",
    "            # If the values are in set form\n",
    "            if isinstance(values, set):\n",
    "                new_values = {label_dict.get(val, val) for val in values}\n",
    "                \n",
    "            # If the values are in dictionary form with concatenated string keys\n",
    "            elif isinstance(values, dict):\n",
    "                new_values = {}\n",
    "                for po_str, count in values.items():\n",
    "                    if '__' in po_str:\n",
    "                        p, o = po_str.split('__')\n",
    "                        new_p = label_dict.get(p, p)\n",
    "                        new_o = label_dict.get(o, o)\n",
    "                        new_po_str = f\"{new_p}__{new_o}\"\n",
    "                        new_values[new_po_str] = count\n",
    "                    else:\n",
    "                        new_po_str = label_dict.get(po_str, po_str)\n",
    "                        new_values[new_po_str] = count\n",
    "            \n",
    "            new_data[new_key] = new_values\n",
    "            \n",
    "        return new_data\n",
    "    \n",
    "    return data\n",
    "\n",
    "\n",
    "\n",
    "def add_labels_to_graph(graph, label_dict):\n",
    "    for uri, label in label_dict.items():\n",
    "        graph.add((URIRef(uri), URIRef('http://www.w3.org/2004/02/skos/core#prefLabel'), Literal(label)))\n",
    "    return graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c215499eab8b4b92bf617cf86f1a2b4c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Processing violations:   0%|          | 0/17911 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from rdflib import Graph\n",
    "import json\n",
    "from routers.utils import get_violation_report_exemplars\n",
    "\n",
    "ontology_union_violation_exemplars_g, edge_count_dict, focus_node_exemplar_dict, exemplar_focus_node_dict = get_violation_report_exemplars(ontology_g, violations_g)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Also replace in joined .ttl of ontology and violations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For the graph\n",
    "ontology_union_violation_exemplars_g = add_labels_to_graph(ontology_union_violation_exemplars_g, label_dict)\n",
    "\n",
    "# For the dictionaries\n",
    "edge_count_dict = replace_uris_with_labels(edge_count_dict, label_dict, data_type=\"dict\")\n",
    "focus_node_exemplar_dict = replace_uris_with_labels(focus_node_exemplar_dict, label_dict, data_type=\"dict\")\n",
    "exemplar_focus_node_dict = replace_uris_with_labels(exemplar_focus_node_dict, label_dict, data_type=\"dict\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Store"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Graph identifier=Ncf4d0effee08468db3116bdffd3cfed6 (<class 'rdflib.graph.Graph'>)>"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ontology_union_violation_exemplars_g.serialize(destination=OMICS_MODEL_UNION_VIOLATION_EXEMPLAR_TTL_PATH, format=\"ttl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "serializing edge count dict. dict:  {rdflib.term.URIRef('http://example.com/exemplar#afb51f95-5b17-45a4-b62d-c58f4998f930_exemplar_1'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1734, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1734, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1734, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/Species>': 1734, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 1734, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasSpecies': 1734, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1734, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1734, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1734, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/afb51f95-5b17-45a4-b62d-c58f4998f930': 1734, 'http://www.w3.org/ns/shacl#value__Homo sapiens': 1734}, rdflib.term.URIRef('http://example.com/exemplar#TranscriptOmicsSampleShape-isPreparedByLibrary_exemplar_2'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1734, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1734, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1734, 'http://www.w3.org/ns/shacl#message__Expected exactly 1 value(s) for omics:isPreparedByLibrary for entity of type omics:LibraryPreparation': 1734, 'http://www.w3.org/ns/shacl#resultMessage__Expected exactly 1 value(s) for omics:isPreparedByLibrary for entity of type omics:LibraryPreparation': 1734, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/isPreparedByLibrary': 1734, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1734, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1734, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1734, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-isPreparedByLibrary': 1734, 'http://www.w3.org/ns/shacl#value__polyA RNA': 1734}, rdflib.term.URIRef('http://example.com/exemplar#TranscriptOmicsSampleShape-hasStrandedness_exemplar_3'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1734, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1734, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1734, 'http://www.w3.org/ns/shacl#message__Expected exactly 1 value(s) for omics:hasStrandedness for entity of type omics:Strandedness': 1734, 'http://www.w3.org/ns/shacl#resultMessage__Expected exactly 1 value(s) for omics:hasStrandedness for entity of type omics:Strandedness': 1734, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasStrandedness': 1734, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1734, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1734, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1734, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-hasStrandedness': 1734, 'http://www.w3.org/ns/shacl#value__unknown': 1734}, rdflib.term.URIRef('http://example.com/exemplar#0c2600b5-30b2-40c7-917a-b402e73f55ae_exemplar_4'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1734, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1734, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1734, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/SampleType>': 1734, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 1734, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasSpecimenType': 1734, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1734, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1734, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1734, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/0c2600b5-30b2-40c7-917a-b402e73f55ae': 1734, 'http://www.w3.org/ns/shacl#value__Primary cells': 1734}, rdflib.term.URIRef('http://example.com/exemplar#ece15faa-30b3-471c-ae51-2d6d1f80e3a9_exemplar_5'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1734, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1734, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1734, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/Organisation>': 1734, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 1734, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/isProcessedByOrganisation': 1734, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1734, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1734, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1734, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/ece15faa-30b3-471c-ae51-2d6d1f80e3a9': 1734, 'http://www.w3.org/ns/shacl#value__Boehringer Ingelheim, Biberach, Germany': 1734}, rdflib.term.URIRef('http://example.com/exemplar#248e9792-8549-4e0d-8665-06c42aff8ffc_exemplar_6'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1734, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1734, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1734, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/SamplePreparation>': 1734, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 1734, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasSpecimenPreparation': 1734, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1734, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1734, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1734, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/248e9792-8549-4e0d-8665-06c42aff8ffc': 1734, 'http://www.w3.org/ns/shacl#value__FACS sorted biological material': 1734}, rdflib.term.URIRef('http://example.com/exemplar#TranscriptOmicsSampleShape-isMeasuredOnPlatform_exemplar_7'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1734, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1734, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1734, 'http://www.w3.org/ns/shacl#message__Expected exactly 1 value(s) for omics:isMeasuredOnPlatform for entity of type omics:Platform': 1734, 'http://www.w3.org/ns/shacl#resultMessage__Expected exactly 1 value(s) for omics:isMeasuredOnPlatform for entity of type omics:Platform': 1734, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/isMeasuredOnPlatform': 1734, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1734, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1734, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1734, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-isMeasuredOnPlatform': 1734, 'http://www.w3.org/ns/shacl#value__HiSeq2500': 1734}, rdflib.term.URIRef('http://example.com/exemplar#af998a22-f11b-43ea-b0ac-44e728baeafb_exemplar_8'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 564, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 564, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 564, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/Cell>': 564, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 564, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasSourceCellType': 564, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 564, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 564, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 564, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/af998a22-f11b-43ea-b0ac-44e728baeafb': 564, 'http://www.w3.org/ns/shacl#value__Peripheral Blood Mononuclear Cell': 564}, rdflib.term.URIRef('http://example.com/exemplar#TranscriptOmicsSampleShape-hasSequencingProtocol_exemplar_9'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1734, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1734, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1734, 'http://www.w3.org/ns/shacl#message__Expected at minimum 0 and at maximum 1 value(s) for omics:hasSequencingProtocol for entity of type omics:SequencingProtocol': 1734, 'http://www.w3.org/ns/shacl#resultMessage__Expected at minimum 0 and at maximum 1 value(s) for omics:hasSequencingProtocol for entity of type omics:SequencingProtocol': 1734, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasSequencingProtocol': 1734, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1734, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1734, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1734, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-hasSequencingProtocol': 1734, 'http://www.w3.org/ns/shacl#value__scRNA-seq': 1734}, rdflib.term.URIRef('http://example.com/exemplar#9589ebf0-6f4f-4b80-9a5a-75aa25b5715f_exemplar_10'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1734, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1734, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1734, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/Disease>': 1734, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 1734, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasDisease': 1734, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1734, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1734, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1734, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/9589ebf0-6f4f-4b80-9a5a-75aa25b5715f': 1734, 'http://www.w3.org/ns/shacl#value__Normal': 1734}, rdflib.term.URIRef('http://example.com/exemplar#38c0a1c7-1c92-489a-a22e-87f5335ccb1a_exemplar_11'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1170, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1170, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1170, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/Cell>': 1170, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 1170, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasCellType': 1170, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1170, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1170, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1170, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/38c0a1c7-1c92-489a-a22e-87f5335ccb1a': 1170, 'http://www.w3.org/ns/shacl#value__Peripheral Blood Mononuclear Cell': 1170}, rdflib.term.URIRef('http://example.com/exemplar#38c0a1c7-1c92-489a-a22e-87f5335ccb1a_exemplar_12'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 192, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 192, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 192, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/Cell>': 192, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 192, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasCellType': 192, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 192, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 192, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 192, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/38c0a1c7-1c92-489a-a22e-87f5335ccb1a': 192, 'http://www.w3.org/ns/shacl#value__Plasmacytoid Dendritic Cell': 192}, rdflib.term.URIRef('http://example.com/exemplar#38c0a1c7-1c92-489a-a22e-87f5335ccb1a_exemplar_13'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 124, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 124, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 124, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/Cell>': 124, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 124, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasCellType': 124, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 124, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 124, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 124, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/38c0a1c7-1c92-489a-a22e-87f5335ccb1a': 124, 'http://www.w3.org/ns/shacl#value__Classical Monocyte': 124}, rdflib.term.URIRef('http://example.com/exemplar#38c0a1c7-1c92-489a-a22e-87f5335ccb1a_exemplar_14'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 124, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 124, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 124, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/Cell>': 124, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 124, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasCellType': 124, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 124, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 124, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 124, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/38c0a1c7-1c92-489a-a22e-87f5335ccb1a': 124, 'http://www.w3.org/ns/shacl#value__Non-classical Monocyte': 124}, rdflib.term.URIRef('http://example.com/exemplar#38c0a1c7-1c92-489a-a22e-87f5335ccb1a_exemplar_15'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 124, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 124, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 124, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/Cell>': 124, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 124, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasCellType': 124, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 124, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 124, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 124, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/38c0a1c7-1c92-489a-a22e-87f5335ccb1a': 124, 'http://www.w3.org/ns/shacl#value__Intermediate Monocyte': 124}, rdflib.term.URIRef('http://example.com/exemplar#OmicsStudyShape-hasStudyType_exemplar_16'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1, 'http://www.w3.org/ns/shacl#message__Expected at minimum 1 and at maximum n value(s) for omics:hasStudyType for entity of type omics:StudyType': 1, 'http://www.w3.org/ns/shacl#resultMessage__Expected at minimum 1 and at maximum n value(s) for omics:hasStudyType for entity of type omics:StudyType': 1, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasStudyType': 1, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/OmicsStudyShape-hasStudyType': 1, 'http://www.w3.org/ns/shacl#value__ex vivo': 1}, rdflib.term.URIRef('http://example.com/exemplar#OmicsStudyShape-hasContactScientist_exemplar_17'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1, 'http://www.w3.org/ns/shacl#message__Expected at minimum 1 and at maximum n value(s) for omics:hasContactScientist for entity of type omics:Person': 1, 'http://www.w3.org/ns/shacl#resultMessage__Expected at minimum 1 and at maximum n value(s) for omics:hasContactScientist for entity of type omics:Person': 1, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasContactScientist': 1, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/OmicsStudyShape-hasContactScientist': 1, 'http://www.w3.org/ns/shacl#value__Stefan Pflanz': 1}, rdflib.term.URIRef('http://example.com/exemplar#OmicsStudyShape-isSequencedForTherapeuticArea_exemplar_18'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1, 'http://www.w3.org/ns/shacl#message__Expected at minimum 1 and at maximum n value(s) for omics:isSequencedForTherapeuticArea for entity of type omics:TherapeuticArea': 1, 'http://www.w3.org/ns/shacl#resultMessage__Expected at minimum 1 and at maximum n value(s) for omics:isSequencedForTherapeuticArea for entity of type omics:TherapeuticArea': 1, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/isSequencedForTherapeuticArea': 1, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/OmicsStudyShape-isSequencedForTherapeuticArea': 1, 'http://www.w3.org/ns/shacl#value__CI&IM': 1}, rdflib.term.URIRef('http://example.com/exemplar#OmicsStudyShape-isPartOfProject_exemplar_19'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1, 'http://www.w3.org/ns/shacl#message__Expected at minimum 1 and at maximum n value(s) for omics:isPartOfProject for entity of type omics:Project': 1, 'http://www.w3.org/ns/shacl#resultMessage__Expected at minimum 1 and at maximum n value(s) for omics:isPartOfProject for entity of type omics:Project': 1, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/isPartOfProject': 1, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/OmicsStudyShape-isPartOfProject': 1, 'http://www.w3.org/ns/shacl#value__Exploratory NDC Immune Modulation': 1}, rdflib.term.URIRef('http://example.com/exemplar#OmicsStudyShape-hasResponsibleSite_exemplar_20'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/Organisation>': 1, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 1, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasResponsibleSite': 1, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/OmicsStudyShape-hasResponsibleSite': 1, 'http://www.w3.org/ns/shacl#value__Boehringer Ingelheim, Biberach, Germany': 1}, rdflib.term.URIRef('http://example.com/exemplar#OmicsStudyShape-hasContactCompBio_exemplar_21'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1, 'http://www.w3.org/ns/shacl#message__Expected at minimum 1 and at maximum n value(s) for omics:hasContactCompBio for entity of type omics:Person': 1, 'http://www.w3.org/ns/shacl#resultMessage__Expected at minimum 1 and at maximum n value(s) for omics:hasContactCompBio for entity of type omics:Person': 1, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasContactCompBio': 1, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/OmicsStudyShape-hasContactCompBio': 1, 'http://www.w3.org/ns/shacl#value__Baerbel Laemmle': 1}, rdflib.term.URIRef('http://example.com/exemplar#OmicsStudyShape-isSequencedForIndication_exemplar_22'): {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1, 'http://www.w3.org/ns/shacl#message__Expected at minimum 1 and at maximum n value(s) for omics:isSequencedForIndication for entity of type omics:Disease': 1, 'http://www.w3.org/ns/shacl#resultMessage__Expected at minimum 1 and at maximum n value(s) for omics:isSequencedForIndication for entity of type omics:Disease': 1, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/isSequencedForIndication': 1, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/OmicsStudyShape-isSequencedForIndication': 1, 'http://www.w3.org/ns/shacl#value__Immune Modulation': 1}}\n",
      "serialized dict:  {'http://example.com/exemplar#afb51f95-5b17-45a4-b62d-c58f4998f930_exemplar_1': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1734, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1734, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1734, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/Species>': 1734, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 1734, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasSpecies': 1734, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1734, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1734, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1734, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/afb51f95-5b17-45a4-b62d-c58f4998f930': 1734, 'http://www.w3.org/ns/shacl#value__Homo sapiens': 1734}, 'http://example.com/exemplar#TranscriptOmicsSampleShape-isPreparedByLibrary_exemplar_2': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1734, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1734, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1734, 'http://www.w3.org/ns/shacl#message__Expected exactly 1 value(s) for omics:isPreparedByLibrary for entity of type omics:LibraryPreparation': 1734, 'http://www.w3.org/ns/shacl#resultMessage__Expected exactly 1 value(s) for omics:isPreparedByLibrary for entity of type omics:LibraryPreparation': 1734, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/isPreparedByLibrary': 1734, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1734, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1734, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1734, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-isPreparedByLibrary': 1734, 'http://www.w3.org/ns/shacl#value__polyA RNA': 1734}, 'http://example.com/exemplar#TranscriptOmicsSampleShape-hasStrandedness_exemplar_3': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1734, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1734, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1734, 'http://www.w3.org/ns/shacl#message__Expected exactly 1 value(s) for omics:hasStrandedness for entity of type omics:Strandedness': 1734, 'http://www.w3.org/ns/shacl#resultMessage__Expected exactly 1 value(s) for omics:hasStrandedness for entity of type omics:Strandedness': 1734, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasStrandedness': 1734, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1734, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1734, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1734, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-hasStrandedness': 1734, 'http://www.w3.org/ns/shacl#value__unknown': 1734}, 'http://example.com/exemplar#0c2600b5-30b2-40c7-917a-b402e73f55ae_exemplar_4': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1734, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1734, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1734, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/SampleType>': 1734, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 1734, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasSpecimenType': 1734, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1734, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1734, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1734, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/0c2600b5-30b2-40c7-917a-b402e73f55ae': 1734, 'http://www.w3.org/ns/shacl#value__Primary cells': 1734}, 'http://example.com/exemplar#ece15faa-30b3-471c-ae51-2d6d1f80e3a9_exemplar_5': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1734, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1734, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1734, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/Organisation>': 1734, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 1734, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/isProcessedByOrganisation': 1734, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1734, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1734, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1734, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/ece15faa-30b3-471c-ae51-2d6d1f80e3a9': 1734, 'http://www.w3.org/ns/shacl#value__Boehringer Ingelheim, Biberach, Germany': 1734}, 'http://example.com/exemplar#248e9792-8549-4e0d-8665-06c42aff8ffc_exemplar_6': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1734, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1734, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1734, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/SamplePreparation>': 1734, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 1734, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasSpecimenPreparation': 1734, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1734, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1734, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1734, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/248e9792-8549-4e0d-8665-06c42aff8ffc': 1734, 'http://www.w3.org/ns/shacl#value__FACS sorted biological material': 1734}, 'http://example.com/exemplar#TranscriptOmicsSampleShape-isMeasuredOnPlatform_exemplar_7': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1734, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1734, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1734, 'http://www.w3.org/ns/shacl#message__Expected exactly 1 value(s) for omics:isMeasuredOnPlatform for entity of type omics:Platform': 1734, 'http://www.w3.org/ns/shacl#resultMessage__Expected exactly 1 value(s) for omics:isMeasuredOnPlatform for entity of type omics:Platform': 1734, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/isMeasuredOnPlatform': 1734, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1734, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1734, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1734, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-isMeasuredOnPlatform': 1734, 'http://www.w3.org/ns/shacl#value__HiSeq2500': 1734}, 'http://example.com/exemplar#af998a22-f11b-43ea-b0ac-44e728baeafb_exemplar_8': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 564, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 564, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 564, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/Cell>': 564, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 564, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasSourceCellType': 564, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 564, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 564, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 564, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/af998a22-f11b-43ea-b0ac-44e728baeafb': 564, 'http://www.w3.org/ns/shacl#value__Peripheral Blood Mononuclear Cell': 564}, 'http://example.com/exemplar#TranscriptOmicsSampleShape-hasSequencingProtocol_exemplar_9': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1734, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1734, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1734, 'http://www.w3.org/ns/shacl#message__Expected at minimum 0 and at maximum 1 value(s) for omics:hasSequencingProtocol for entity of type omics:SequencingProtocol': 1734, 'http://www.w3.org/ns/shacl#resultMessage__Expected at minimum 0 and at maximum 1 value(s) for omics:hasSequencingProtocol for entity of type omics:SequencingProtocol': 1734, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasSequencingProtocol': 1734, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1734, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1734, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1734, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-hasSequencingProtocol': 1734, 'http://www.w3.org/ns/shacl#value__scRNA-seq': 1734}, 'http://example.com/exemplar#9589ebf0-6f4f-4b80-9a5a-75aa25b5715f_exemplar_10': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1734, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1734, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1734, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/Disease>': 1734, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 1734, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasDisease': 1734, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1734, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1734, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1734, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/9589ebf0-6f4f-4b80-9a5a-75aa25b5715f': 1734, 'http://www.w3.org/ns/shacl#value__Normal': 1734}, 'http://example.com/exemplar#38c0a1c7-1c92-489a-a22e-87f5335ccb1a_exemplar_11': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1170, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1170, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1170, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/Cell>': 1170, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 1170, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasCellType': 1170, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1170, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1170, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1170, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/38c0a1c7-1c92-489a-a22e-87f5335ccb1a': 1170, 'http://www.w3.org/ns/shacl#value__Peripheral Blood Mononuclear Cell': 1170}, 'http://example.com/exemplar#38c0a1c7-1c92-489a-a22e-87f5335ccb1a_exemplar_12': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 192, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 192, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 192, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/Cell>': 192, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 192, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasCellType': 192, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 192, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 192, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 192, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/38c0a1c7-1c92-489a-a22e-87f5335ccb1a': 192, 'http://www.w3.org/ns/shacl#value__Plasmacytoid Dendritic Cell': 192}, 'http://example.com/exemplar#38c0a1c7-1c92-489a-a22e-87f5335ccb1a_exemplar_13': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 124, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 124, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 124, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/Cell>': 124, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 124, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasCellType': 124, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 124, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 124, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 124, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/38c0a1c7-1c92-489a-a22e-87f5335ccb1a': 124, 'http://www.w3.org/ns/shacl#value__Classical Monocyte': 124}, 'http://example.com/exemplar#38c0a1c7-1c92-489a-a22e-87f5335ccb1a_exemplar_14': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 124, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 124, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 124, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/Cell>': 124, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 124, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasCellType': 124, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 124, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 124, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 124, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/38c0a1c7-1c92-489a-a22e-87f5335ccb1a': 124, 'http://www.w3.org/ns/shacl#value__Non-classical Monocyte': 124}, 'http://example.com/exemplar#38c0a1c7-1c92-489a-a22e-87f5335ccb1a_exemplar_15': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 124, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 124, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 124, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/Cell>': 124, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 124, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasCellType': 124, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 124, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 124, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 124, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/38c0a1c7-1c92-489a-a22e-87f5335ccb1a': 124, 'http://www.w3.org/ns/shacl#value__Intermediate Monocyte': 124}, 'http://example.com/exemplar#OmicsStudyShape-hasStudyType_exemplar_16': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1, 'http://www.w3.org/ns/shacl#message__Expected at minimum 1 and at maximum n value(s) for omics:hasStudyType for entity of type omics:StudyType': 1, 'http://www.w3.org/ns/shacl#resultMessage__Expected at minimum 1 and at maximum n value(s) for omics:hasStudyType for entity of type omics:StudyType': 1, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasStudyType': 1, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/OmicsStudyShape-hasStudyType': 1, 'http://www.w3.org/ns/shacl#value__ex vivo': 1}, 'http://example.com/exemplar#OmicsStudyShape-hasContactScientist_exemplar_17': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1, 'http://www.w3.org/ns/shacl#message__Expected at minimum 1 and at maximum n value(s) for omics:hasContactScientist for entity of type omics:Person': 1, 'http://www.w3.org/ns/shacl#resultMessage__Expected at minimum 1 and at maximum n value(s) for omics:hasContactScientist for entity of type omics:Person': 1, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasContactScientist': 1, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/OmicsStudyShape-hasContactScientist': 1, 'http://www.w3.org/ns/shacl#value__Stefan Pflanz': 1}, 'http://example.com/exemplar#OmicsStudyShape-isSequencedForTherapeuticArea_exemplar_18': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1, 'http://www.w3.org/ns/shacl#message__Expected at minimum 1 and at maximum n value(s) for omics:isSequencedForTherapeuticArea for entity of type omics:TherapeuticArea': 1, 'http://www.w3.org/ns/shacl#resultMessage__Expected at minimum 1 and at maximum n value(s) for omics:isSequencedForTherapeuticArea for entity of type omics:TherapeuticArea': 1, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/isSequencedForTherapeuticArea': 1, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/OmicsStudyShape-isSequencedForTherapeuticArea': 1, 'http://www.w3.org/ns/shacl#value__CI&IM': 1}, 'http://example.com/exemplar#OmicsStudyShape-isPartOfProject_exemplar_19': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1, 'http://www.w3.org/ns/shacl#message__Expected at minimum 1 and at maximum n value(s) for omics:isPartOfProject for entity of type omics:Project': 1, 'http://www.w3.org/ns/shacl#resultMessage__Expected at minimum 1 and at maximum n value(s) for omics:isPartOfProject for entity of type omics:Project': 1, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/isPartOfProject': 1, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/OmicsStudyShape-isPartOfProject': 1, 'http://www.w3.org/ns/shacl#value__Exploratory NDC Immune Modulation': 1}, 'http://example.com/exemplar#OmicsStudyShape-hasResponsibleSite_exemplar_20': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1, 'http://www.w3.org/ns/shacl#message__value not of type <http://data.boehringer.com/ontology/omics/Organisation>': 1, 'http://www.w3.org/ns/shacl#resultMessage__value not of type {$class}': 1, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasResponsibleSite': 1, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/OmicsStudyShape-hasResponsibleSite': 1, 'http://www.w3.org/ns/shacl#value__Boehringer Ingelheim, Biberach, Germany': 1}, 'http://example.com/exemplar#OmicsStudyShape-hasContactCompBio_exemplar_21': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1, 'http://www.w3.org/ns/shacl#message__Expected at minimum 1 and at maximum n value(s) for omics:hasContactCompBio for entity of type omics:Person': 1, 'http://www.w3.org/ns/shacl#resultMessage__Expected at minimum 1 and at maximum n value(s) for omics:hasContactCompBio for entity of type omics:Person': 1, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/hasContactCompBio': 1, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/OmicsStudyShape-hasContactCompBio': 1, 'http://www.w3.org/ns/shacl#value__Baerbel Laemmle': 1}, 'http://example.com/exemplar#OmicsStudyShape-isSequencedForIndication_exemplar_22': {'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://www.w3.org/ns/shacl#ValidationResult': 1, 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type__http://rdfunit.aksw.org/ns/core#TestCaseResult': 1, 'http://www.w3.org/ns/prov#wasGeneratedBy__urn:uuid:2d10a433-714b-411f-a69a-113c1e322d5e': 1, 'http://www.w3.org/ns/shacl#message__Expected at minimum 1 and at maximum n value(s) for omics:isSequencedForIndication for entity of type omics:Disease': 1, 'http://www.w3.org/ns/shacl#resultMessage__Expected at minimum 1 and at maximum n value(s) for omics:isSequencedForIndication for entity of type omics:Disease': 1, 'http://www.w3.org/ns/shacl#resultPath__http://data.boehringer.com/ontology/omics/isSequencedForIndication': 1, 'http://www.w3.org/ns/shacl#resultSeverity__http://www.w3.org/ns/shacl#Violation': 1, 'http://www.w3.org/ns/shacl#severity__http://persistence.uni-leipzig.org/nlp2rdf/ontologies/rlog#ERROR': 1, 'http://www.w3.org/ns/shacl#sourceConstraintComponent__http://www.w3.org/ns/shacl#ClassConstraintComponent': 1, 'http://www.w3.org/ns/shacl#sourceShape__http://data.boehringer.com/ontology/omics/OmicsStudyShape-isSequencedForIndication': 1, 'http://www.w3.org/ns/shacl#value__Immune Modulation': 1}}\n"
     ]
    }
   ],
   "source": [
    "from routers.utils import save_edge_count_json, save_uri_set_of_uris_dict\n",
    "\n",
    "save_edge_count_json(edge_count_dict, EXEMPLAR_EDGE_COUNT_DICT_PATH)\n",
    "save_uri_set_of_uris_dict(focus_node_exemplar_dict, FOCUS_NODE_EXEMPLAR_DICT_PATH)\n",
    "save_uri_set_of_uris_dict(exemplar_focus_node_dict, EXEMPLAR_FOCUS_NODE_DICT_PATH)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py38",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
