{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports & Constants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from rdflib import Graph, Namespace\n",
    "from collections import defaultdict\n",
    "from tqdm.notebook import tqdm\n",
    "import json\n",
    "# input\n",
    "STUDY_TTL_PATH = './ttl/study.ttl' # contains the actual study data\n",
    "VIOLATION_REPORT_TTL_PATH = './ttl/violation_report.ttl' # contains the shacl violation test case result\n",
    "OMICS_MODEL_TTL_PATH = './ttl/omics_model.ttl' # contains the ontology\n",
    "\n",
    "# output\n",
    "VIOLATION_LIST_FILE = \"./json/violation_list.json\" # path to store the violation list\n",
    "STUDY_CSV_FILE = \"./csv/study.csv\" # path to store the tabluraized study csv file\n",
    "OMICS_MODEL_UNION_VIOLATION_EXEMPLAR_TTL_PATH = './ttl/omics_model_union_violation_exemplar.ttl' # contains the union of the ontology and the violation exemplars\n",
    "EXEMPLAR_EDGE_COUNT_DICT_PATH = './json/exemplar_edge_count_dict.json' # contains the nested dict of exemplar edge counts\n",
    "FOCUS_NODE_EXEMPLAR_DICT_PATH = './json/focus_node_exemplar_dict.json' # contains the dict of exemplars per focus node\n",
    "EXEMPLAR_FOCUS_NODE_DICT_PATH = './json/exemplar_focus_node_dict.json' # contains the dict of focus nodes per exemplar\n",
    "VIOLATION_EXEMPLAR_DICT_PATH = './json/violation_exemplar_dict.json' # contains the dict of exemplars and their counts per violation"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read & Parse Graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading study graph...\n",
      "Reading violations graph...\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Graph identifier=N578fd1b4bbd741a9ac30d99e3607f3d0 (<class 'rdflib.graph.Graph'>)>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Reading study graph...\")\n",
    "study_g = Graph()\n",
    "study_g.parse(STUDY_TTL_PATH, format=\"ttl\")\n",
    "\n",
    "print(\"Reading violations graph...\")\n",
    "violations_g = Graph()\n",
    "violations_g.parse(VIOLATION_REPORT_TTL_PATH, format=\"ttl\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Give Enumarting Names to Unnamed NodeShapes\n",
    "example: [] a sh:NodeShape -> NodeShape1 a sh:NodeShape\n",
    "etc\n",
    "this could be done with rdflib but then when storing the graph to a .ttl file it would automatically mess with the namespaces and remove unused ones\n",
    "so until these names are provided by BI, we decided to do simple string replacement"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"color:orange\">Careful: If the omics_model.ttl contains [] anywhere else, e.g., in a label, it will also be replaced there!</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "# Read the input file\n",
    "with open(OMICS_MODEL_TTL_PATH, 'r') as file:\n",
    "    data = file.read()\n",
    "\n",
    "# Define backup file path\n",
    "backup_file_path = OMICS_MODEL_TTL_PATH + \".bak\"\n",
    "\n",
    "# Ensure the output directory exists\n",
    "os.makedirs(os.path.dirname(OMICS_MODEL_TTL_PATH), exist_ok=True)\n",
    "\n",
    "# store backup file\n",
    "with open(backup_file_path, 'w') as file:\n",
    "    file.write(data)\n",
    "\n",
    "# Initialize the counter\n",
    "counter = 1\n",
    "\n",
    "# Prepend '@prefix cns: <http://custom_node_shapes/> .'\n",
    "data = \"@prefix cns: <http://customnamespace.com/> .\\n\" + data\n",
    "\n",
    "# Replace occurrences of [] with NodeShape{counter}\n",
    "while \"[]\" in data:\n",
    "    data = data.replace(\"[]\", f\"cns:NodeShape{counter}\", 1)\n",
    "    counter += 1\n",
    "\n",
    "# Write the updated string to the new file\n",
    "with open(OMICS_MODEL_TTL_PATH, 'w') as file:\n",
    "    file.write(data)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Differentiate NaN Cases - Predicate cannot Exist on Class vs. Predicate is Missing vs. Predicta Exists Even Though It Cannot\n",
    "A node can have an edge if\n",
    "\n",
    "s1 a sh:NodeShape\n",
    "\n",
    "s1 sh:targetClass node\n",
    "\n",
    "s1 sh:property s2\n",
    "\n",
    "s2 a sh:PropertyShape\n",
    "\n",
    "s2 sh:Path edge\n",
    "\n",
    "OR \n",
    "\n",
    "(\n",
    "node rdfs:subClassOf s3\n",
    "AND\n",
    "s3 can have the edge\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Roots: ['omics:Thing', 'omics:KeyValueHolder', 'omics:ReferenceData']\n",
      "Edges: ['omics:hasCellType', 'omics:hasSourceTissue', 'omics:hasSourceCellType']\n",
      "Parents: ['omics:Specimen']\n",
      "Children: ['omics:PDXSpecimen', 'omics:PrimaryCellSpecimen']\n"
     ]
    }
   ],
   "source": [
    "from rdflib import Graph\n",
    "from rdflib.namespace import Namespace\n",
    "import networkx as nx\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# read ontology into a graph\n",
    "ontology_g = Graph()\n",
    "ontology_g.parse(OMICS_MODEL_TTL_PATH, format=\"ttl\")\n",
    "\n",
    "# take namespace prefix for sh and omics from the graph\n",
    "sh = Namespace(\"http://www.w3.org/ns/shacl#\")\n",
    "omics = Namespace(\"http://data.boehringer.com/ontology/omics#\")\n",
    "owl = Namespace(\"http://www.w3.org/2002/07/owl#\")\n",
    "\n",
    "# define prefixes and corresponding namespaces\n",
    "prefixes = {\n",
    "    \"sh\": \"http://www.w3.org/ns/shacl#\",\n",
    "    \"omics\": \"http://data.boehringer.com/ontology/omics#\",\n",
    "    \"owl\": \"http://www.w3.org/2002/07/owl#\"\n",
    "}\n",
    "\n",
    "def convert_to_prefixed(graph, uri):\n",
    "    \"\"\"Convert a URI to its prefixed form using a rdflib graph's namespace manager.\"\"\"\n",
    "    prefix, namespace, name = graph.namespace_manager.compute_qname(uri)\n",
    "    return f\"{prefix}:{name}\"\n",
    "\n",
    "def select_direct_possible_edges(node):\n",
    "    select_statement = \"\"\"\n",
    "    SELECT ?edge\n",
    "    WHERE {\n",
    "        ?ns a sh:NodeShape .\n",
    "        ?ns sh:targetClass omics:%s .\n",
    "        ?ns sh:property ?ps .\n",
    "        ?ps a sh:PropertyShape .\n",
    "        ?ps sh:path ?edge .\n",
    "    }\n",
    "    \"\"\" % node\n",
    "    return [str(result[0].n3(ontology_g.namespace_manager)) for result in ontology_g.query(select_statement)] # type: ignore\n",
    "\n",
    "def select_direct_parents(node):\n",
    "    select_statement = \"\"\"\n",
    "    SELECT ?parent\n",
    "    WHERE {\n",
    "        omics:%s rdfs:subClassOf ?parent .\n",
    "    }\n",
    "    \"\"\" % node\n",
    "    return [str(result[0].n3(ontology_g.namespace_manager)) for result in ontology_g.query(select_statement)] # type: ignore\n",
    "\n",
    "def select_direct_children(node):\n",
    "    select_statement = \"\"\"\n",
    "    SELECT ?child\n",
    "    WHERE {\n",
    "        ?child rdfs:subClassOf omics:%s .\n",
    "    }\n",
    "    \"\"\" % node\n",
    "    return [str(result[0].n3(ontology_g.namespace_manager)) for result in ontology_g.query(select_statement)] # type: ignore\n",
    "\n",
    "def select_roots():\n",
    "    select_statement = \"\"\"\n",
    "    SELECT ?root\n",
    "    WHERE {\n",
    "        ?root rdfs:subClassOf owl:Thing .\n",
    "    }\n",
    "    \"\"\"\n",
    "    return [str(result[0].n3(ontology_g.namespace_manager)) for result in ontology_g.query(select_statement)] # type: ignore\n",
    "\n",
    "def select_root_classes():\n",
    "    select_statement = \"\"\"\n",
    "    SELECT ?class\n",
    "    WHERE {\n",
    "        ?class a owl:Class .\n",
    "        FILTER NOT EXISTS { ?class rdfs:subClassOf ?parent . }\n",
    "    }\n",
    "    \"\"\"\n",
    "    return [str(result[0].n3(ontology_g.namespace_manager)) for result in ontology_g.query(select_statement)] # type: ignore\n",
    "\n",
    "\n",
    "# use the functions\n",
    "roots = select_root_classes()\n",
    "edges = select_direct_possible_edges('PrimaryCellSpecimen')\n",
    "parents = select_direct_parents('BiopsySpecimen')\n",
    "children = select_direct_children('BiopsySpecimen')\n",
    "\n",
    "print('Roots:', roots)\n",
    "print('Edges:', edges)\n",
    "print('Parents:', parents)\n",
    "print('Children:', children)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "class OntologyTree:\n",
    "    def __init__(self):\n",
    "        self.tree = {}\n",
    "\n",
    "    def add_node(self, node, edges, children):\n",
    "        if node not in self.tree:\n",
    "            self.tree[node] = {'edges': edges, 'children': children}\n",
    "\n",
    "    def add_child(self, parent, child):\n",
    "        if parent in self.tree:\n",
    "            self.tree[parent]['children'].append(child)\n",
    "        else:\n",
    "            self.tree[parent] = {'edges': [], 'children': [child]}\n",
    "\n",
    "def aggregate_edges(node, tree):\n",
    "    edges = set(tree[node]['edges'])\n",
    "    queue = [parent_node for parent_node in tree if node in tree[parent_node]['children']]\n",
    "    while queue:\n",
    "        parent_node = queue.pop(0)\n",
    "        edges |= set(tree[parent_node]['edges'])\n",
    "        queue += [grand_parent for grand_parent in tree if parent_node in tree[grand_parent]['children']]\n",
    "    return edges\n",
    "\n",
    "def create_aggregated_edges_dict(tree):\n",
    "    agg_edges_dict = {}\n",
    "    for node in tree.keys():\n",
    "        agg_edges_dict[node] = aggregate_edges(node, tree)\n",
    "    return agg_edges_dict\n",
    "\n",
    "def convert_to_full_uri(graph, abbreviated_uri):\n",
    "    \"\"\"Converts a prefixed URI to a full URI using a rdflib graph's namespace manager.\"\"\"\n",
    "    prefix, uri = abbreviated_uri.split(':')\n",
    "    namespace = graph.namespace_manager.store.namespace(prefix)\n",
    "    return namespace + uri\n",
    "\n",
    "def remove_prefix(uri):\n",
    "    if ':' in uri:\n",
    "        return uri.split(':', 1)[1]\n",
    "    return uri\n",
    "\n",
    "def build_tree(root_nodes, ontology_tree):\n",
    "    queue = root_nodes\n",
    "    while queue:\n",
    "        current_node = queue.pop(0)\n",
    "        unprefixed_node = remove_prefix(current_node)\n",
    "        edges = select_direct_possible_edges(unprefixed_node)\n",
    "        children = select_direct_children(unprefixed_node)\n",
    "\n",
    "        ontology_tree.add_node(current_node, edges, children)\n",
    "\n",
    "        for child in children:\n",
    "            queue.append(child)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "roots = select_root_classes()\n",
    "ontology_tree = OntologyTree()\n",
    "build_tree(roots, ontology_tree)\n",
    "agg_edges_dict = create_aggregated_edges_dict(ontology_tree.tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "http://data.boehringer.com/ontology/omics/hasCellType\n"
     ]
    }
   ],
   "source": [
    "full_uri = convert_to_full_uri(ontology_g, \"omics:hasCellType\")\n",
    "print(full_uri)  # Outputs: http://[blabla...]/omics:hasCellType\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def select_all_focus_nodes_and_classes(graph):\n",
    "    select_statement = \"\"\"\n",
    "    SELECT ?node ?class\n",
    "    WHERE {\n",
    "        ?node a ?class .\n",
    "    }\n",
    "    \"\"\"\n",
    "    results = graph.query(select_statement)\n",
    "    focus_nodes_classes_dict = {str(result[0]): str(result[1]) for result in results}\n",
    "    return focus_nodes_classes_dict\n",
    "\n",
    "\n",
    "focus_nodes_class_dict = select_all_focus_nodes_and_classes(study_g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "http://data.boehringer.com/ontology/omics/TranscriptOmicsSample\n",
      "omics:TranscriptOmicsSample\n",
      "{'omics:processingDate', 'omics:isPartOfStudy', 'omics:hasStrandedness', 'omics:isPreparedByLibrary', 'omics:isExcluded', 'omics:isProcessedByOrganisation', 'omics:sourceId', 'omics:isPartOfSampleSet', 'omics:name', 'omics:isDerivedFromSpecimen', 'omics:hasSequencingProtocol', 'omics:sampleOrder', 'omics:id', 'omics:comment', 'omics:hasKeyValue', 'omics:isMeasuredOnPlatform', 'omics:hasRIN', 'omics:hasExclusionReason'}\n"
     ]
    }
   ],
   "source": [
    "x = focus_nodes_class_dict['http://data.boehringer.com/uuid/TranscriptOmicsSample/sample-EX51-EX51_545']\n",
    "print(x) # this returns http://data.boehringer.com/ontology/omics/TranscriptOmicsSample\n",
    "print(convert_to_prefixed(ontology_g, x)) # should return omics:TranscriptOmicsSample\n",
    "print(agg_edges_dict[convert_to_prefixed(ontology_g, x)])\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tabularize & Join Study + Violations Graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "study graph triples: 97794\n",
      "violations graph triples: 286592\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing Graph Once: 100%|██████████| 286592/286592 [00:02<00:00, 136771.80it/s]\n",
      "Going Over Violation Instances: 100%|██████████| 17911/17911 [00:00<00:00, 691846.75it/s]\n",
      "Processing study: 100%|██████████| 97794/97794 [00:01<00:00, 65591.22it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count_skipped = 19186\n",
      "count_not_skipped = 78608\n",
      "count_allowed = 50295\n",
      "count_not_allowed_exists = 28313\n",
      "count_not_allowed_empty = 0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>focus_node</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-isMeasuredOnPlatform</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/ece15faa-30b3-471c-ae51-2d6d1f80e3a9</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-hasStrandedness</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-hasSequencingProtocol</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-isPreparedByLibrary</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/248e9792-8549-4e0d-8665-06c42aff8ffc</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/9589ebf0-6f4f-4b80-9a5a-75aa25b5715f</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/0c2600b5-30b2-40c7-917a-b402e73f55ae</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/38c0a1c7-1c92-489a-a22e-87f5335ccb1a</th>\n",
       "      <th>...</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/externalStudyId</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/sequencingDate</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/keywords</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/isSequencedForTherapeuticArea</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/studyId</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/studyTitle</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/isSequencedForIndication</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/hasStudyType</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/hasContactCompBio</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/hasContactScientist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/TranscriptOmicsSample/sample-EX51-EX51_263</th>\n",
       "      <td>http://data.boehringer.com/uuid/TranscriptOmic...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/PrimaryCellSpecimen/fff3049c-5e43-4d60-905b-4bb483f212b7</th>\n",
       "      <td>http://data.boehringer.com/uuid/PrimaryCellSpe...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/PrimaryCellSpecimen/3843350f-45b7-474d-baa0-9fe2e1fc6642</th>\n",
       "      <td>http://data.boehringer.com/uuid/PrimaryCellSpe...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/PrimaryCellSpecimen/0ff82279-a862-483e-acb6-c175201d6e35</th>\n",
       "      <td>http://data.boehringer.com/uuid/PrimaryCellSpe...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/PrimaryCellSpecimen/2333e9c1-8515-40c3-b39d-881b5cf22aae</th>\n",
       "      <td>http://data.boehringer.com/uuid/PrimaryCellSpe...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                           focus_node  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...  http://data.boehringer.com/uuid/TranscriptOmic...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  http://data.boehringer.com/uuid/PrimaryCellSpe...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  http://data.boehringer.com/uuid/PrimaryCellSpe...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  http://data.boehringer.com/uuid/PrimaryCellSpe...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  http://data.boehringer.com/uuid/PrimaryCellSpe...   \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-isMeasuredOnPlatform  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                1.0                                           \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                           \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                           \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                           \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                           \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/ece15faa-30b3-471c-ae51-2d6d1f80e3a9  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-hasStrandedness  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                1.0                                      \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                      \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                      \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                      \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                      \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-hasSequencingProtocol  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                1.0                                            \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                            \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                            \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                            \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                            \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-isPreparedByLibrary  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                1.0                                          \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                          \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                          \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                          \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                          \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/248e9792-8549-4e0d-8665-06c42aff8ffc  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/9589ebf0-6f4f-4b80-9a5a-75aa25b5715f  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/0c2600b5-30b2-40c7-917a-b402e73f55ae  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/38c0a1c7-1c92-489a-a22e-87f5335ccb1a  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "\n",
       "                                                    ...  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...  ...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  ...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  ...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  ...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  ...   \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/externalStudyId  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent           \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent           \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent           \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent           \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent           \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/sequencingDate  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent          \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent          \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent          \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent          \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent          \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/keywords  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent    \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent    \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent    \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent    \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent    \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/isSequencedForTherapeuticArea  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent                         \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent                         \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent                         \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent                         \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent                         \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/studyId  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent   \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/studyTitle  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent      \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent      \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent      \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent      \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent      \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/isSequencedForIndication  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent                    \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent                    \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent                    \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent                    \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent                    \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/hasStudyType  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent        \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent        \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent        \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent        \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent        \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/hasContactCompBio  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent             \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent             \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent             \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent             \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent             \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/hasContactScientist  \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent              \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent              \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent              \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent              \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent              \n",
       "\n",
       "[5 rows x 64 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from rdflib import Graph, Namespace\n",
    "from rdflib.namespace import NamespaceManager\n",
    "from collections import defaultdict\n",
    "import json\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "SH = Namespace(\"http://www.w3.org/ns/shacl#\")\n",
    "\n",
    "def count_violations(violations: Graph):\n",
    "    d_focus_node_d_source_shape_counts = defaultdict(lambda: defaultdict(int))\n",
    "    d_violation_focus_node = defaultdict((str))\n",
    "    d_violation_source_shape = defaultdict((str))\n",
    "    violation_list = []\n",
    "\n",
    "    for s, p, o in tqdm(violations, desc=\"Processing Graph Once\"):\n",
    "        if p == SH.sourceShape:\n",
    "            d_violation_source_shape[str(s)] = str(o)\n",
    "            violation_list += [str(o)]\n",
    "        elif p == SH.focusNode:\n",
    "            d_violation_focus_node[str(s)] = str(o)\n",
    "\n",
    "    # for key in violation_focus_node dict\n",
    "    for key in tqdm(d_violation_focus_node, desc=\"Going Over Violation Instances\"):\n",
    "        # init with 1 or add 1 if already exists\n",
    "        d_focus_node_d_source_shape_counts[d_violation_focus_node[key]][d_violation_source_shape[key]] += 1\n",
    "\n",
    "    violation_list = list(set(violation_list))\n",
    "    return d_focus_node_d_source_shape_counts, violation_list\n",
    "\n",
    "\n",
    "def save_violation_list(violation_list):\n",
    "    with open(VIOLATION_LIST_FILE, 'w') as f:\n",
    "        json.dump(violation_list, f)  \n",
    "\n",
    "\n",
    "def create_study_dataframe(study: Graph, d_focus_node_d_source_shape_counts, focus_nodes_class_dict, agg_edges_dict):\n",
    "    # for convenience\n",
    "    pre = lambda x: convert_to_prefixed(ontology_g, str(x))\n",
    "\n",
    "    # create an empty study_df, fill the violation counts\n",
    "    study_df = pd.DataFrame(d_focus_node_d_source_shape_counts).T.fillna(0)\n",
    "\n",
    "    count_not_skipped, count_skipped, count_allowed, count_not_allowed_exists, count_not_allowed_empty = 0, 0, 0, 0, 0\n",
    "\n",
    "    # iterate all study triples\n",
    "    for s, p, o in tqdm(study, desc=\"Processing study\"):\n",
    "        # if s is not in d_focus_node_d_source_shape_counts, continue as it is not a focus node\n",
    "        if str(s) not in d_focus_node_d_source_shape_counts:\n",
    "            count_skipped += 1\n",
    "            continue\n",
    "        count_not_skipped += 1\n",
    "\n",
    "        # if p is not in study_df.columns, add it and set it to EdgeNotPresent for all rows because so far there has not been a value for it for anything, and future values will overwrite it anyway\n",
    "        if str(p) not in study_df.columns:\n",
    "            study_df[str(p)] = \"EdgeNotPresent\"\n",
    "\n",
    "        # if object is empty throw exception and print details\n",
    "        if str(o) == \"\":\n",
    "            print(\"empty object for \" + str(s) + \" \" + str(p))\n",
    "            raise Exception(\"empty object for \" + str(s) + \" \" + str(p))\n",
    "\n",
    "        # get class of focus node\n",
    "        focus_node_class = focus_nodes_class_dict.get(str(s), None)\n",
    "        assert focus_node_class is not None, f\"The focus_node_class for {str(s)} was None, but expected a value.\"\n",
    "        # get allowed edges\n",
    "        focus_node_edges = agg_edges_dict.get(convert_to_prefixed(ontology_g, focus_node_class), set())\n",
    "        assert len(focus_node_edges) > 0\n",
    "\n",
    "        # if edge is allowed (and object is present otherwise we wouldn't get here), set value study_df.at[str(s), str(p)] = str(o)\n",
    "        if pre(p) in focus_node_edges:\n",
    "            study_df.at[str(s), str(p)] = str(o)\n",
    "            count_allowed += 1\n",
    "        # if edge is not allowed, and there is no object, set value study_df.at[str(s), str(p)] = \"EdgeNotPresent\"\n",
    "        elif pre(p) not in focus_node_edges and str(o) == \"\":\n",
    "            study_df.at[str(s), str(p)] = \"EdgeNotPresent\"\n",
    "            count_not_allowed_empty += 1\n",
    "        # if edge is not allowed, and there is an object, set value study_df.at[str(s), str(p)] = \"InvalidEdgeWUnexpectedEdgeWithDataithData: str(o)\"\n",
    "        elif pre(p) not in focus_node_edges and str(o) != \"\":\n",
    "            # study_df.at[str(s), str(p)] = \"UnexpectedEdgeWithData: \" + str(o)\n",
    "            study_df.at[str(s), str(p)] = str(o)\n",
    "            count_not_allowed_exists += 1\n",
    "        else:\n",
    "            raise Exception(\"should not happen: \" + str(s) + \" \" + str(p) + \" \" + str(o))\n",
    "\n",
    "    print(f\"count_skipped = {count_skipped}\")\n",
    "    print(f\"count_not_skipped = {count_not_skipped}\")\n",
    "    print(f\"count_allowed = {count_allowed}\")\n",
    "    print(f\"count_not_allowed_exists = {count_not_allowed_exists}\")\n",
    "    print(f\"count_not_allowed_empty = {count_not_allowed_empty}\")\n",
    "\n",
    "    assert study_df.isin(['']).sum().sum() == 0\n",
    "    assert study_df.isnull().sum().sum() == 0, f\"study_df.isnull().sum().sum() = {study_df.isnull().sum().sum()}\"\n",
    "\n",
    "    # add focus_node column, set it to the study_df.index, and move it to the very left of the dataframe\n",
    "    study_df['focus_node'] = study_df.index\n",
    "    cols = study_df.columns.tolist()\n",
    "    cols = cols[-1:] + cols[:-1]  # move the last column to the first\n",
    "    study_df = study_df[cols]\n",
    "\n",
    "    return study_df\n",
    "\n",
    "\n",
    "def save_study_dataframe(study_df):\n",
    "    study_df.to_csv(STUDY_CSV_FILE)\n",
    "\n",
    "\n",
    "def tabularize_graphs(study: Graph, violations: Graph):\n",
    "    print('study graph triples:', len(study))  # type: ignore\n",
    "    print('violations graph triples:', len(violations))  # type: ignore\n",
    "    d_focus_node_d_source_shape_counts, violation_list = count_violations(violations)\n",
    "    save_violation_list(violation_list)\n",
    "    study_df = create_study_dataframe(study, d_focus_node_d_source_shape_counts, focus_nodes_class_dict, agg_edges_dict)\n",
    "    save_study_dataframe(study_df)\n",
    "    return d_focus_node_d_source_shape_counts\n",
    "\n",
    "\n",
    "d_focus_node_d_source_shape_counts = tabularize_graphs(study_g, violations_g)\n",
    "\n",
    "study_df = pd.read_csv(STUDY_CSV_FILE, index_col=0)\n",
    "study_df.head()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Unit Tests for Violation Counting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing Graph Once: 100%|██████████| 6/6 [00:00<?, ?it/s]\n",
      "Going Over Violation Instances: 100%|██████████| 3/3 [00:00<00:00, 3003.80it/s]\n",
      "Processing Graph Once: 100%|██████████| 8/8 [00:00<?, ?it/s]\n",
      "Going Over Violation Instances: 100%|██████████| 4/4 [00:00<?, ?it/s]\n",
      ".\n",
      "----------------------------------------------------------------------\n",
      "Ran 2 tests in 0.013s\n",
      "\n",
      "OK\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<unittest.runner.TextTestResult run=2 errors=0 failures=0>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import unittest\n",
    "from rdflib import Graph, Literal, Namespace, BNode, URIRef\n",
    "\n",
    "SH = Namespace(\"http://www.w3.org/ns/shacl#\")\n",
    "\n",
    "class TestCountViolations(unittest.TestCase):\n",
    "    def setUp(self):\n",
    "        self.violations = Graph()\n",
    "\n",
    "    def test_count_violations_node_with_multiple_and_node_with_double(self):\n",
    "        self.focusNode1 = BNode()\n",
    "        self.focusNode2 = BNode()\n",
    "        self.sourceShape1 = Literal('shape1')\n",
    "        self.sourceShape2 = Literal('shape2')\n",
    "        self.violation1 = BNode()\n",
    "        self.violation2 = BNode()\n",
    "        self.violation3 = BNode()\n",
    "        self.violation4 = BNode()\n",
    "\n",
    "        # Add triples to the violations graph\n",
    "        self.violations.add((self.violation1, SH.focusNode, self.focusNode1))\n",
    "        self.violations.add((self.violation1, SH.sourceShape, self.sourceShape1))\n",
    "\n",
    "        self.violations.add((self.violation2, SH.focusNode, self.focusNode2))\n",
    "        self.violations.add((self.violation2, SH.sourceShape, self.sourceShape2))\n",
    "\n",
    "        self.violations.add((self.violation3, SH.focusNode, self.focusNode1))\n",
    "        self.violations.add((self.violation3, SH.sourceShape, self.sourceShape1))\n",
    "\n",
    "        self.violations.add((self.violation4, SH.focusNode, self.focusNode2))\n",
    "        self.violations.add((self.violation4, SH.sourceShape, self.sourceShape1))\n",
    "        violation_counts, violation_list = count_violations(self.violations)\n",
    "        expected_counts = {\n",
    "            str(self.focusNode1): {str(self.sourceShape1): 2},\n",
    "            str(self.focusNode2): {str(self.sourceShape2): 1, str(self.sourceShape1): 1}\n",
    "        }\n",
    "        expected_list = [str(self.sourceShape1), str(self.sourceShape2)] \n",
    "        \n",
    "        self.assertEqual(violation_counts, expected_counts)\n",
    "        self.assertEqual(set(violation_list), set(expected_list))  # Use set to ignore order\n",
    "\n",
    "    def test_count_violations_basic(self):\n",
    "        self.focusNode1 = BNode()\n",
    "        self.focusNode2 = BNode()\n",
    "        self.focusNode3 = BNode()\n",
    "        self.sourceShape1 = Literal('shape1')\n",
    "        self.sourceShape2 = Literal('shape2')\n",
    "        self.sourceShape3 = Literal('shape3')\n",
    "        self.violation1 = BNode()\n",
    "        self.violation2 = BNode()\n",
    "        self.violation3 = BNode()\n",
    "\n",
    "        # Add triples to the violations graph\n",
    "        self.violations.add((self.violation1, SH.focusNode, self.focusNode1))\n",
    "        self.violations.add((self.violation1, SH.sourceShape, self.sourceShape1))\n",
    "\n",
    "        self.violations.add((self.violation2, SH.focusNode, self.focusNode2))\n",
    "        self.violations.add((self.violation2, SH.sourceShape, self.sourceShape2))\n",
    "\n",
    "        self.violations.add((self.violation3, SH.focusNode, self.focusNode3))\n",
    "        self.violations.add((self.violation3, SH.sourceShape, self.sourceShape3))\n",
    "\n",
    "        violation_counts, violation_list = count_violations(self.violations)\n",
    "        expected_counts = {\n",
    "            str(self.focusNode1): {str(self.sourceShape1): 1},\n",
    "            str(self.focusNode2): {str(self.sourceShape2): 1},\n",
    "            str(self.focusNode3): {str(self.sourceShape3): 1}\n",
    "        }\n",
    "        expected_list = [str(self.sourceShape1), str(self.sourceShape2), str(self.sourceShape3)] \n",
    "        \n",
    "        self.assertEqual(violation_counts, expected_counts)\n",
    "        self.assertEqual(set(violation_list), set(expected_list))  # Use set to ignore order\n",
    "\n",
    "suite = unittest.TestLoader().loadTestsFromTestCase(TestCountViolations)\n",
    "unittest.TextTestRunner().run(suite)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Abbreviation using QNames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from rdflib import Graph\n",
    "from rdflib.namespace import split_uri\n",
    "\n",
    "def namespace_in_nsm(nsm, namespace):\n",
    "    \"\"\"\n",
    "    Checks if a given namespace is in the provided NamespaceManager.\n",
    "\n",
    "    Parameters:\n",
    "    nsm (NamespaceManager): the NamespaceManager object to check\n",
    "    namespace (str): the namespace to look for in the NamespaceManager\n",
    "\n",
    "    Returns:\n",
    "    bool: True if the namespace is in the NamespaceManager, False otherwise\n",
    "    \"\"\"\n",
    "    for prefix, ns in nsm.namespaces():\n",
    "        if str(ns) == str(namespace):\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "def get_qname(nsm, uri):\n",
    "    \"\"\"\n",
    "    Returns a QName for a given URI if its namespace is in the provided NamespaceManager.\n",
    "    If the namespace isn't found in the NamespaceManager, it returns the original URI.\n",
    "\n",
    "    Parameters:\n",
    "    nsm (NamespaceManager): the NamespaceManager object to use\n",
    "    uri (str): the URI to transform into a QName\n",
    "\n",
    "    Returns:\n",
    "    str: a QName if the namespace of the URI is in the NamespaceManager, the original URI otherwise\n",
    "    \"\"\"\n",
    "    namespace, name = split_uri(uri)\n",
    "    if namespace_in_nsm(nsm, namespace):\n",
    "        try:\n",
    "            prefix, _, _ = nsm.compute_qname(uri)\n",
    "            return f\"{prefix}:{name}\"\n",
    "        except Exception as e:\n",
    "            return uri\n",
    "    else:\n",
    "        return uri\n",
    "\n",
    "def abbreviate_using_namespaces(study_graph: Graph, violations_graph: Graph):\n",
    "    \"\"\"\n",
    "    Replaces URIs in a DataFrame and a list of violations with QNames if their namespaces\n",
    "    are found in two provided RDF graphs. Saves the updated DataFrame and the list back to disk.\n",
    "\n",
    "    Parameters:\n",
    "    study_graph (Graph): the first RDF graph to use for namespaces\n",
    "    violations_graph (Graph): the second RDF graph to use for namespaces\n",
    "    \"\"\"\n",
    "\n",
    "    # load violation_list\n",
    "    with open(VIOLATION_LIST_FILE, 'r') as f:\n",
    "        violation_list = json.load(f)\n",
    "\n",
    "    # read study_df tabularized data\n",
    "    study_df = pd.read_csv(STUDY_CSV_FILE, index_col=0)\n",
    "\n",
    "    # Create a new graph to combine namespaces\n",
    "    combined_graph = Graph()\n",
    "\n",
    "    # Get namespaces from both graphs and bind them to the combined graph\n",
    "    for prefix, ns_uri in study_graph.namespace_manager.namespaces():\n",
    "        combined_graph.namespace_manager.bind(prefix, ns_uri)\n",
    "\n",
    "    for prefix, ns_uri in violations_graph.namespace_manager.namespaces():\n",
    "        combined_graph.namespace_manager.bind(prefix, ns_uri)\n",
    "\n",
    "    # Now use the NamespaceManager of the combined graph\n",
    "    nsm_combined = NamespaceManager(combined_graph)\n",
    "\n",
    "    # Change column names\n",
    "    for col in study_df.columns:\n",
    "        try:\n",
    "            study_df.rename(columns={col: get_qname(nsm_combined, col)}, inplace=True)\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "    # Change indices and cell values\n",
    "    for col in study_df.columns:\n",
    "        for idx in study_df.index:\n",
    "            try:\n",
    "                study_df.loc[idx, col] = get_qname(nsm_combined, study_df.loc[idx, col])\n",
    "            except:\n",
    "                pass\n",
    "\n",
    "    # change index\n",
    "    for idx in study_df.index:\n",
    "        try:\n",
    "            study_df.rename(index={idx: get_qname(nsm_combined, str(idx))}, inplace=True)\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "    # change violation_list\n",
    "    for i in range(len(violation_list)):\n",
    "        try:\n",
    "            violation_list[i] = get_qname(nsm_combined, violation_list[i])\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "    # store tabular graph as csv\n",
    "    study_df.to_csv(STUDY_CSV_FILE)    \n",
    "\n",
    "    # store violation_list as json\n",
    "    with open(VIOLATION_LIST_FILE, 'w') as f:\n",
    "        json.dump(violation_list, f)\n",
    "\n",
    "abbreviate_using_namespaces(study_g, violations_g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "unique violations (18):\n",
      "\n",
      "omics:OmicsStudyShape-hasContactScientist\n",
      "omics:afb51f95-5b17-45a4-b62d-c58f4998f930\n",
      "omics:TranscriptOmicsSampleShape-hasStrandedness\n",
      "omics:38c0a1c7-1c92-489a-a22e-87f5335ccb1a\n",
      "omics:ece15faa-30b3-471c-ae51-2d6d1f80e3a9\n",
      "omics:OmicsStudyShape-hasResponsibleSite\n",
      "omics:248e9792-8549-4e0d-8665-06c42aff8ffc\n",
      "omics:TranscriptOmicsSampleShape-isPreparedByLibrary\n",
      "omics:TranscriptOmicsSampleShape-hasSequencingProtocol\n",
      "omics:OmicsStudyShape-hasContactCompBio\n",
      "omics:OmicsStudyShape-hasStudyType\n",
      "omics:af998a22-f11b-43ea-b0ac-44e728baeafb\n",
      "omics:9589ebf0-6f4f-4b80-9a5a-75aa25b5715f\n",
      "omics:OmicsStudyShape-isSequencedForIndication\n",
      "omics:TranscriptOmicsSampleShape-isMeasuredOnPlatform\n",
      "omics:OmicsStudyShape-isSequencedForTherapeuticArea\n",
      "omics:OmicsStudyShape-isPartOfProject\n",
      "omics:0c2600b5-30b2-40c7-917a-b402e73f55ae\n"
     ]
    }
   ],
   "source": [
    "with open(VIOLATION_LIST_FILE, 'r') as f:\n",
    "    violation_list = json.load(f)\n",
    "\n",
    "print('unique violations ('+str(len(violation_list))+'):\\n')\n",
    "print('\\n'.join([str(v) for v in violation_list]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>focus_node</th>\n",
       "      <th>omics:TranscriptOmicsSampleShape-isMeasuredOnPlatform</th>\n",
       "      <th>omics:ece15faa-30b3-471c-ae51-2d6d1f80e3a9</th>\n",
       "      <th>omics:TranscriptOmicsSampleShape-hasStrandedness</th>\n",
       "      <th>omics:TranscriptOmicsSampleShape-hasSequencingProtocol</th>\n",
       "      <th>omics:TranscriptOmicsSampleShape-isPreparedByLibrary</th>\n",
       "      <th>omics:248e9792-8549-4e0d-8665-06c42aff8ffc</th>\n",
       "      <th>omics:9589ebf0-6f4f-4b80-9a5a-75aa25b5715f</th>\n",
       "      <th>omics:0c2600b5-30b2-40c7-917a-b402e73f55ae</th>\n",
       "      <th>omics:38c0a1c7-1c92-489a-a22e-87f5335ccb1a</th>\n",
       "      <th>...</th>\n",
       "      <th>omics:externalStudyId</th>\n",
       "      <th>omics:sequencingDate</th>\n",
       "      <th>omics:keywords</th>\n",
       "      <th>omics:isSequencedForTherapeuticArea</th>\n",
       "      <th>omics:studyId</th>\n",
       "      <th>omics:studyTitle</th>\n",
       "      <th>omics:isSequencedForIndication</th>\n",
       "      <th>omics:hasStudyType</th>\n",
       "      <th>omics:hasContactCompBio</th>\n",
       "      <th>omics:hasContactScientist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/TranscriptOmicsSample/sample-EX51-EX51_263</th>\n",
       "      <td>http://data.boehringer.com/uuid/TranscriptOmic...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/PrimaryCellSpecimen/fff3049c-5e43-4d60-905b-4bb483f212b7</th>\n",
       "      <td>http://data.boehringer.com/uuid/PrimaryCellSpe...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/PrimaryCellSpecimen/3843350f-45b7-474d-baa0-9fe2e1fc6642</th>\n",
       "      <td>http://data.boehringer.com/uuid/PrimaryCellSpe...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/PrimaryCellSpecimen/0ff82279-a862-483e-acb6-c175201d6e35</th>\n",
       "      <td>http://data.boehringer.com/uuid/PrimaryCellSpe...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/PrimaryCellSpecimen/2333e9c1-8515-40c3-b39d-881b5cf22aae</th>\n",
       "      <td>http://data.boehringer.com/uuid/PrimaryCellSpe...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                           focus_node  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...  http://data.boehringer.com/uuid/TranscriptOmic...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  http://data.boehringer.com/uuid/PrimaryCellSpe...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  http://data.boehringer.com/uuid/PrimaryCellSpe...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  http://data.boehringer.com/uuid/PrimaryCellSpe...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  http://data.boehringer.com/uuid/PrimaryCellSpe...   \n",
       "\n",
       "                                                    omics:TranscriptOmicsSampleShape-isMeasuredOnPlatform  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                1.0       \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0       \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0       \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0       \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0       \n",
       "\n",
       "                                                    omics:ece15faa-30b3-471c-ae51-2d6d1f80e3a9  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                         1.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         0.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         0.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         0.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         0.0   \n",
       "\n",
       "                                                    omics:TranscriptOmicsSampleShape-hasStrandedness  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                               1.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                               0.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                               0.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                               0.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                               0.0   \n",
       "\n",
       "                                                    omics:TranscriptOmicsSampleShape-hasSequencingProtocol  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                1.0        \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0        \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0        \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0        \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0        \n",
       "\n",
       "                                                    omics:TranscriptOmicsSampleShape-isPreparedByLibrary  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                1.0      \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0      \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0      \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0      \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0      \n",
       "\n",
       "                                                    omics:248e9792-8549-4e0d-8665-06c42aff8ffc  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                         0.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "\n",
       "                                                    omics:9589ebf0-6f4f-4b80-9a5a-75aa25b5715f  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                         0.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "\n",
       "                                                    omics:0c2600b5-30b2-40c7-917a-b402e73f55ae  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                         0.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "\n",
       "                                                    omics:38c0a1c7-1c92-489a-a22e-87f5335ccb1a  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                         0.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                         1.0   \n",
       "\n",
       "                                                    ...  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...  ...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  ...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  ...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  ...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  ...   \n",
       "\n",
       "                                                    omics:externalStudyId  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...         EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...         EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...         EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...         EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...         EdgeNotPresent   \n",
       "\n",
       "                                                    omics:sequencingDate  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...        EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...        EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...        EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...        EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...        EdgeNotPresent   \n",
       "\n",
       "                                                    omics:keywords  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  EdgeNotPresent   \n",
       "\n",
       "                                                    omics:isSequencedForTherapeuticArea  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                       EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                       EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                       EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                       EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                       EdgeNotPresent   \n",
       "\n",
       "                                                     omics:studyId  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  EdgeNotPresent   \n",
       "\n",
       "                                                    omics:studyTitle  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...    EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...    EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...    EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...    EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...    EdgeNotPresent   \n",
       "\n",
       "                                                    omics:isSequencedForIndication  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                  EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                  EdgeNotPresent   \n",
       "\n",
       "                                                    omics:hasStudyType  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...      EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...      EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...      EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...      EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...      EdgeNotPresent   \n",
       "\n",
       "                                                    omics:hasContactCompBio  \\\n",
       "http://data.boehringer.com/uuid/TranscriptOmics...           EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...           EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...           EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...           EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...           EdgeNotPresent   \n",
       "\n",
       "                                                    omics:hasContactScientist  \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...             EdgeNotPresent  \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...             EdgeNotPresent  \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...             EdgeNotPresent  \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...             EdgeNotPresent  \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...             EdgeNotPresent  \n",
       "\n",
       "[5 rows x 64 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study_df = pd.read_csv(STUDY_CSV_FILE, index_col=0)\n",
    "study_df.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compute Violation Embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating UMAP embedding... (this may take some time)\n",
      "UMAP(n_neighbors=72, random_state=0, verbose=True)\n",
      "Mon Oct 23 12:18:17 2023 Construct fuzzy simplicial set\n",
      "Mon Oct 23 12:18:17 2023 Finding Nearest Neighbors\n",
      "Mon Oct 23 12:18:17 2023 Building RP forest with 9 trees\n",
      "Mon Oct 23 12:18:17 2023 NN descent for 12 iterations\n",
      "\t 1  /  12\n",
      "\tStopping threshold met -- exiting after 1 iterations\n",
      "Mon Oct 23 12:18:41 2023 Finished Nearest Neighbor Search\n",
      "Mon Oct 23 12:18:44 2023 Construct embedding\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "24702cc24abc4da2a48b285006aad67e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Epochs completed:   0%|            0/500 [00:00]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mon Oct 23 12:19:51 2023 Finished embedding\n",
      "UMAP embedding created and saved to ./csv/study.csv\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAioAAAGdCAYAAAA8F1jjAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/NK7nSAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAn6UlEQVR4nO3de3xU9Z3/8fckkCGRZCCQZBIJEGAFMQpIAQNKRdiEi1SUul5qhdalgsEuF63QWi62GMQta7Vc6u4K+lPAh/soWG+0ECWsJYACEUFJBUFASEA0MxBhcju/P1ymRBJMYM6c7ySv5+NxHg/nXL7fj19jzjvnfM8Zl2VZlgAAAAwU5XQBAAAA9SGoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACM1cLpAi5VTU2Njhw5ovj4eLlcLqfLAQAADWBZlk6ePKm0tDRFRdV/3STig8qRI0eUnp7udBkAAOAiHDp0SB06dKh3e8QHlfj4eEnf/IsmJCQ4XA0AAGgIv9+v9PT04Hm8PhEfVM7e7klISCCoAAAQYb5r2gaTaQEAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGMvWoJKXl6d+/fopPj5eycnJGjNmjIqLi2vtc+ONN8rlctVaJk6caGdZAAAgQtgaVAoKCpSbm6vNmzdr3bp1qqysVHZ2tsrLy2vtN2HCBB09ejS4LFiwwM6yAABAhLD1PSpr166t9Xn58uVKTk7Wtm3bNHjw4OD6uLg4eb1eO0sBAAARKKxzVHw+nyQpMTGx1vqXXnpJ7du3V2ZmpmbOnKmvv/663jYCgYD8fn+tBYC9qmssFe47oVeLPlfhvhOqrrGcLglAMxG2N9PW1NRoypQpGjRokDIzM4Pr7777bnXq1ElpaWnauXOnHnnkERUXF+tPf/pTne3k5eVp7ty54SobaPbW7jqqX/7PDn15pnY4yR2arof/+RqHqgLQXLgsywrLn0aTJk3SW2+9pXffffeCXz709ttva+jQodq7d6+6du163vZAIKBAIBD8fPa7Anw+H6/QB0Js7a6jmvji9gvuc2D+qDBVA6Ap8fv98ng833n+DssVlcmTJ+v111/Xxo0bLxhSJGnAgAGSVG9QcbvdcrvdttQJ4B+qa6zvDCmS1HnGG/r30T310Gsf1Vq/ZuIg9e7cxqbqADQXtgYVy7L04IMPavXq1dqwYYMyMjK+85iioiJJUmpqqp2lAfiWU2eq9LPnN2vTfl+jj/12SJGkMUv/JokrLgAuja1BJTc3VytWrNCrr76q+Ph4lZSUSJI8Ho9iY2O1b98+rVixQiNHjlS7du20c+dOTZ06VYMHD9Y113DvGwiXH/zhf7XzsD0T0zvPeIOwAuCi2frUz5IlS+Tz+XTjjTcqNTU1uLz88suSpJiYGK1fv17Z2dnq0aOHpk+frrFjx+q1116zsywA57AzpJxVdKDM1vYBNF1hm0xrl4ZOxgFwvlNnqpQ55y9h6YurKgDO1dDzN9/1AzRjP1+5zekSAOCCCCpAM7btYJnTJQDABRFUgGasJkx3fkde2TYs/QBoeggqQDPWM6V1WPr53V0DwtIPgKaHoAI0U2t3HdWWzxr/zpSLERsTHZZ+ADQ9BBWgGWrIq/FDqfOMN8LWF4CmhaACNDPVNZYefCl8IeUswgqAi0FQAZqZTXu/UKVDb0+a+FyBMx0DiFgEFaCZ+dP2w471vfbvp1RRVeNY/wAiD0EFaGbKK6od7f/5TQcc7R9AZCGoAM1Mv87OvtPkvQMnHO0fQGQhqADNzLiBGY72Hxdj65e2A2hiCCpAMxPTIkr3Xd/Zsf7HXtvBsb4BRB6CCtAM/frmq3RNh/B/2/hlMdEa2K192PsFELkIKkAz9efJN+hfB4X3NtDv/qWXoqNcYe0TQGQjqADN2KOje+qBIZ1t76dNbAstvedaDc9Mtb0vAE0Ls9qAZu4/Cw7Y0u7kIV0luZTVtZ2u69KOKykALgpBBWjmKm14/9rKCdcpq2u70DcMoNnh1g+AkOufkeh0CQCaCIIK0Mxlpl4W8ja5zQMgVAgqQDP30oRBTpcAAPUiqADNnCeupTq1iw1Ze/ff0DFkbQEAQQWACh6+6YJhJTGu4fPup+dcFYqSAEASQQXA/yl4+CZ9MCtbvTskyB3tUqsWLt34T4naNSdH22flNOhNtvcPzlBMC36tAAgdl2VZltNFXAq/3y+PxyOfz6eEhPC/EhxoTn7z+m7997sH6tx2/+AMzRzZM7wFAYhYDT1/E1QANEpFVY2W/W2/1n1UKslS9pVejb+eKykAGoegAgAAjNXQ8zd/AgEAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjGVrUMnLy1O/fv0UHx+v5ORkjRkzRsXFxbX2OXPmjHJzc9WuXTu1bt1aY8eOVWlpqZ1lAQCACGFrUCkoKFBubq42b96sdevWqbKyUtnZ2SovLw/uM3XqVL322mt65ZVXVFBQoCNHjui2226zsywAABAhwvpm2uPHjys5OVkFBQUaPHiwfD6fkpKStGLFCv3whz+UJO3Zs0dXXnmlCgsLdd11131nm7yZFgCAyGPkm2l9Pp8kKTExUZK0bds2VVZWatiwYcF9evTooY4dO6qwsLDONgKBgPx+f60FAAA0TWELKjU1NZoyZYoGDRqkzMxMSVJJSYliYmLUpk2bWvumpKSopKSkznby8vLk8XiCS3p6ut2lAwAAh4QtqOTm5mrXrl1atWrVJbUzc+ZM+Xy+4HLo0KEQVQgAAEzTIhydTJ48Wa+//ro2btyoDh06BNd7vV5VVFSorKys1lWV0tJSeb3eOttyu91yu912lwwAAAxg6xUVy7I0efJkrV69Wm+//bYyMjJqbe/bt69atmyp/Pz84Lri4mIdPHhQWVlZdpYGAAAigK1XVHJzc7VixQq9+uqrio+PD8478Xg8io2Nlcfj0X333adp06YpMTFRCQkJevDBB5WVldWgJ34AAEDTZuvjyS6Xq871y5Yt0/jx4yV988K36dOna+XKlQoEAsrJydHixYvrvfXzbTyeDABA5Gno+Tus71GxA0EFAIDIY+R7VAAAABqDoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjGVrUNm4caNGjx6ttLQ0uVwurVmzptb28ePHy+Vy1VqGDx9uZ0kAACCC2BpUysvL1atXLy1atKjefYYPH66jR48Gl5UrV9pZEgAAiCAt7Gx8xIgRGjFixAX3cbvd8nq9dpYBAAAilONzVDZs2KDk5GR1795dkyZN0okTJy64fyAQkN/vr7UAAICmydGgMnz4cL3wwgvKz8/XE088oYKCAo0YMULV1dX1HpOXlyePxxNc0tPTw1gxAAAIJ5dlWVZYOnK5tHr1ao0ZM6befT799FN17dpV69ev19ChQ+vcJxAIKBAIBD/7/X6lp6fL5/MpISEh1GUDAAAb+P1+eTye7zx/O37r51xdunRR+/bttXfv3nr3cbvdSkhIqLUAAICmyaigcvjwYZ04cUKpqalOlwIAAAxg61M/p06dqnV1ZP/+/SoqKlJiYqISExM1d+5cjR07Vl6vV/v27dMvfvELdevWTTk5OXaWBQAAIoStQeX999/XkCFDgp+nTZsmSRo3bpyWLFminTt36vnnn1dZWZnS0tKUnZ2t3/zmN3K73XaWBQAAIkTYJtPapaGTcQAAgDkicjItAADAuQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYtgaVjRs3avTo0UpLS5PL5dKaNWtqbbcsS7NmzVJqaqpiY2M1bNgwffLJJ3aWBAAAIoitQaW8vFy9evXSokWL6ty+YMECPf3001q6dKm2bNmiyy67TDk5OTpz5oydZQEAgAjRws7GR4wYoREjRtS5zbIsPfXUU3r00Ud1yy23SJJeeOEFpaSkaM2aNbrzzjvtLA0AAEQAx+ao7N+/XyUlJRo2bFhwncfj0YABA1RYWFjvcYFAQH6/v9YCAACaJluvqFxISUmJJCklJaXW+pSUlOC2uuTl5Wnu3Lm21gYAQKQ6daZKE//fVv1t31eyJMXHRGntlBt1eWKs06VdFMeCysWaOXOmpk2bFvzs9/uVnp7uYEUAAJjhB3/4X+08XPtOw8mKGg1a8LZaREm/+5feSo5vpb6d2mr9rhI9sGpHrX0fH9lddw/uFs6Sv5NjQcXr9UqSSktLlZqaGlxfWlqq3r1713uc2+2W2+22uzwAACJKXSHlXFU10r+tKrpgG798s1i/fLNYB+aPCnF1F8+xOSoZGRnyer3Kz88PrvP7/dqyZYuysrKcKgsAgIhz6kzVBUNKY3We8UbI2rpUtgaVU6dOqaioSEVFRZK+mUBbVFSkgwcPyuVyacqUKfrtb3+rP//5z/rwww917733Ki0tTWPGjLGzLAAAmoyKqhr94OmNIW939qtbVF1jhbzdxnJZlmVbFRs2bNCQIUPOWz9u3DgtX75clmVp9uzZevbZZ1VWVqbrr79eixcv1hVXXNHgPvx+vzwej3w+nxISEkJZPgAARnvwhb/ptY/KbO3jzusSNX9M6O90NPT8bWtQCQeCCgCgOQr37ZlQz1tp6Pmb7/oBACDCODGHxKl5KwQVAAAiyOa/n3Cs7xlr6n8hq10i7j0qAAA0N9U1lh5dU6iVW79ytI5Vm7/UvB9Yio5yha1PggoAAAZbu+uoJr643ekygrbu/1JZXduFrT9u/QAAYCjTQookHTt5Jqz9EVQAADBQdY1lXEiRpN/n/z2s/RFUAAAw0Nb9XzpdQp0+Pf61Tp2pClt/BBUAAAwU7lssjfHDP6wPW18EFQAADJQc38rpEuq154vqsPVFUAEAwED9MxKdLsEIBBUAAAwUHeXSH+7s7XQZjiOoAABgqJt7X65/7pnsdBnnefDGy8PWF0EFAACDPX3ntU6XcJ7bv3dF2PoiqAAAYLA7/32t0yWcZ/C/vxO2vggqAAAY7AO/0xU4i6ACAACMRVABAADGIqgAAGCwsX3jnS7hPI+P7B62vggqAAAYbMHYG5wu4Tx3D+4Wtr4IKgAAGCw6yqWl95jziPKB+aPC2h9BBQAAww3PTHU8rNzWzxP2kCJJLcLeY4Q7XVGtua99qA3FX8gl6cbuyZo1+irFxkQ7XRoAoAnr29G57/7Z+NAQdWwf50jfBJXvUF1jafO+Eyr89Au9sfOo9p/4utb2le8d0sr3DqlPWqxW//wmh6oEADR1t/xhoyP9toiSYyFFIqhc0MotBzRz9e4G7bvjyGl1nvGGI5fFAABN31F/hSP9/u5fejvS71kElXpc8as3VVFtNfo4wgoAwA4uSY0/K1265PhWDvT6D0ymrcOVv37rokLKWT9/aXMIqwEAQLoq1ZnbL/0znJsbIxFUzlNSdkanK2suqY0/f3hCFVWX1gYAAOd6ccL1jvQbHeVypN+zCCrfct38/JC00/uxv4akHQAAJMkT11Kd2sWGtc+NDw0Ja391Iaic49SZqpC19XVFtfr9dl3I2gMAoODhmy4YVob2SNYzY69Ry3PO7j2TYvS3X9ykbgmN68vpp33OclmW5cTcnJDx+/3yeDzy+XxKSGjkf4Vv+dfn39P6j4+FqLJvLBx7tW7r1zGkbQIAmjff15Uat2yziktOKTrKpRGZXj12y9WNeqfX2l1HNfHF7XVuaxEl7X3c3gdDGnr+Jqic44Yn3tahr06HqLJ/2Pf4SMfv8QEA8G3VNZbe3HZY09Z8qKpqS7Eto7T2374flispDT1/83jyOTyxLW0JKu8WH9f3r0wOebsAAFyK6CiXRvdL1+h+6U6XUi/mqJxj+tArbGn32Xc/taVdAACaOoLKOQZfmSw7btD4z1Ta0CoAAE2f40Flzpw5crlctZYePXo4Ukt0lEvpiaF/9KtXhzYhbxMAgObAiDkqV111ldavXx/83KKFc2V1aBOrg1+Gdp7Kr0b1DGl7AAA0F0YElRYtWsjr9TpdhiRp8D8ladOnX4asvaE92jfqcTEAAPAPjt/6kaRPPvlEaWlp6tKli370ox/p4MGD9e4bCATk9/trLaGU0iZ0t346tYvVf48fELL2AABobhwPKgMGDNDy5cu1du1aLVmyRPv379cNN9ygkydP1rl/Xl6ePB5PcElPD+0jVd6E0HxL5E8HdVbBwzeFpC0AAJor4174VlZWpk6dOmnhwoW67777ztseCAQUCASCn/1+v9LT00Pywjfpm5ff9Ju3Xl+WVzT62ChJM0deqXEDOyumheMZEAAAY0XsC9/atGmjK664Qnv37q1zu9vtltvttq3/6CiXfntLph5YUfdrhS/kE95ACwBASBn3Z/+pU6e0b98+paamOlbDyGtSdf/gjEYds/juawkpAACEmONB5aGHHlJBQYEOHDigTZs26dZbb1V0dLTuuusuR+uaObKnFt/dp0H73j84QyOvcS5YAQDQVDl+6+fw4cO66667dOLECSUlJen666/X5s2blZSU5HRpGnlNmvZlpmrem7v03LvnP4nUKlpaeEcfjbwmzYHqAABo+oybTNtYofz25AuprrG0+dMTKtx3QpKlrC7tdV3XdtzuAQDgIkTsZFpTRUe5NKhbew3q1t7pUgAAaDYcn6MCAABQH4IKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsI4LKokWL1LlzZ7Vq1UoDBgzQ1q1bnS4JAAAYwPGg8vLLL2vatGmaPXu2tm/frl69eiknJ0fHjh1zujQAAOAwx4PKwoULNWHCBP3kJz9Rz549tXTpUsXFxem5555zujQAAOAwR4NKRUWFtm3bpmHDhgXXRUVFadiwYSosLHSwMgAAYIIWTnb+xRdfqLq6WikpKbXWp6SkaM+ePXUeEwgEFAgEgp/9fr+tNQIAAOc4fuunsfLy8uTxeIJLenq60yUBAACbOBpU2rdvr+joaJWWltZaX1paKq/XW+cxM2fOlM/nCy6HDh0KR6kAAMABjgaVmJgY9e3bV/n5+cF1NTU1ys/PV1ZWVp3HuN1uJSQk1FoAAEDT5OgcFUmaNm2axo0bp+9973vq37+/nnrqKZWXl+snP/mJ06UBAACHOR5U7rjjDh0/flyzZs1SSUmJevfurbVr1543wRYAADQ/LsuyLKeLuBR+v18ej0c+n4/bQAAARIiGnr8j7qkfAADQfBBUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxHA0qnTt3lsvlqrXMnz/fyZIAAIBBWjhdwGOPPaYJEyYEP8fHxztYDQAAMInjQSU+Pl5er9fpMgAAgIEcn6Myf/58tWvXTn369NGTTz6pqqoqp0sCAACGcPSKys9//nNde+21SkxM1KZNmzRz5kwdPXpUCxcurPeYQCCgQCAQ/Oz3+8NRKgAAcIDLsiwrlA3OmDFDTzzxxAX3+fjjj9WjR4/z1j/33HO6//77derUKbnd7jqPnTNnjubOnXveep/Pp4SEhIsrGgAAhJXf75fH4/nO83fIg8rx48d14sSJC+7TpUsXxcTEnLd+9+7dyszM1J49e9S9e/c6j63rikp6ejpBBQCACNLQoBLyWz9JSUlKSkq6qGOLiooUFRWl5OTkevdxu931Xm0BAABNi2NzVAoLC7VlyxYNGTJE8fHxKiws1NSpU3XPPfeobdu2TpUFAAAM4lhQcbvdWrVqlebMmaNAIKCMjAxNnTpV06ZNc6okAABgGMeCyrXXXqvNmzc71T0AAIgAjr9HBQAAoD4EFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAY7VwugAAOKu6xtLW/V/q2MkzSo5vpf4ZiYqOcjldFgAHEVQAGGHtrqOa+OL289b/4c7eurn35Q5UBMAEBBUAYVVdY2njx8f0u/y/y3e6Ut298br5Kq+m/M/OOvefvKpIr+48ov+8tx9XXIBmyGVZluV0EZfC7/fL4/HI5/MpISHB6XIAXMDaXUf1wEvbVROi3zqpnlaaPbqnhmemhqZBAGHT0PO3bZNp582bp4EDByouLk5t2rSpc5+DBw9q1KhRiouLU3Jysh5++GFVVVXZVRIAB529tROqkCJJR31nNOnF7Vq762joGgVgFNuCSkVFhW6//XZNmjSpzu3V1dUaNWqUKioqtGnTJj3//PNavny5Zs2aZVdJABxSXWNp8kvnzz8JBUvS3Nc+UnUoExAAY9h+62f58uWaMmWKysrKaq1/6623dPPNN+vIkSNKSUmRJC1dulSPPPKIjh8/rpiYmAa1z60fwFyb9nyhu5dvCUtfKydcp6yu7cLSF4BL5/itn+9SWFioq6++OhhSJCknJ0d+v1+7d++u97hAICC/319rAWCezjPeCFtIkaRjJ8+ErS8A4eNYUCkpKakVUiQFP5eUlNR7XF5enjweT3BJT0+3tU4Ajdd5xhth7zM5vlXY+wRgv0YFlRkzZsjlcl1w2bNnj121SpJmzpwpn88XXA4dOmRrfwAa56n1HzrS71flFY70C8BejXqPyvTp0zV+/PgL7tOlS5cGteX1erV169Za60pLS4Pb6uN2u+V2uxvUB4Dwqq6x9NT6g470/cCK7VoadS2PKgNNTKOCSlJSkpKSkkLScVZWlubNm6djx44pOTlZkrRu3TolJCSoZ8+eIekDQHht3f+lo/1PfHG79j0+kpfAAU2IbXNUDh48qKKiIh08eFDV1dUqKipSUVGRTp06JUnKzs5Wz5499eMf/1gffPCB/vKXv+jRRx9Vbm4uV0yACGXChNbF+bucLgFACNkWVGbNmqU+ffpo9uzZOnXqlPr06aM+ffro/ffflyRFR0fr9ddfV3R0tLKysnTPPffo3nvv1WOPPWZXSQBsZsKE1t/lO3PrCYA9eIU+gJD51+e3av3Hx50uQwfmj3K6BADfwfj3qABoWk5XVBsRUgA0LQQVACEx8vG1TpcgSeoU73QFAEKJoALgklXXWNrv/DxaSdKsH/R1ugQAIURQAXDJnH4s+VyTVm5zugQAIURQAXDJTHgs+ayKGmde4Q/AHgQVAJfMhMeSv63no4QVoCkgqAC4ZP0zEpXcOsbpMmr5uko67g84XQaAS0RQAXDJoqNcumtAJ6fLOM+ti991ugQAl4igAiAkqmpqnC7hPF+WVzpdAoBLRFABEBJHvjrtdAnnSbyspdMlALhEBBUAIZHWJtbpEs6z+oHrnS4BwCUiqAAIiYFd2ztdQi0JrVooKYFvYgciHUEFQEhc17Wd2sSZcasltmWUds7JcboMACFAUAEQEtFRLs2/7Wqny9ATt2Xq49+McLoMACFCUAEQMsMzU7X0nmvldeCWS0p8jPY9PlJ39DfvMWkAF4+gAiCkhmem6m8zhiq5dXhvA/nPVCk6yhXWPgHYj6ACIOSio1zKu+WasPZ5WUx0WPsDEB4EFQC2uPGqFIXzAsf4gZ3D1xmAsCGoALBFdJRLi390bdj6+9n3u4WtLwDhQ1ABYJuzk2vtfkHs/YMzFNOCX2dAU9TC6QIANG3DM1P1zz1HavO+Eyr89AstLdinqhB+LdD9gzM0c2TP0DUIwCguy7Isp4u4FH6/Xx6PRz6fTwkJCU6XA6ABrvz1Wzpd2bi00qal5KuULEkuSRMGd9RD2VdxJQWIUA09f3NFBUDYffybESopO6Pr5ufXWn9jhkup7ZK18v3S4Lp+aXFa9rMb1LoVv66A5ogrKgAAIOwaev7mmikAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMFbEv5P67It1/X6/w5UAAICGOnve/q4X5Ed8UDl58qQkKT093eFKAABAY508eVIej6fe7RH/XT81NTU6cuSI4uPj5XK5nC4nyO/3Kz09XYcOHeI7iGzCGIcH42w/xth+jHF4NGacLcvSyZMnlZaWpqio+meiRPwVlaioKHXo0MHpMuqVkJDA/xQ2Y4zDg3G2H2NsP8Y4PBo6zhe6knIWk2kBAICxCCoAAMBYBBWbuN1uzZ49W2632+lSmizGODwYZ/sxxvZjjMPDjnGO+Mm0AACg6eKKCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKohNiBAwd03333KSMjQ7Gxseratatmz56tioqKWvvt3LlTN9xwg1q1aqX09HQtWLDAoYoj07x58zRw4EDFxcWpTZs2de5z8OBBjRo1SnFxcUpOTtbDDz+sqqqq8BYa4RYtWqTOnTurVatWGjBggLZu3ep0SRFt48aNGj16tNLS0uRyubRmzZpa2y3L0qxZs5SamqrY2FgNGzZMn3zyiTPFRqi8vDz169dP8fHxSk5O1pgxY1RcXFxrnzNnzig3N1ft2rVT69atNXbsWJWWljpUceRZsmSJrrnmmuBL3bKysvTWW28Ft4d6fAkqIbZnzx7V1NToj3/8o3bv3q3/+I//0NKlS/XLX/4yuI/f71d2drY6deqkbdu26cknn9ScOXP07LPPOlh5ZKmoqNDtt9+uSZMm1bm9urpao0aNUkVFhTZt2qTnn39ey5cv16xZs8JcaeR6+eWXNW3aNM2ePVvbt29Xr169lJOTo2PHjjldWsQqLy9Xr169tGjRojq3L1iwQE8//bSWLl2qLVu26LLLLlNOTo7OnDkT5kojV0FBgXJzc7V582atW7dOlZWVys7OVnl5eXCfqVOn6rXXXtMrr7yigoICHTlyRLfddpuDVUeWDh06aP78+dq2bZvef/993XTTTbrlllu0e/duSTaMrwXbLViwwMrIyAh+Xrx4sdW2bVsrEAgE1z3yyCNW9+7dnSgvoi1btszyeDznrX/zzTetqKgoq6SkJLhuyZIlVkJCQq1xR/369+9v5ebmBj9XV1dbaWlpVl5enoNVNR2SrNWrVwc/19TUWF6v13ryySeD68rKyiy3222tXLnSgQqbhmPHjlmSrIKCAsuyvhnTli1bWq+88kpwn48//tiSZBUWFjpVZsRr27at9V//9V+2jC9XVMLA5/MpMTEx+LmwsFCDBw9WTExMcF1OTo6Ki4v11VdfOVFik1NYWKirr75aKSkpwXU5OTny+/3B1I/6VVRUaNu2bRo2bFhwXVRUlIYNG6bCwkIHK2u69u/fr5KSklpj7vF4NGDAAMb8Evh8PkkK/g7etm2bKisra41zjx491LFjR8b5IlRXV2vVqlUqLy9XVlaWLeNLULHZ3r179cwzz+j+++8PrispKal1ApUU/FxSUhLW+poqxvjSfPHFF6qurq5zDBk/e5wdV8Y8dGpqajRlyhQNGjRImZmZkr4Z55iYmPPmtjHOjfPhhx+qdevWcrvdmjhxolavXq2ePXvaMr4ElQaaMWOGXC7XBZc9e/bUOubzzz/X8OHDdfvtt2vChAkOVR45LmaMAaA+ubm52rVrl1atWuV0KU1O9+7dVVRUpC1btmjSpEkaN26cPvroI1v6amFLq03Q9OnTNX78+Avu06VLl+A/HzlyREOGDNHAgQPPmyTr9XrPmwF99rPX6w1NwRGosWN8IV6v97wnVBjjhmvfvr2io6Pr/Dll/OxxdlxLS0uVmpoaXF9aWqrevXs7VFXkmjx5sl5//XVt3LhRHTp0CK73er2qqKhQWVlZrb/6+dlunJiYGHXr1k2S1LdvX7333nv6/e9/rzvuuCPk40tQaaCkpCQlJSU1aN/PP/9cQ4YMUd++fbVs2TJFRdW+cJWVlaVf/epXqqysVMuWLSVJ69atU/fu3dW2bduQ1x4pGjPG3yUrK0vz5s3TsWPHlJycLOmbMU5ISFDPnj1D0kdTFhMTo759+yo/P19jxoyR9M1l9Pz8fE2ePNnZ4pqojIwMeb1e5efnB4OJ3+8P/sWKhrEsSw8++KBWr16tDRs2KCMjo9b2vn37qmXLlsrPz9fYsWMlScXFxTp48KCysrKcKLlJqKmpUSAQsGd8QzThF//n8OHDVrdu3ayhQ4dahw8fto4ePRpcziorK7NSUlKsH//4x9auXbusVatWWXFxcdYf//hHByuPLJ999pm1Y8cOa+7cuVbr1q2tHTt2WDt27LBOnjxpWZZlVVVVWZmZmVZ2drZVVFRkrV271kpKSrJmzpzpcOWRY9WqVZbb7baWL19uffTRR9bPfvYzq02bNrWepELjnDx5MvizKslauHChtWPHDuuzzz6zLMuy5s+fb7Vp08Z69dVXrZ07d1q33HKLlZGRYZ0+fdrhyiPHpEmTLI/HY23YsKHW79+vv/46uM/EiROtjh07Wm+//bb1/vvvW1lZWVZWVpaDVUeWGTNmWAUFBdb+/futnTt3WjNmzLBcLpf117/+1bKs0I8vQSXEli1bZkmqcznXBx98YF1//fWW2+22Lr/8cmv+/PkOVRyZxo0bV+cYv/POO8F9Dhw4YI0YMcKKjY212rdvb02fPt2qrKx0rugI9Mwzz1gdO3a0YmJirP79+1ubN292uqSI9s4779T5cztu3DjLsr55RPnXv/61lZKSYrndbmvo0KFWcXGxs0VHmPp+/y5btiy4z+nTp60HHnjAatu2rRUXF2fdeuuttf6YxIX99Kc/tTp16mTFxMRYSUlJ1tChQ4MhxbJCP74uy7Ksi7sWAwAAYC+e+gEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWP8fWrpxIfwGjC0AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "import umap\n",
    "import pandas as pd\n",
    "import json\n",
    "import numpy as np\n",
    "\n",
    "from scipy.sparse import coo_matrix\n",
    "\n",
    "def create_embedding():\n",
    "    # read study.csv\n",
    "    study_df = pd.read_csv(STUDY_CSV_FILE, index_col=0)\n",
    "    \n",
    "    # read violation_list.json\n",
    "    with open(VIOLATION_LIST_FILE, 'r') as f:\n",
    "        violation_columns = json.load(f)\n",
    "    \n",
    "    view_df = study_df[violation_columns]  # view of study dataframe with only violation columns\n",
    "\n",
    "    print(\"Creating UMAP embedding... (this may take some time)\")\n",
    "    reducer = umap.UMAP(n_neighbors=int(np.sqrt(len(view_df))), min_dist=0.1, n_components=2, random_state=0, verbose=True)\n",
    "    result = reducer.fit_transform(view_df)  # 2D projection\n",
    "\n",
    "    # Check if the result is a tuple and unpack accordingly\n",
    "    if isinstance(result, tuple):\n",
    "        embedding, *_ = result\n",
    "    else:\n",
    "        embedding = result\n",
    "\n",
    "    # Convert to dense numpy array if it's a sparse matrix\n",
    "    if isinstance(embedding, coo_matrix):\n",
    "        embedding = embedding.toarray()\n",
    "\n",
    "    # Assuming that the embedding is now a NumPy array, based on the docstring\n",
    "    study_df[\"x\"] = embedding[:, 0]\n",
    "    study_df[\"y\"] = embedding[:, 1]\n",
    "\n",
    "    study_df.to_csv(STUDY_CSV_FILE)\n",
    "    print(\"UMAP embedding created and saved to\", STUDY_CSV_FILE)\n",
    "\n",
    "\n",
    "def plot_embedding():\n",
    "    study_df = pd.read_csv(STUDY_CSV_FILE, index_col=0)\n",
    "\n",
    "    plt.scatter(study_df[\"x\"], study_df[\"y\"])\n",
    "    plt.show()\n",
    "\n",
    "create_embedding()\n",
    "plot_embedding()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Replace URIs with Label Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "from rdflib import URIRef\n",
    "\n",
    "# list of label predicates in ascending order of priority\n",
    "label_predicates = [\n",
    "    # URIRef('http://data.boehringer.com/ontology/omics/hasValue'),\n",
    "    # URIRef('http://data.boehringer.com/ontology/omics/name'),\n",
    "    URIRef('http://www.w3.org/2004/02/skos/core#prefLabel')\n",
    "]\n",
    "\n",
    "# read tabularized graph .csv as a dataframe, and violations list .json as a list\n",
    "study_df = pd.read_csv(STUDY_CSV_FILE, index_col=0)\n",
    "\n",
    "with open(VIOLATION_LIST_FILE, 'r') as f:\n",
    "    violation_columns = json.load(f)\n",
    "\n",
    "# Create a new graph to combine namespaces\n",
    "combined_graph = Graph()\n",
    "\n",
    "# Get namespaces from both graphs and bind them to the combined graph\n",
    "for prefix, ns_uri in study_g.namespace_manager.namespaces():\n",
    "    combined_graph.namespace_manager.bind(prefix, ns_uri)\n",
    "\n",
    "for prefix, ns_uri in violations_g.namespace_manager.namespaces():\n",
    "    combined_graph.namespace_manager.bind(prefix, ns_uri)\n",
    "\n",
    "# Now use the NamespaceManager of the combined graph\n",
    "nsm_combined = NamespaceManager(combined_graph)\n",
    "\n",
    "# create a dictionary of {s: o} pairs for translating source shapes to their labels\n",
    "label_dict = {}\n",
    "for label_predicate in label_predicates:\n",
    "    temp_dict = {str(s): str(o) for s, p, o in study_g.triples((None, label_predicate, None))}\n",
    "    # replace the temp_dict keys with their corresponding QNames\n",
    "    temp_dict = {get_qname(nsm_combined, k): v for k, v in temp_dict.items()}\n",
    "    # Update label_dict with temp_dict, overwriting existing keys\n",
    "    label_dict.update(temp_dict)\n",
    "\n",
    "# replace all column names, indices, and cell values in the dataframe with their corresponding labels\n",
    "study_df.columns = [label_dict.get(col, col) for col in study_df.columns]\n",
    "study_df.index = pd.Index([label_dict.get(idx, idx) for idx in study_df.index])\n",
    "for col in study_df.columns:\n",
    "    study_df[col] = study_df[col].apply(lambda x: label_dict.get(x, x))\n",
    "\n",
    "# replace all dictionary keys and values with their corresponding labels\n",
    "violation_columns = [label_dict.get(x, x) for x in violation_columns]\n",
    "\n",
    "# store dataframe back to .csv and violations list back to .json\n",
    "study_df.to_csv(STUDY_CSV_FILE)\n",
    "with open(VIOLATION_LIST_FILE, 'w') as f:\n",
    "    json.dump(violation_columns, f)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create Union of Ontology and Violation Report Exemplars"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rdflib import URIRef, RDFS\n",
    "\n",
    "def replace_uris_with_labels(data, label_dict, data_type=\"graph\"):\n",
    "    if data_type == \"graph\":\n",
    "        # Iterate through each triple\n",
    "        for s, p, o in list(data.triples((None, None, None))):\n",
    "            new_s = URIRef(label_dict.get(str(s), s))\n",
    "            new_o = URIRef(label_dict.get(str(o), o))\n",
    "            \n",
    "            if new_s != s or new_o != o:\n",
    "                data.remove((s, p, o))\n",
    "                data.add((new_s, p, new_o))\n",
    "                \n",
    "    elif data_type == \"dict\":\n",
    "        new_data = {}\n",
    "        for key, values in data.items():\n",
    "            new_key = label_dict.get(key, key)\n",
    "            \n",
    "            # If the values are in set form\n",
    "            if isinstance(values, set):\n",
    "                new_values = {label_dict.get(val, val) for val in values}\n",
    "                \n",
    "            # If the values are in dictionary form with concatenated string keys\n",
    "            elif isinstance(values, dict):\n",
    "                new_values = {}\n",
    "                for po_str, count in values.items():\n",
    "                    if '__' in po_str:\n",
    "                        p, o = po_str.split('__')\n",
    "                        new_p = label_dict.get(p, p)\n",
    "                        new_o = label_dict.get(o, o)\n",
    "                        new_po_str = f\"{new_p}__{new_o}\"\n",
    "                        new_values[new_po_str] = count\n",
    "                    else:\n",
    "                        new_po_str = label_dict.get(po_str, po_str)\n",
    "                        new_values[new_po_str] = count\n",
    "            \n",
    "            new_data[new_key] = new_values\n",
    "            \n",
    "        return new_data\n",
    "    \n",
    "    return data\n",
    "\n",
    "\n",
    "\n",
    "def add_labels_to_graph(graph, label_dict):\n",
    "    for uri, label in label_dict.items():\n",
    "        graph.add((URIRef(uri), URIRef('http://www.w3.org/2004/02/skos/core#prefLabel'), Literal(label)))\n",
    "    return graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "986de1c7a565482aa8d215afdb719802",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Processing violations:   0%|          | 0/17911 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from rdflib import Graph\n",
    "import json\n",
    "from routers.utils import get_violation_report_exemplars\n",
    "\n",
    "ontology_union_violation_exemplars_g, edge_count_dict, focus_node_exemplar_dict, exemplar_focus_node_dict, violation_exemplar_dict = get_violation_report_exemplars(ontology_g, violations_g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "http://data.boehringer.com/ontology/omics/afb51f95-5b17-45a4-b62d-c58f4998f930 defaultdict(<class 'int'>, {rdflib.term.URIRef('http://example.com/exemplar#afb51f95-5b17-45a4-b62d-c58f4998f930_exemplar_1'): 1734})\n",
      "http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-isPreparedByLibrary defaultdict(<class 'int'>, {rdflib.term.URIRef('http://example.com/exemplar#TranscriptOmicsSampleShape-isPreparedByLibrary_exemplar_2'): 1734})\n",
      "http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-hasStrandedness defaultdict(<class 'int'>, {rdflib.term.URIRef('http://example.com/exemplar#TranscriptOmicsSampleShape-hasStrandedness_exemplar_3'): 1734})\n",
      "http://data.boehringer.com/ontology/omics/0c2600b5-30b2-40c7-917a-b402e73f55ae defaultdict(<class 'int'>, {rdflib.term.URIRef('http://example.com/exemplar#0c2600b5-30b2-40c7-917a-b402e73f55ae_exemplar_4'): 1734})\n",
      "http://data.boehringer.com/ontology/omics/ece15faa-30b3-471c-ae51-2d6d1f80e3a9 defaultdict(<class 'int'>, {rdflib.term.URIRef('http://example.com/exemplar#ece15faa-30b3-471c-ae51-2d6d1f80e3a9_exemplar_5'): 1734})\n",
      "http://data.boehringer.com/ontology/omics/248e9792-8549-4e0d-8665-06c42aff8ffc defaultdict(<class 'int'>, {rdflib.term.URIRef('http://example.com/exemplar#248e9792-8549-4e0d-8665-06c42aff8ffc_exemplar_6'): 1734})\n",
      "http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-isMeasuredOnPlatform defaultdict(<class 'int'>, {rdflib.term.URIRef('http://example.com/exemplar#TranscriptOmicsSampleShape-isMeasuredOnPlatform_exemplar_7'): 1734})\n",
      "http://data.boehringer.com/ontology/omics/af998a22-f11b-43ea-b0ac-44e728baeafb defaultdict(<class 'int'>, {rdflib.term.URIRef('http://example.com/exemplar#af998a22-f11b-43ea-b0ac-44e728baeafb_exemplar_8'): 564})\n",
      "http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-hasSequencingProtocol defaultdict(<class 'int'>, {rdflib.term.URIRef('http://example.com/exemplar#TranscriptOmicsSampleShape-hasSequencingProtocol_exemplar_9'): 1734})\n",
      "http://data.boehringer.com/ontology/omics/9589ebf0-6f4f-4b80-9a5a-75aa25b5715f defaultdict(<class 'int'>, {rdflib.term.URIRef('http://example.com/exemplar#9589ebf0-6f4f-4b80-9a5a-75aa25b5715f_exemplar_10'): 1734})\n"
     ]
    }
   ],
   "source": [
    "c = 0\n",
    "for k,v in violation_exemplar_dict.items():\n",
    "    if c == 10:\n",
    "        break\n",
    "    print(k, v)\n",
    "    c += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Also replace in joined .ttl of ontology and violations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For the graph\n",
    "ontology_union_violation_exemplars_g = add_labels_to_graph(ontology_union_violation_exemplars_g, label_dict)\n",
    "\n",
    "# For the dictionaries\n",
    "edge_count_dict = replace_uris_with_labels(edge_count_dict, label_dict, data_type=\"dict\")\n",
    "focus_node_exemplar_dict = replace_uris_with_labels(focus_node_exemplar_dict, label_dict, data_type=\"dict\")\n",
    "exemplar_focus_node_dict = replace_uris_with_labels(exemplar_focus_node_dict, label_dict, data_type=\"dict\")\n",
    "# if necessary add code for replacing uris with labels in violation_exemplar_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Store"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Graph identifier=Nbadf4592f1764c7dbb70f44fa6f9f3ea (<class 'rdflib.graph.Graph'>)>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ontology_union_violation_exemplars_g.serialize(destination=OMICS_MODEL_UNION_VIOLATION_EXEMPLAR_TTL_PATH, format=\"ttl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the the current cell or a previous cell. Please review the code in the cell(s) to identify a possible cause of the failure. Click <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. View Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "from routers.utils import save_nested_counts_dict_json, save_lists_dict\n",
    "\n",
    "save_nested_counts_dict_json(edge_count_dict, EXEMPLAR_EDGE_COUNT_DICT_PATH)\n",
    "save_lists_dict(focus_node_exemplar_dict, FOCUS_NODE_EXEMPLAR_DICT_PATH)\n",
    "save_lists_dict(exemplar_focus_node_dict, EXEMPLAR_FOCUS_NODE_DICT_PATH)\n",
    "save_nested_counts_dict_json(violation_exemplar_dict, VIOLATION_EXEMPLAR_DICT_PATH)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py38",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
