{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports & Constants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from rdflib import Graph, Namespace\n",
    "from collections import defaultdict\n",
    "from tqdm.notebook import tqdm\n",
    "import json\n",
    "\n",
    "# Base data path\n",
    "data_path = './data/ex51/'\n",
    "\n",
    "# Function to join paths and create directories if they don't exist\n",
    "def create_and_get_path(*args):\n",
    "    path = os.path.join(*args)\n",
    "    os.makedirs(os.path.dirname(path), exist_ok=True)\n",
    "    return path\n",
    "\n",
    "# Input paths\n",
    "STUDY_TTL_PATH = create_and_get_path(data_path, 'input', 'instances.ttl') # contains the actual study data\n",
    "VIOLATION_REPORT_TTL_PATH = create_and_get_path(data_path, 'input', 'violations.ttl') # contains the shacl violation test case result\n",
    "OMICS_MODEL_TTL_PATH = create_and_get_path(data_path, 'input', 'omics_model.ttl') # contains the ontology\n",
    "\n",
    "# Output paths\n",
    "VIOLATION_LIST_FILE = create_and_get_path(data_path, 'output', 'json', 'violation_list.json') # path to store the violation list\n",
    "STUDY_CSV_FILE = create_and_get_path(data_path, 'output', 'csv', 'study.csv') # path to store the tabularized study csv file\n",
    "OMICS_MODEL_UNION_VIOLATION_EXEMPLAR_TTL_PATH = create_and_get_path(data_path, 'output', 'ttl', 'omics_model_union_violation_exemplar.ttl') # contains the union of the ontology and the violation exemplars\n",
    "EXEMPLAR_EDGE_COUNT_DICT_PATH = create_and_get_path(data_path, 'output', 'json', 'exemplar_edge_count_dict.json') # contains the nested dict of exemplar edge counts\n",
    "FOCUS_NODE_EXEMPLAR_DICT_PATH = create_and_get_path(data_path, 'output', 'json', 'focus_node_exemplar_dict.json') # contains the dict of exemplars per focus node\n",
    "EXEMPLAR_FOCUS_NODE_DICT_PATH = create_and_get_path(data_path, 'output', 'json', 'exemplar_focus_node_dict.json') # contains the dict of focus nodes per exemplar\n",
    "VIOLATION_EXEMPLAR_DICT_PATH = create_and_get_path(data_path, 'output', 'json', 'violation_exemplar_dict.json') # contains the dict of exemplars and their counts per violation\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read & Parse Graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading study graph...\n",
      "Reading violations graph...\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Graph identifier=N0d4278d93ed64835a498a77633d2a7b1 (<class 'rdflib.graph.Graph'>)>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Reading study graph...\")\n",
    "study_g = Graph()\n",
    "study_g.parse(STUDY_TTL_PATH, format=\"ttl\")\n",
    "\n",
    "print(\"Reading violations graph...\")\n",
    "violations_g = Graph()\n",
    "violations_g.parse(VIOLATION_REPORT_TTL_PATH, format=\"ttl\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Give Enumarting Names to Unnamed NodeShapes\n",
    "example: [] a sh:NodeShape -> NodeShape1 a sh:NodeShape\n",
    "etc\n",
    "this could be done with rdflib but then when storing the graph to a .ttl file it would automatically mess with the namespaces and remove unused ones\n",
    "so until these names are provided by BI, we decided to do simple string replacement"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"color:orange\">Careful: If the omics_model.ttl contains [] anywhere else, e.g., in a label, it will also be replaced there!</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "# Read the input file\n",
    "with open(OMICS_MODEL_TTL_PATH, 'r') as file:\n",
    "    data = file.read()\n",
    "\n",
    "# Define backup file path\n",
    "backup_file_path = OMICS_MODEL_TTL_PATH + \".bak\"\n",
    "\n",
    "# Ensure the output directory exists\n",
    "os.makedirs(os.path.dirname(OMICS_MODEL_TTL_PATH), exist_ok=True)\n",
    "\n",
    "# store backup file\n",
    "with open(backup_file_path, 'w') as file:\n",
    "    file.write(data)\n",
    "\n",
    "# Initialize the counter\n",
    "counter = 1\n",
    "\n",
    "# Prepend '@prefix cns: <http://custom_node_shapes/> .'\n",
    "data = \"@prefix cns: <http://customnamespace.com/> .\\n\" + data\n",
    "\n",
    "# Replace occurrences of [] with NodeShape{counter}\n",
    "while \"[]\" in data:\n",
    "    data = data.replace(\"[]\", f\"cns:NodeShape{counter}\", 1)\n",
    "    counter += 1\n",
    "\n",
    "# Write the updated string to the new file\n",
    "with open(OMICS_MODEL_TTL_PATH, 'w') as file:\n",
    "    file.write(data)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Differentiate NaN Cases - Predicate cannot Exist on Class vs. Predicate is Missing vs. Predicta Exists Even Though It Cannot\n",
    "A node can have an edge if\n",
    "\n",
    "s1 a sh:NodeShape\n",
    "\n",
    "s1 sh:targetClass node\n",
    "\n",
    "s1 sh:property s2\n",
    "\n",
    "s2 a sh:PropertyShape\n",
    "\n",
    "s2 sh:Path edge\n",
    "\n",
    "OR \n",
    "\n",
    "(\n",
    "node rdfs:subClassOf s3\n",
    "AND\n",
    "s3 can have the edge\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rdflib import Graph\n",
    "from rdflib.namespace import Namespace\n",
    "import networkx as nx\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# read ontology into a graph\n",
    "ontology_g = Graph()\n",
    "ontology_g.parse(OMICS_MODEL_TTL_PATH, format=\"ttl\")\n",
    "\n",
    "# take namespace prefix for sh and omics from the graph\n",
    "sh = Namespace(\"http://www.w3.org/ns/shacl#\")\n",
    "omics = Namespace(\"http://data.boehringer.com/ontology/omics#\")\n",
    "owl = Namespace(\"http://www.w3.org/2002/07/owl#\")\n",
    "\n",
    "# define prefixes and corresponding namespaces\n",
    "prefixes = {\n",
    "    \"sh\": \"http://www.w3.org/ns/shacl#\",\n",
    "    \"omics\": \"http://data.boehringer.com/ontology/omics#\",\n",
    "    \"owl\": \"http://www.w3.org/2002/07/owl#\"\n",
    "}\n",
    "\n",
    "def convert_to_prefixed(graph, uri):\n",
    "    \"\"\"Convert a URI to its prefixed form using a rdflib graph's namespace manager.\"\"\"\n",
    "    prefix, namespace, name = graph.namespace_manager.compute_qname(uri)\n",
    "    return f\"{prefix}:{name}\"\n",
    "\n",
    "def select_direct_possible_edges(node):\n",
    "    select_statement = \"\"\"\n",
    "    SELECT ?edge\n",
    "    WHERE {\n",
    "        ?ns a sh:NodeShape .\n",
    "        ?ns sh:targetClass omics:%s .\n",
    "        ?ns sh:property ?ps .\n",
    "        ?ps a sh:PropertyShape .\n",
    "        ?ps sh:path ?edge .\n",
    "    }\n",
    "    \"\"\" % node\n",
    "    return [str(result[0].n3(ontology_g.namespace_manager)) for result in ontology_g.query(select_statement)] # type: ignore\n",
    "\n",
    "def select_direct_parents(node):\n",
    "    select_statement = \"\"\"\n",
    "    SELECT ?parent\n",
    "    WHERE {\n",
    "        omics:%s rdfs:subClassOf ?parent .\n",
    "    }\n",
    "    \"\"\" % node\n",
    "    return [str(result[0].n3(ontology_g.namespace_manager)) for result in ontology_g.query(select_statement)] # type: ignore\n",
    "\n",
    "def select_direct_children(node):\n",
    "    select_statement = \"\"\"\n",
    "    SELECT ?child\n",
    "    WHERE {\n",
    "        ?child rdfs:subClassOf omics:%s .\n",
    "    }\n",
    "    \"\"\" % node\n",
    "    return [str(result[0].n3(ontology_g.namespace_manager)) for result in ontology_g.query(select_statement)] # type: ignore\n",
    "\n",
    "def select_roots():\n",
    "    select_statement = \"\"\"\n",
    "    SELECT ?root\n",
    "    WHERE {\n",
    "        ?root rdfs:subClassOf owl:Thing .\n",
    "    }\n",
    "    \"\"\"\n",
    "    return [str(result[0].n3(ontology_g.namespace_manager)) for result in ontology_g.query(select_statement)] # type: ignore\n",
    "\n",
    "def select_root_classes():\n",
    "    select_statement = \"\"\"\n",
    "    SELECT ?class\n",
    "    WHERE {\n",
    "        ?class a owl:Class .\n",
    "        FILTER NOT EXISTS { ?class rdfs:subClassOf ?parent . }\n",
    "    }\n",
    "    \"\"\"\n",
    "    return [str(result[0].n3(ontology_g.namespace_manager)) for result in ontology_g.query(select_statement)] # type: ignore\n",
    "\n",
    "\n",
    "# use the functions\n",
    "roots = select_root_classes()\n",
    "#commented out block 2: by L.B\n",
    "#edges = select_direct_possible_edges('PrimaryCellSpecimen')\n",
    "#parents = select_direct_parents('BiopsySpecimen')\n",
    "#children = select_direct_children('BiopsySpecimen')\n",
    "\n",
    "#print('Roots:', roots)\n",
    "#print('Edges:', edges)\n",
    "#print('Parents:', parents)\n",
    "#print('Children:', children)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "class OntologyTree:\n",
    "    def __init__(self):\n",
    "        self.tree = {}\n",
    "\n",
    "    def add_node(self, node, edges, children):\n",
    "        if node not in self.tree:\n",
    "            self.tree[node] = {'edges': edges, 'children': children}\n",
    "\n",
    "    def add_child(self, parent, child):\n",
    "        if parent in self.tree:\n",
    "            self.tree[parent]['children'].append(child)\n",
    "        else:\n",
    "            self.tree[parent] = {'edges': [], 'children': [child]}\n",
    "\n",
    "def aggregate_edges(node, tree):\n",
    "    edges = set(tree[node]['edges'])\n",
    "    queue = [parent_node for parent_node in tree if node in tree[parent_node]['children']]\n",
    "    while queue:\n",
    "        parent_node = queue.pop(0)\n",
    "        edges |= set(tree[parent_node]['edges'])\n",
    "        queue += [grand_parent for grand_parent in tree if parent_node in tree[grand_parent]['children']]\n",
    "    return edges\n",
    "\n",
    "def create_aggregated_edges_dict(tree):\n",
    "    agg_edges_dict = {}\n",
    "    for node in tree.keys():\n",
    "        agg_edges_dict[node] = aggregate_edges(node, tree)\n",
    "    return agg_edges_dict\n",
    "\n",
    "def convert_to_full_uri(graph, abbreviated_uri):\n",
    "    \"\"\"Converts a prefixed URI to a full URI using a rdflib graph's namespace manager.\"\"\"\n",
    "    prefix, uri = abbreviated_uri.split(':')\n",
    "    namespace = graph.namespace_manager.store.namespace(prefix)\n",
    "    return namespace + uri\n",
    "\n",
    "def remove_prefix(uri):\n",
    "    if ':' in uri:\n",
    "        return uri.split(':', 1)[1]\n",
    "    return uri\n",
    "\n",
    "def build_tree(root_nodes, ontology_tree):\n",
    "    queue = root_nodes\n",
    "    while queue:\n",
    "        current_node = queue.pop(0)\n",
    "        unprefixed_node = remove_prefix(current_node)\n",
    "        edges = select_direct_possible_edges(unprefixed_node)\n",
    "        children = select_direct_children(unprefixed_node)\n",
    "\n",
    "        ontology_tree.add_node(current_node, edges, children)\n",
    "\n",
    "        for child in children:\n",
    "            queue.append(child)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "roots = select_root_classes()\n",
    "ontology_tree = OntologyTree()\n",
    "build_tree(roots, ontology_tree)\n",
    "agg_edges_dict = create_aggregated_edges_dict(ontology_tree.tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def select_all_focus_nodes_and_classes(graph):\n",
    "    select_statement = \"\"\"\n",
    "    SELECT ?node ?class\n",
    "    WHERE {\n",
    "        ?node a ?class .\n",
    "    }\n",
    "    \"\"\"\n",
    "    results = graph.query(select_statement)\n",
    "    focus_nodes_classes_dict = {str(result[0]): str(result[1]) for result in results}\n",
    "    return focus_nodes_classes_dict\n",
    "\n",
    "\n",
    "focus_nodes_class_dict = select_all_focus_nodes_and_classes(study_g)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tabularize & Join Study + Violations Graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "study graph triples: 97794\n",
      "violations graph triples: 286592\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing Graph Once: 100%|██████████| 286592/286592 [00:01<00:00, 212378.15it/s]\n",
      "Going Over Violation Instances: 100%|██████████| 17911/17911 [00:00<00:00, 987826.15it/s]\n",
      "Processing study: 100%|██████████| 97794/97794 [00:02<00:00, 36707.05it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count_skipped = 19186\n",
      "count_not_skipped = 78608\n",
      "count_allowed = 50295\n",
      "count_not_allowed_exists = 28313\n",
      "count_not_allowed_empty = 0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>focus_node</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/38c0a1c7-1c92-489a-a22e-87f5335ccb1a</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/248e9792-8549-4e0d-8665-06c42aff8ffc</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/9589ebf0-6f4f-4b80-9a5a-75aa25b5715f</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/0c2600b5-30b2-40c7-917a-b402e73f55ae</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/afb51f95-5b17-45a4-b62d-c58f4998f930</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-isMeasuredOnPlatform</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-hasSequencingProtocol</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/ece15faa-30b3-471c-ae51-2d6d1f80e3a9</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-isPreparedByLibrary</th>\n",
       "      <th>...</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/technicalStudyTitle</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/isPartOfProject</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/studyId</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/hasStudyType</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/sequencingDate</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/isSequencedForIndication</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/hasResponsibleSite</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/hasContactScientist</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/studyDescription</th>\n",
       "      <th>http://data.boehringer.com/ontology/omics/hasContactCompBio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/PrimaryCellSpecimen/0444023a-6a21-4afe-bb75-78941dfb5766</th>\n",
       "      <td>http://data.boehringer.com/uuid/PrimaryCellSpe...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/Donor/946b8e10-ad3c-492d-9c50-aff294cd122b</th>\n",
       "      <td>http://data.boehringer.com/uuid/Donor/946b8e10...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/Donor/e441da98-c03e-4688-8999-d2b048b3d4ad</th>\n",
       "      <td>http://data.boehringer.com/uuid/Donor/e441da98...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/PrimaryCellSpecimen/16ad2a0e-37e5-496a-960a-ca33df46d9cd</th>\n",
       "      <td>http://data.boehringer.com/uuid/PrimaryCellSpe...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>http://data.boehringer.com/uuid/TranscriptOmicsSample/sample-EX51-EX51_902</th>\n",
       "      <td>http://data.boehringer.com/uuid/TranscriptOmic...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "      <td>EdgeNotPresent</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                           focus_node  \\\n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  http://data.boehringer.com/uuid/PrimaryCellSpe...   \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...  http://data.boehringer.com/uuid/Donor/946b8e10...   \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...  http://data.boehringer.com/uuid/Donor/e441da98...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  http://data.boehringer.com/uuid/PrimaryCellSpe...   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...  http://data.boehringer.com/uuid/TranscriptOmic...   \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/38c0a1c7-1c92-489a-a22e-87f5335ccb1a  \\\n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                0.0                                \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/248e9792-8549-4e0d-8665-06c42aff8ffc  \\\n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                0.0                                \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/9589ebf0-6f4f-4b80-9a5a-75aa25b5715f  \\\n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                0.0                                \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/0c2600b5-30b2-40c7-917a-b402e73f55ae  \\\n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                0.0                                \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/afb51f95-5b17-45a4-b62d-c58f4998f930  \\\n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...                                                1.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                0.0                                \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-isMeasuredOnPlatform  \\\n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                           \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...                                                0.0                                           \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...                                                0.0                                           \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                           \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                1.0                                           \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-hasSequencingProtocol  \\\n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                            \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...                                                0.0                                            \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...                                                0.0                                            \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                            \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                1.0                                            \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/ece15faa-30b3-471c-ae51-2d6d1f80e3a9  \\\n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                1.0                                \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-isPreparedByLibrary  \\\n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                          \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...                                                0.0                                          \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...                                                0.0                                          \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                                0.0                                          \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                                1.0                                          \n",
       "\n",
       "                                                    ...  \\\n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  ...   \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...  ...   \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...  ...   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...  ...   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...  ...   \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/technicalStudyTitle  \\\n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent               \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...                                     EdgeNotPresent               \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...                                     EdgeNotPresent               \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent               \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent               \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/isPartOfProject  \\\n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent           \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...                                     EdgeNotPresent           \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...                                     EdgeNotPresent           \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent           \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent           \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/studyId  \\\n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...                                     EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...                                     EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent   \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent   \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/hasStudyType  \\\n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent        \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...                                     EdgeNotPresent        \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...                                     EdgeNotPresent        \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent        \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent        \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/sequencingDate  \\\n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent          \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...                                     EdgeNotPresent          \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...                                     EdgeNotPresent          \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent          \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent          \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/isSequencedForIndication  \\\n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent                    \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...                                     EdgeNotPresent                    \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...                                     EdgeNotPresent                    \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent                    \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent                    \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/hasResponsibleSite  \\\n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent              \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...                                     EdgeNotPresent              \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...                                     EdgeNotPresent              \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent              \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent              \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/hasContactScientist  \\\n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent               \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...                                     EdgeNotPresent               \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...                                     EdgeNotPresent               \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent               \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent               \n",
       "\n",
       "                                                    http://data.boehringer.com/ontology/omics/studyDescription  \\\n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent            \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...                                     EdgeNotPresent            \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...                                     EdgeNotPresent            \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent            \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent            \n",
       "\n",
       "                                                   http://data.boehringer.com/ontology/omics/hasContactCompBio  \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent           \n",
       "http://data.boehringer.com/uuid/Donor/946b8e10-...                                     EdgeNotPresent           \n",
       "http://data.boehringer.com/uuid/Donor/e441da98-...                                     EdgeNotPresent           \n",
       "http://data.boehringer.com/uuid/PrimaryCellSpec...                                     EdgeNotPresent           \n",
       "http://data.boehringer.com/uuid/TranscriptOmics...                                     EdgeNotPresent           \n",
       "\n",
       "[5 rows x 64 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from rdflib import Graph, Namespace\n",
    "from rdflib.namespace import NamespaceManager\n",
    "from collections import defaultdict\n",
    "import json\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "empty_edge_string = \"EdgeNotPresent\"\n",
    "\n",
    "SH = Namespace(\"http://www.w3.org/ns/shacl#\")\n",
    "\n",
    "def count_violations(violations: Graph):\n",
    "    d_focus_node_d_source_shape_counts = defaultdict(lambda: defaultdict(int))\n",
    "    d_violation_focus_node = defaultdict((str))\n",
    "    d_violation_source_shape = defaultdict((str))\n",
    "    violation_list = []\n",
    "\n",
    "    for s, p, o in tqdm(violations, desc=\"Processing Graph Once\"):\n",
    "        if p == SH.sourceShape:\n",
    "            d_violation_source_shape[str(s)] = str(o)\n",
    "            violation_list += [str(o)]\n",
    "        elif p == SH.focusNode:\n",
    "            d_violation_focus_node[str(s)] = str(o)\n",
    "\n",
    "    # for key in violation_focus_node dict\n",
    "    for key in tqdm(d_violation_focus_node, desc=\"Going Over Violation Instances\"):\n",
    "        # init with 1 or add 1 if already exists\n",
    "        d_focus_node_d_source_shape_counts[d_violation_focus_node[key]][d_violation_source_shape[key]] += 1\n",
    "\n",
    "    violation_list = list(set(violation_list))\n",
    "    return d_focus_node_d_source_shape_counts, violation_list\n",
    "\n",
    "\n",
    "def save_violation_list(violation_list):\n",
    "    with open(VIOLATION_LIST_FILE, 'w') as f:\n",
    "        json.dump(violation_list, f)  \n",
    "\n",
    "\n",
    "def create_study_dataframe(study: Graph, d_focus_node_d_source_shape_counts, focus_nodes_class_dict, agg_edges_dict):\n",
    "    # for convenience\n",
    "    pre = lambda x: convert_to_prefixed(ontology_g, str(x))\n",
    "\n",
    "    # create an empty study_df, fill the violation counts\n",
    "    study_df = pd.DataFrame(d_focus_node_d_source_shape_counts).T.fillna(0)\n",
    "\n",
    "    count_not_skipped, count_skipped, count_allowed, count_not_allowed_exists, count_not_allowed_empty = 0, 0, 0, 0, 0\n",
    "\n",
    "    # iterate all study triples\n",
    "    for s, p, o in tqdm(study, desc=\"Processing study\"):\n",
    "        # if s is not in d_focus_node_d_source_shape_counts, continue as it is not a focus node\n",
    "        if str(s) not in d_focus_node_d_source_shape_counts:\n",
    "            count_skipped += 1\n",
    "            continue\n",
    "        count_not_skipped += 1\n",
    "\n",
    "        # if p is not in study_df.columns, add it and set it to empty_edge_string for all rows because so far there has not been a value for it for anything, and future values will overwrite it anyway\n",
    "        if str(p) not in study_df.columns:\n",
    "            study_df[str(p)] = empty_edge_string\n",
    "\n",
    "        # if object is empty throw exception and print details\n",
    "        if str(o) == \"\":\n",
    "            print(\"empty object for \" + str(s) + \" \" + str(p))\n",
    "            raise Exception(\"empty object for \" + str(s) + \" \" + str(p))\n",
    "\n",
    "        # get class of focus node\n",
    "        focus_node_class = focus_nodes_class_dict.get(str(s), None)\n",
    "        assert focus_node_class is not None, f\"The focus_node_class for {str(s)} was None, but expected a value.\"\n",
    "        # get allowed edges\n",
    "        focus_node_edges = agg_edges_dict.get(convert_to_prefixed(ontology_g, focus_node_class), set())\n",
    "        # assert len(focus_node_edges) > 0\n",
    "\n",
    "        def write_object_to_study_df(study_df, s, p, o):\n",
    "            current_value = study_df.at[str(s), str(p)]\n",
    "            # Check if the current cell already contains a value that is not empty_edge_string\n",
    "            if current_value == empty_edge_string:\n",
    "                # If the cell is empty, directly assign the new value\n",
    "                study_df.at[str(s), str(p)] = str(o)\n",
    "            else:\n",
    "                # If the cell already has a value, ensure it's a list and append the new value\n",
    "                if not isinstance(current_value, list):\n",
    "                    current_value = [current_value] if current_value != str(o) else current_value\n",
    "                if isinstance(current_value, list):\n",
    "                    current_value.append(str(o))\n",
    "                # Update the DataFrame cell with the new value\n",
    "                study_df.at[str(s), str(p)] = current_value\n",
    "\n",
    "\n",
    "        # if edge is allowed (and object is present otherwise we wouldn't get here), set value study_df.at[str(s), str(p)] = str(o)\n",
    "        if pre(p) in focus_node_edges:\n",
    "            write_object_to_study_df(study_df, s, p, o)\n",
    "            count_allowed += 1\n",
    "        # if edge is not allowed, and there is no object, set value study_df.at[str(s), str(p)] = empty_edge_string\n",
    "        elif pre(p) not in focus_node_edges and str(o) == \"\":\n",
    "            study_df.at[str(s), str(p)] = empty_edge_string\n",
    "            count_not_allowed_empty += 1\n",
    "        # if edge is not allowed, and there is an object, set value study_df.at[str(s), str(p)] = \"InvalidEdgeWUnexpectedEdgeWithDataithData: str(o)\"\n",
    "        elif pre(p) not in focus_node_edges and str(o) != \"\":\n",
    "            # study_df.at[str(s), str(p)] = \"UnexpectedEdgeWithData: \" + str(o)\n",
    "            write_object_to_study_df(study_df, s, p, o)\n",
    "            count_not_allowed_exists += 1\n",
    "        else:\n",
    "            raise Exception(\"should not happen: \" + str(s) + \" \" + str(p) + \" \" + str(o))\n",
    "\n",
    "    print(f\"count_skipped = {count_skipped}\")\n",
    "    print(f\"count_not_skipped = {count_not_skipped}\")\n",
    "    print(f\"count_allowed = {count_allowed}\")\n",
    "    print(f\"count_not_allowed_exists = {count_not_allowed_exists}\")\n",
    "    print(f\"count_not_allowed_empty = {count_not_allowed_empty}\")\n",
    "\n",
    "    assert study_df.isin(['']).sum().sum() == 0\n",
    "    assert study_df.isnull().sum().sum() == 0, f\"study_df.isnull().sum().sum() = {study_df.isnull().sum().sum()}\"\n",
    "\n",
    "    # add focus_node column, set it to the study_df.index, and move it to the very left of the dataframe\n",
    "    study_df['focus_node'] = study_df.index\n",
    "    cols = study_df.columns.tolist()\n",
    "    cols = cols[-1:] + cols[:-1]  # move the last column to the first\n",
    "    study_df = study_df[cols]\n",
    "\n",
    "    return study_df\n",
    "\n",
    "\n",
    "def save_study_dataframe(study_df):\n",
    "    study_df.to_csv(STUDY_CSV_FILE)\n",
    "\n",
    "def replace_empty_edge_with_empty_array(df, empty_edge_string=\"EdgeNotPresent\"):\n",
    "    # Iterate through each column of the DataFrame\n",
    "    for column in df.columns:\n",
    "        # Determine if the column is a set column (contains at least one list)\n",
    "        is_set_column = any(df[column].apply(lambda x: isinstance(x, list)))\n",
    "        print(is_set_column, column)\n",
    "        \n",
    "        # Apply conditional replacement based on whether the column is a set column\n",
    "        if is_set_column:\n",
    "            # For set columns, replace empty_edge_string with an empty list\n",
    "            df[column] = df[column].apply(lambda x: [] if x == empty_edge_string else x)\n",
    "        else:\n",
    "            # For non-set columns, ensure that empty_edge_string remains for missing values\n",
    "            # This else block might not perform any action, but it's included for clarity.\n",
    "            # Adjust or remove based on your specific requirements or if further logic is needed.\n",
    "            pass\n",
    "    return df\n",
    "\n",
    "def post_process_convert_to_arrays(df, empty_edge_string=\"EdgeNotPresent\"):\n",
    "    for column in df.columns:\n",
    "        # Check if any cell in the column contains a list\n",
    "        contains_list = df[column].apply(lambda x: isinstance(x, list)).any()\n",
    "        if contains_list:\n",
    "            # Convert all non-list cells in the column into a list\n",
    "            df[column] = df[column].apply(lambda x: [x] if not isinstance(x, list) else x)\n",
    "        else:\n",
    "            # For columns without lists, no action is needed\n",
    "            pass\n",
    "    return df\n",
    "\n",
    "\n",
    "def tabularize_graphs(study: Graph, violations: Graph):\n",
    "    print('study graph triples:', len(study))  # type: ignore\n",
    "    print('violations graph triples:', len(violations))  # type: ignore\n",
    "    d_focus_node_d_source_shape_counts, violation_list = count_violations(violations)\n",
    "    save_violation_list(violation_list)\n",
    "    study_df = create_study_dataframe(study, d_focus_node_d_source_shape_counts, focus_nodes_class_dict, agg_edges_dict)\n",
    "    study_df = post_process_convert_to_arrays(study_df, empty_edge_string)\n",
    "    # study_df = replace_empty_edge_with_empty_array(study_df, empty_edge_string)    \n",
    "    save_study_dataframe(study_df)\n",
    "    return d_focus_node_d_source_shape_counts\n",
    "\n",
    "\n",
    "d_focus_node_d_source_shape_counts = tabularize_graphs(study_g, violations_g)\n",
    "\n",
    "study_df = pd.read_csv(STUDY_CSV_FILE, index_col=0)\n",
    "study_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "focus_node ['http://data.boehringer.com/uuid/PrimaryCellSpecimen/0444023a-6a21-4afe-bb75-78941dfb5766'\n",
      " 'http://data.boehringer.com/uuid/Donor/946b8e10-ad3c-492d-9c50-aff294cd122b'\n",
      " 'http://data.boehringer.com/uuid/Donor/e441da98-c03e-4688-8999-d2b048b3d4ad'\n",
      " ...\n",
      " 'http://data.boehringer.com/uuid/Donor/9d9a89be-08f3-40a4-8545-f4f2f5598334'\n",
      " 'http://data.boehringer.com/uuid/Donor/5dd45b95-d5f6-4301-976c-0a4ad50e1ca8'\n",
      " 'http://data.boehringer.com/uuid/Donor/c4e77f24-c57e-4b52-a8e4-ea4dcf16f61c']\n",
      "http://data.boehringer.com/ontology/omics/38c0a1c7-1c92-489a-a22e-87f5335ccb1a [1. 0.]\n",
      "http://data.boehringer.com/ontology/omics/248e9792-8549-4e0d-8665-06c42aff8ffc [1. 0.]\n",
      "http://data.boehringer.com/ontology/omics/9589ebf0-6f4f-4b80-9a5a-75aa25b5715f [1. 0.]\n",
      "http://data.boehringer.com/ontology/omics/0c2600b5-30b2-40c7-917a-b402e73f55ae [1. 0.]\n",
      "http://data.boehringer.com/ontology/omics/afb51f95-5b17-45a4-b62d-c58f4998f930 [0. 1.]\n",
      "http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-isMeasuredOnPlatform [0. 1.]\n",
      "http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-hasSequencingProtocol [0. 1.]\n",
      "http://data.boehringer.com/ontology/omics/ece15faa-30b3-471c-ae51-2d6d1f80e3a9 [0. 1.]\n",
      "http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-isPreparedByLibrary [0. 1.]\n",
      "http://data.boehringer.com/ontology/omics/TranscriptOmicsSampleShape-hasStrandedness [0. 1.]\n",
      "http://data.boehringer.com/ontology/omics/af998a22-f11b-43ea-b0ac-44e728baeafb [0. 1.]\n",
      "http://data.boehringer.com/ontology/omics/OmicsStudyShape-isPartOfProject [0. 1.]\n",
      "http://data.boehringer.com/ontology/omics/OmicsStudyShape-hasResponsibleSite [0. 1.]\n",
      "http://data.boehringer.com/ontology/omics/OmicsStudyShape-hasContactCompBio [0. 1.]\n",
      "http://data.boehringer.com/ontology/omics/OmicsStudyShape-isSequencedForTherapeuticArea [0. 1.]\n",
      "http://data.boehringer.com/ontology/omics/OmicsStudyShape-isSequencedForIndication [0. 1.]\n",
      "http://data.boehringer.com/ontology/omics/OmicsStudyShape-hasContactScientist [0. 1.]\n",
      "http://data.boehringer.com/ontology/omics/OmicsStudyShape-hasStudyType [0. 1.]\n",
      "http://data.boehringer.com/ontology/omics/hasSpecimenPreparation ['http://vocabulary.boehringer.com/OMICS/12eec321-ca7e-4866-982f-0a0677fed966'\n",
      " 'EdgeNotPresent']\n",
      "http://data.boehringer.com/ontology/omics/hasSpecimenType ['http://vocabulary.boehringer.com/OMICS/14884101-7675-49ca-b50c-32da5aeb965a'\n",
      " 'EdgeNotPresent']\n",
      "http://data.boehringer.com/ontology/omics/isPreparedByLibrary ['EdgeNotPresent'\n",
      " 'http://vocabulary.boehringer.com/OMICS/ac828eb0-cb7b-4f5b-866e-197bcf7a8e32']\n",
      "http://data.boehringer.com/ontology/omics/name ['Primary Cell Specimen EX51_702_CD1CminusCD141minus'\n",
      " 'Donor EX51_1597_CD123NEG_D7' 'Donor EX51_305_CD141' ...\n",
      " 'Donor EX51_787_Mono_classical' 'Donor EX51_845_Mono_classical'\n",
      " 'Donor EX51_1539_CD123NEG_D7']\n",
      "http://data.boehringer.com/ontology/omics/isPartOfSampleSet ['EdgeNotPresent'\n",
      " 'http://data.boehringer.com/uuid/TranscriptOmicsSampleSet/3f78957e-d666-4a6d-8e39-66768d9e9807'\n",
      " 'http://data.boehringer.com/uuid/TranscriptOmicsSampleSet/1cfa3874-3969-4abb-9778-406d02c9eebe'\n",
      " 'http://data.boehringer.com/uuid/TranscriptOmicsSampleSet/988f9edf-6cb1-4459-a3bf-fe9aca9e6ff3'\n",
      " 'http://data.boehringer.com/uuid/TranscriptOmicsSampleSet/4e991235-d2ce-4313-bdef-f830c94d3a86'\n",
      " 'http://data.boehringer.com/uuid/TranscriptOmicsSampleSet/08b24327-0226-4487-960a-918ef73e8521'\n",
      " 'http://data.boehringer.com/uuid/TranscriptOmicsSampleSet/c2aa1cc9-44af-40ca-b9d1-09928706da45'\n",
      " 'http://data.boehringer.com/uuid/TranscriptOmicsSampleSet/00a14164-e90f-42d7-aeca-f980c8b17ecb'\n",
      " 'http://data.boehringer.com/uuid/TranscriptOmicsSampleSet/b00528aa-7161-439d-97c4-b75898eb095b'\n",
      " 'http://data.boehringer.com/uuid/TranscriptOmicsSampleSet/499f195e-918c-4e39-861b-c536cfde7af2'\n",
      " 'http://data.boehringer.com/uuid/TranscriptOmicsSampleSet/fb690cca-a49d-4ad7-87c6-e6dec1e85bef'\n",
      " 'http://data.boehringer.com/uuid/TranscriptOmicsSampleSet/57575d69-1589-426d-8367-bfdcb5e457a6'\n",
      " 'http://data.boehringer.com/uuid/TranscriptOmicsSampleSet/da107d01-29d4-449d-aec5-41b444c05708']\n",
      "http://data.boehringer.com/ontology/omics/hasSourceCellType ['EdgeNotPresent'\n",
      " 'http://vocabulary.boehringer.com/CellType/64bd586f-8f04-4190-bab0-28a92c5122f1']\n",
      "http://data.boehringer.com/ontology/omics/hasHistologyType ['EdgeNotPresent'\n",
      " 'http://vocabulary.boehringer.com/OMICS/9b0b88f4-a879-48a6-ae49-5734f0a4222c']\n",
      "http://data.boehringer.com/ontology/omics/isSequencedBySite ['EdgeNotPresent'\n",
      " 'http://vocabulary.boehringer.com/OMICS/fe0dd6cf-162c-4c04-bf13-9986eb8f1f85']\n",
      "http://data.boehringer.com/ontology/omics/isExcluded [False]\n",
      "http://data.boehringer.com/ontology/omics/hasSpecies ['EdgeNotPresent'\n",
      " 'http://vocabulary.boehringer.com/Strain/b7f294dd-cdbc-42fe-954f-9db0e1466e4e']\n",
      "http://data.boehringer.com/ontology/omics/isPartOfStudy ['EdgeNotPresent'\n",
      " 'http://data.boehringer.com/uuid/TranscriptOmicsStudy/EX51']\n",
      "http://data.boehringer.com/ontology/omics/hasStrandedness ['EdgeNotPresent'\n",
      " 'http://vocabulary.boehringer.com/OMICS/6f2a8408-2031-471a-bd42-5d1f4564cf6a']\n",
      "http://data.boehringer.com/ontology/omics/hasSampleType ['EdgeNotPresent'\n",
      " 'http://vocabulary.boehringer.com/OMICS/14884101-7675-49ca-b50c-32da5aeb965a']\n",
      "http://data.boehringer.com/ontology/omics/hasCellType ['http://vocabulary.boehringer.com/CellType/64bd586f-8f04-4190-bab0-28a92c5122f1'\n",
      " 'EdgeNotPresent'\n",
      " 'http://vocabulary.boehringer.com/CellType/d524dbbe-c918-457c-8447-fb137b4acebb'\n",
      " 'http://vocabulary.boehringer.com/CellType/7fac0c84-baca-4ff8-9c6b-326c06d25757'\n",
      " 'http://vocabulary.boehringer.com/CellType/f2908a66-6f13-465b-9caa-ea69f53c2a37'\n",
      " 'http://vocabulary.boehringer.com/CellType/db30f3b2-82f0-4ad6-b0f5-c5fbf4e93191']\n",
      "http://data.boehringer.com/ontology/omics/sourceId [\"['EdgeNotPresent']\" \"['EX51_902', 'Mono_intermediate_S107']\"\n",
      " \"['EX51_308', 'CD141_P4_S44']\" ... \"['AXLSIGLEC6_S24', 'EX51_1270']\"\n",
      " \"['EX51_476', 'CD1C_P3_S5']\" \"['EX51_865', 'Mono_classical_S74']\"]\n",
      "http://data.boehringer.com/ontology/omics/id ['0444023a-6a21-4afe-bb75-78941dfb5766'\n",
      " '946b8e10-ad3c-492d-9c50-aff294cd122b'\n",
      " 'e441da98-c03e-4688-8999-d2b048b3d4ad' ...\n",
      " '9d9a89be-08f3-40a4-8545-f4f2f5598334'\n",
      " '5dd45b95-d5f6-4301-976c-0a4ad50e1ca8'\n",
      " 'c4e77f24-c57e-4b52-a8e4-ea4dcf16f61c']\n",
      "http://data.boehringer.com/ontology/omics/hasKeyValue [\"['EdgeNotPresent']\"\n",
      " \"['http://data.boehringer.com/uuid/KeyValueHolder/e29a0ae3cf8e80d2b17105c29eb175a8', 'http://data.boehringer.com/uuid/KeyValueHolder/ce495590157ea6a8a523a2cb877ebac0', 'http://data.boehringer.com/uuid/KeyValueHolder/a40232060dfa08cf3aee506ef45f283a']\"\n",
      " \"['http://data.boehringer.com/uuid/KeyValueHolder/ee5bcd3132f73d9b8a3386a69b8fe3ba', 'http://data.boehringer.com/uuid/KeyValueHolder/8660eb062efb7d587f75a28b61e5dc9f', 'http://data.boehringer.com/uuid/KeyValueHolder/dab859a8b7583f1546334ba4e0c3af75', 'http://data.boehringer.com/uuid/KeyValueHolder/2d78ad559715901b70efa359ebebe830']\"\n",
      " ...\n",
      " \"['http://data.boehringer.com/uuid/KeyValueHolder/4f58245760fdb32cf7f34e2a20b63e8e', 'http://data.boehringer.com/uuid/KeyValueHolder/1d80555a6ac07958e916b2e4ea2705a0', 'http://data.boehringer.com/uuid/KeyValueHolder/0210c969f6a13320c68fbbd858c95175', 'http://data.boehringer.com/uuid/KeyValueHolder/b24d6d488e1525bc42355075da13ee3b']\"\n",
      " \"['http://data.boehringer.com/uuid/KeyValueHolder/f0810411ec3df873fb6eebaa88b02043', 'http://data.boehringer.com/uuid/KeyValueHolder/00992f2fac168d115cfb3f369e21bdf2', 'http://data.boehringer.com/uuid/KeyValueHolder/5a26a46c03905bfb658f04444ca8e87f', 'http://data.boehringer.com/uuid/KeyValueHolder/2d5b7eb23476dec9e1639f34f2947e39']\"\n",
      " \"['http://data.boehringer.com/uuid/KeyValueHolder/8ecbb3c3437f8633eaac99521b6c7f91', 'http://data.boehringer.com/uuid/KeyValueHolder/2918783b0e6d63364e4f23ebcda40e09', 'http://data.boehringer.com/uuid/KeyValueHolder/f13e1f77d6dba1290ecba130a414df91']\"]\n",
      "http://data.boehringer.com/ontology/omics/sampleExcluded ['EdgeNotPresent' 'false']\n",
      "http://www.w3.org/1999/02/22-rdf-syntax-ns#type [\"['http://data.boehringer.com/ontology/omics/PrimaryCellSpecimen']\"\n",
      " \"['http://data.boehringer.com/ontology/omics/Donor']\"\n",
      " \"['http://data.boehringer.com/ontology/omics/TranscriptOmicsSample']\"\n",
      " \"['http://data.boehringer.com/ontology/omics/Study', 'http://data.boehringer.com/ontology/omics/TranscriptOmicsStudy']\"]\n",
      "http://data.boehringer.com/ontology/omics/hasPreparationProtocol ['EdgeNotPresent'\n",
      " 'http://vocabulary.boehringer.com/OMICS/12eec321-ca7e-4866-982f-0a0677fed966']\n",
      "http://data.boehringer.com/ontology/omics/hasSequencingProtocol ['EdgeNotPresent'\n",
      " 'http://vocabulary.boehringer.com/OMICS/4ddc9bf7-73bd-45de-aa78-ad3ccfc24420']\n",
      "http://data.boehringer.com/ontology/omics/isDerivedFromSpecimen ['EdgeNotPresent'\n",
      " 'http://data.boehringer.com/uuid/PrimaryCellSpecimen/0de429a9-39ac-4862-988a-d5243098996c'\n",
      " 'http://data.boehringer.com/uuid/PrimaryCellSpecimen/b8d8e0a2-00c7-46db-a006-8d8c72b6c983'\n",
      " ...\n",
      " 'http://data.boehringer.com/uuid/PrimaryCellSpecimen/03a828aa-405a-45ea-bfcd-acc6f921fa30'\n",
      " 'http://data.boehringer.com/uuid/PrimaryCellSpecimen/e05632ff-8d05-42aa-a3af-2de35db98ba7'\n",
      " 'http://data.boehringer.com/uuid/PrimaryCellSpecimen/65662a8c-c175-4ff8-96a8-18f115672433']\n",
      "http://data.boehringer.com/ontology/omics/hasDisease ['http://vocabulary.boehringer.com/TherapeuticArea/8cb5cb83-020f-4e0e-b19c-9008cf92f40e'\n",
      " 'EdgeNotPresent']\n",
      "http://data.boehringer.com/ontology/omics/labSourceId ['EdgeNotPresent' 'Mono_intermediate_S107' 'CD141_P4_S44' ...\n",
      " 'AXLSIGLEC6_S24' 'CD1C_P3_S5' 'Mono_classical_S74']\n",
      "http://data.boehringer.com/ontology/omics/isProvidedBy ['http://vocabulary.boehringer.com/OMICS/fe0dd6cf-162c-4c04-bf13-9986eb8f1f85'\n",
      " 'EdgeNotPresent']\n",
      "http://data.boehringer.com/ontology/omics/isMeasuredOnPlatform ['EdgeNotPresent'\n",
      " 'http://vocabulary.boehringer.com/OMICS/67ecb0ea-0b44-4357-9af7-131d787d87bc']\n",
      "http://data.boehringer.com/ontology/omics/isDerivedFromDonor ['http://data.boehringer.com/uuid/Donor/fc336226-2fb8-4d50-a015-877115c29cbd'\n",
      " 'EdgeNotPresent'\n",
      " 'http://data.boehringer.com/uuid/Donor/75478ad4-f7dc-4428-a74a-1a77d9333169'\n",
      " ...\n",
      " 'http://data.boehringer.com/uuid/Donor/3c053db0-2460-459a-a037-2b7f53153193'\n",
      " 'http://data.boehringer.com/uuid/Donor/def9e5b9-0307-4eb1-8bf8-32d18b595279'\n",
      " 'http://data.boehringer.com/uuid/Donor/884bdf8b-f41d-4312-a626-0b22d3761327']\n",
      "http://data.boehringer.com/ontology/omics/isProcessedByOrganisation ['EdgeNotPresent'\n",
      " 'http://vocabulary.boehringer.com/OMICS/fe0dd6cf-162c-4c04-bf13-9986eb8f1f85']\n",
      "http://data.boehringer.com/ontology/omics/sampleId ['EdgeNotPresent' 'EX51_902' 'EX51_308' ... 'EX51_1270' 'EX51_476'\n",
      " 'EX51_865']\n",
      "http://data.boehringer.com/ontology/omics/sampleName ['EdgeNotPresent' 'EX51_902_Mono_intermediate' 'EX51_308_CD141' ...\n",
      " 'EX51_1270_AXLSIGLEC6' 'EX51_476_CD1C' 'EX51_865_Mono_classical']\n",
      "http://data.boehringer.com/ontology/omics/studyTitle ['EdgeNotPresent'\n",
      " 'Human Blood Monoyctes Dendritic Cells Atlas GSE94820 (GEO)']\n",
      "http://data.boehringer.com/ontology/omics/isSequencedForTherapeuticArea ['EdgeNotPresent'\n",
      " 'http://vocabulary.boehringer.com/TherapeuticArea/3d54ad47-8840-40f2-a2c4-43eaefa0834a']\n",
      "http://data.boehringer.com/ontology/omics/keywords [\"['EdgeNotPresent']\"\n",
      " \"['Plasmacytoid Dendritic Cell', 'Immune Modulation']\"]\n",
      "http://data.boehringer.com/ontology/omics/externalStudyId ['EdgeNotPresent' 'GSE94820']\n",
      "http://data.boehringer.com/ontology/omics/sourceStudyId ['EdgeNotPresent' 'EX51']\n",
      "http://data.boehringer.com/ontology/omics/technicalStudyTitle ['EdgeNotPresent' 'EX51_Human_Blood_Monocytes_DendriticCells_Atlas']\n",
      "http://data.boehringer.com/ontology/omics/isPartOfProject ['EdgeNotPresent' 'http://data.boehringer.com/lov/project/2913']\n",
      "http://data.boehringer.com/ontology/omics/studyId ['EdgeNotPresent' 'EX51']\n",
      "http://data.boehringer.com/ontology/omics/hasStudyType ['EdgeNotPresent'\n",
      " 'http://vocabulary.boehringer.com/OMICS/07e0e9f3-9564-4f58-b36b-a3affa14a356']\n",
      "http://data.boehringer.com/ontology/omics/sequencingDate ['EdgeNotPresent' '2017-01-01T00:00:00+00:00']\n",
      "http://data.boehringer.com/ontology/omics/isSequencedForIndication ['EdgeNotPresent'\n",
      " 'http://vocabulary.boehringer.com/TherapeuticArea/e5534260-f54c-4607-afb5-37f447a1352c']\n",
      "http://data.boehringer.com/ontology/omics/hasResponsibleSite ['EdgeNotPresent'\n",
      " 'http://vocabulary.boehringer.com/OMICS/fe0dd6cf-162c-4c04-bf13-9986eb8f1f85']\n",
      "http://data.boehringer.com/ontology/omics/hasContactScientist ['EdgeNotPresent'\n",
      " 'http://vocabulary.boehringer.com/OMICS/d3e43a12-403d-4299-95a5-7e3dbe50918a']\n",
      "http://data.boehringer.com/ontology/omics/studyDescription ['EdgeNotPresent'\n",
      " 'Healthy donors were recruited from the Boston-based PhenoGenetic project a resource of healthy subjects that are recontactable by genotype  and the Newcastle community. Individuals were excluded if they had a history of cancer allergies inflammatory disease autoimmune disease chronic metabolic disorders or infectious disorders. All healthy donors were nonsmokers had a normal BMI and normal blood pressure and were between 25 and 40 years of age. PBMCs were isolated from fresh blood within 2 hours of collection using Ficoll-Paque density gradient centrifugation as described . Single-cell suspensions were stained per manufacturer ecommendations with different panels of antibodies designed to enrich for certain population for single-cell sorting and single-cell RNA sequencing (scRNA-seq). Aim: Immune cell atlas a tool box for CI&IM target identification /validation ( e.g. IRRs ) Next steps: Expressuion analysis of IRRs .']\n",
      "http://data.boehringer.com/ontology/omics/hasContactCompBio ['EdgeNotPresent'\n",
      " 'http://vocabulary.boehringer.com/OMICS/bb2440ad-e80e-424d-ad29-fc83a277611b']\n"
     ]
    }
   ],
   "source": [
    "# for each column print all unique values\n",
    "for column in study_df.columns:\n",
    "    print(column, study_df[column].unique())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Unit Tests for Violation Counting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing Graph Once: 100%|██████████| 6/6 [00:00<00:00, 5826.77it/s]\n",
      "Going Over Violation Instances: 100%|██████████| 3/3 [00:00<?, ?it/s]\n",
      "Processing Graph Once: 100%|██████████| 8/8 [00:00<?, ?it/s]\n",
      "Going Over Violation Instances: 100%|██████████| 4/4 [00:00<?, ?it/s]\n",
      ".\n",
      "----------------------------------------------------------------------\n",
      "Ran 2 tests in 0.007s\n",
      "\n",
      "OK\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<unittest.runner.TextTestResult run=2 errors=0 failures=0>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import unittest\n",
    "from rdflib import Graph, Literal, Namespace, BNode, URIRef\n",
    "\n",
    "SH = Namespace(\"http://www.w3.org/ns/shacl#\")\n",
    "\n",
    "class TestCountViolations(unittest.TestCase):\n",
    "    def setUp(self):\n",
    "        self.violations = Graph()\n",
    "\n",
    "    def test_count_violations_node_with_multiple_and_node_with_double(self):\n",
    "        self.focusNode1 = BNode()\n",
    "        self.focusNode2 = BNode()\n",
    "        self.sourceShape1 = Literal('shape1')\n",
    "        self.sourceShape2 = Literal('shape2')\n",
    "        self.violation1 = BNode()\n",
    "        self.violation2 = BNode()\n",
    "        self.violation3 = BNode()\n",
    "        self.violation4 = BNode()\n",
    "\n",
    "        # Add triples to the violations graph\n",
    "        self.violations.add((self.violation1, SH.focusNode, self.focusNode1))\n",
    "        self.violations.add((self.violation1, SH.sourceShape, self.sourceShape1))\n",
    "\n",
    "        self.violations.add((self.violation2, SH.focusNode, self.focusNode2))\n",
    "        self.violations.add((self.violation2, SH.sourceShape, self.sourceShape2))\n",
    "\n",
    "        self.violations.add((self.violation3, SH.focusNode, self.focusNode1))\n",
    "        self.violations.add((self.violation3, SH.sourceShape, self.sourceShape1))\n",
    "\n",
    "        self.violations.add((self.violation4, SH.focusNode, self.focusNode2))\n",
    "        self.violations.add((self.violation4, SH.sourceShape, self.sourceShape1))\n",
    "        violation_counts, violation_list = count_violations(self.violations)\n",
    "        expected_counts = {\n",
    "            str(self.focusNode1): {str(self.sourceShape1): 2},\n",
    "            str(self.focusNode2): {str(self.sourceShape2): 1, str(self.sourceShape1): 1}\n",
    "        }\n",
    "        expected_list = [str(self.sourceShape1), str(self.sourceShape2)] \n",
    "        \n",
    "        self.assertEqual(violation_counts, expected_counts)\n",
    "        self.assertEqual(set(violation_list), set(expected_list))  # Use set to ignore order\n",
    "\n",
    "    def test_count_violations_basic(self):\n",
    "        self.focusNode1 = BNode()\n",
    "        self.focusNode2 = BNode()\n",
    "        self.focusNode3 = BNode()\n",
    "        self.sourceShape1 = Literal('shape1')\n",
    "        self.sourceShape2 = Literal('shape2')\n",
    "        self.sourceShape3 = Literal('shape3')\n",
    "        self.violation1 = BNode()\n",
    "        self.violation2 = BNode()\n",
    "        self.violation3 = BNode()\n",
    "\n",
    "        # Add triples to the violations graph\n",
    "        self.violations.add((self.violation1, SH.focusNode, self.focusNode1))\n",
    "        self.violations.add((self.violation1, SH.sourceShape, self.sourceShape1))\n",
    "\n",
    "        self.violations.add((self.violation2, SH.focusNode, self.focusNode2))\n",
    "        self.violations.add((self.violation2, SH.sourceShape, self.sourceShape2))\n",
    "\n",
    "        self.violations.add((self.violation3, SH.focusNode, self.focusNode3))\n",
    "        self.violations.add((self.violation3, SH.sourceShape, self.sourceShape3))\n",
    "\n",
    "        violation_counts, violation_list = count_violations(self.violations)\n",
    "        expected_counts = {\n",
    "            str(self.focusNode1): {str(self.sourceShape1): 1},\n",
    "            str(self.focusNode2): {str(self.sourceShape2): 1},\n",
    "            str(self.focusNode3): {str(self.sourceShape3): 1}\n",
    "        }\n",
    "        expected_list = [str(self.sourceShape1), str(self.sourceShape2), str(self.sourceShape3)] \n",
    "        \n",
    "        self.assertEqual(violation_counts, expected_counts)\n",
    "        self.assertEqual(set(violation_list), set(expected_list))  # Use set to ignore order\n",
    "\n",
    "suite = unittest.TestLoader().loadTestsFromTestCase(TestCountViolations)\n",
    "unittest.TextTestRunner().run(suite)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Abbreviation using QNames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Can't split 'EX51_902_Mono_intermediate'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[13], line 61\u001b[0m, in \u001b[0;36mabbreviate_cell_value\u001b[1;34m(nsm, cell_value)\u001b[0m\n\u001b[0;32m     59\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m     60\u001b[0m     \u001b[38;5;66;03m# Attempt to evaluate the cell_value as a list\u001b[39;00m\n\u001b[1;32m---> 61\u001b[0m     uris \u001b[38;5;241m=\u001b[39m \u001b[43mast\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mliteral_eval\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcell_value\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     62\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(uris, \u001b[38;5;28mlist\u001b[39m):\n\u001b[0;32m     63\u001b[0m         \u001b[38;5;66;03m# Process each URI in the list\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python310\\lib\\ast.py:108\u001b[0m, in \u001b[0;36mliteral_eval\u001b[1;34m(node_or_string)\u001b[0m\n\u001b[0;32m    107\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _convert_signed_num(node)\n\u001b[1;32m--> 108\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_convert\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnode_or_string\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python310\\lib\\ast.py:107\u001b[0m, in \u001b[0;36mliteral_eval.<locals>._convert\u001b[1;34m(node)\u001b[0m\n\u001b[0;32m    106\u001b[0m             \u001b[38;5;28;01mreturn\u001b[39;00m left \u001b[38;5;241m-\u001b[39m right\n\u001b[1;32m--> 107\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_convert_signed_num\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnode\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python310\\lib\\ast.py:81\u001b[0m, in \u001b[0;36mliteral_eval.<locals>._convert_signed_num\u001b[1;34m(node)\u001b[0m\n\u001b[0;32m     80\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;241m-\u001b[39m operand\n\u001b[1;32m---> 81\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_convert_num\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnode\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python310\\lib\\ast.py:72\u001b[0m, in \u001b[0;36mliteral_eval.<locals>._convert_num\u001b[1;34m(node)\u001b[0m\n\u001b[0;32m     71\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(node, Constant) \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mtype\u001b[39m(node\u001b[38;5;241m.\u001b[39mvalue) \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m (\u001b[38;5;28mint\u001b[39m, \u001b[38;5;28mfloat\u001b[39m, \u001b[38;5;28mcomplex\u001b[39m):\n\u001b[1;32m---> 72\u001b[0m     \u001b[43m_raise_malformed_node\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnode\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     73\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m node\u001b[38;5;241m.\u001b[39mvalue\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python310\\lib\\ast.py:69\u001b[0m, in \u001b[0;36mliteral_eval.<locals>._raise_malformed_node\u001b[1;34m(node)\u001b[0m\n\u001b[0;32m     68\u001b[0m     msg \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m on line \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mlno\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m---> 69\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(msg \u001b[38;5;241m+\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mnode\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[1;31mValueError\u001b[0m: malformed node or string on line 1: <ast.Name object at 0x000002669462F5B0>",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[13], line 148\u001b[0m\n\u001b[0;32m    145\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(VIOLATION_LIST_FILE, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mw\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m f:\n\u001b[0;32m    146\u001b[0m         json\u001b[38;5;241m.\u001b[39mdump(violation_list, f)\n\u001b[1;32m--> 148\u001b[0m \u001b[43mabbreviate_using_namespaces\u001b[49m\u001b[43m(\u001b[49m\u001b[43mstudy_g\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mviolations_g\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[1;32mIn[13], line 117\u001b[0m, in \u001b[0;36mabbreviate_using_namespaces\u001b[1;34m(study_graph, violations_graph)\u001b[0m\n\u001b[0;32m    115\u001b[0m         \u001b[38;5;66;03m# Check if the cell contains a string that needs abbreviation\u001b[39;00m\n\u001b[0;32m    116\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(cell_value, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m--> 117\u001b[0m             study_df\u001b[38;5;241m.\u001b[39mloc[idx, col] \u001b[38;5;241m=\u001b[39m \u001b[43mabbreviate_cell_value\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnsm_combined\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcell_value\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    119\u001b[0m \u001b[38;5;66;03m# Change indices and cell values\u001b[39;00m\n\u001b[0;32m    120\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m col \u001b[38;5;129;01min\u001b[39;00m study_df\u001b[38;5;241m.\u001b[39mcolumns:\n",
      "Cell \u001b[1;32mIn[13], line 71\u001b[0m, in \u001b[0;36mabbreviate_cell_value\u001b[1;34m(nsm, cell_value)\u001b[0m\n\u001b[0;32m     68\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m get_qname(nsm, cell_value)\n\u001b[0;32m     69\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:  \u001b[38;5;66;03m# Catching any exception explicitly\u001b[39;00m\n\u001b[0;32m     70\u001b[0m     \u001b[38;5;66;03m# If evaluation fails or it's a single URI not in list form, process as a single URI\u001b[39;00m\n\u001b[1;32m---> 71\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mget_qname\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnsm\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcell_value\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[1;32mIn[13], line 34\u001b[0m, in \u001b[0;36mget_qname\u001b[1;34m(nsm, uri)\u001b[0m\n\u001b[0;32m     22\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_qname\u001b[39m(nsm, uri):\n\u001b[0;32m     23\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m     24\u001b[0m \u001b[38;5;124;03m    Returns a QName for a given URI if its namespace is in the provided NamespaceManager.\u001b[39;00m\n\u001b[0;32m     25\u001b[0m \u001b[38;5;124;03m    If the namespace isn't found in the NamespaceManager, it returns the original URI.\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m     32\u001b[0m \u001b[38;5;124;03m    str: a QName if the namespace of the URI is in the NamespaceManager, the original URI otherwise\u001b[39;00m\n\u001b[0;32m     33\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m---> 34\u001b[0m     namespace, name \u001b[38;5;241m=\u001b[39m \u001b[43msplit_uri\u001b[49m\u001b[43m(\u001b[49m\u001b[43muri\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     35\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m namespace_in_nsm(nsm, namespace):\n\u001b[0;32m     36\u001b[0m         \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "File \u001b[1;32mc:\\Work\\ICG\\BI\\bikg_app\\.venv\\lib\\site-packages\\rdflib\\namespace\\__init__.py:809\u001b[0m, in \u001b[0;36msplit_uri\u001b[1;34m(uri, split_start)\u001b[0m\n\u001b[0;32m    807\u001b[0m                 \u001b[38;5;28;01mreturn\u001b[39;00m (ns, ln)\n\u001b[0;32m    808\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n\u001b[1;32m--> 809\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCan\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mt split \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(uri))\n",
      "\u001b[1;31mValueError\u001b[0m: Can't split 'EX51_902_Mono_intermediate'"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import ast\n",
    "from rdflib import Graph\n",
    "from rdflib.namespace import split_uri\n",
    "\n",
    "def namespace_in_nsm(nsm, namespace):\n",
    "    \"\"\"\n",
    "    Checks if a given namespace is in the provided NamespaceManager.\n",
    "\n",
    "    Parameters:\n",
    "    nsm (NamespaceManager): the NamespaceManager object to check\n",
    "    namespace (str): the namespace to look for in the NamespaceManager\n",
    "\n",
    "    Returns:\n",
    "    bool: True if the namespace is in the NamespaceManager, False otherwise\n",
    "    \"\"\"\n",
    "    for prefix, ns in nsm.namespaces():\n",
    "        if str(ns) == str(namespace):\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "def get_qname(nsm, uri):\n",
    "    \"\"\"\n",
    "    Returns a QName for a given URI if its namespace is in the provided NamespaceManager.\n",
    "    If the namespace isn't found in the NamespaceManager, it returns the original URI.\n",
    "\n",
    "    Parameters:\n",
    "    nsm (NamespaceManager): the NamespaceManager object to use\n",
    "    uri (str): the URI to transform into a QName\n",
    "\n",
    "    Returns:\n",
    "    str: a QName if the namespace of the URI is in the NamespaceManager, the original URI otherwise\n",
    "    \"\"\"\n",
    "    namespace, name = split_uri(uri)\n",
    "    if namespace_in_nsm(nsm, namespace):\n",
    "        try:\n",
    "            prefix, _, _ = nsm.compute_qname(uri)\n",
    "            return f\"{prefix}:{name}\"\n",
    "        except Exception as e:\n",
    "            return uri\n",
    "    else:\n",
    "        return uri\n",
    "    \n",
    "def abbreviate_cell_value(nsm, cell_value):\n",
    "    \"\"\"\n",
    "    Attempts to abbreviate URIs in a cell that may contain a string representation\n",
    "    of a list of URIs or a single URI. If abbreviation is not possible, the original\n",
    "    cell value is returned.\n",
    "    \n",
    "    Parameters:\n",
    "    nsm (NamespaceManager): The NamespaceManager to use for abbreviation.\n",
    "    cell_value (str): The cell value to abbreviate.\n",
    "    \n",
    "    Returns:\n",
    "    str: The abbreviated cell value if possible; otherwise, the original cell value.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # Attempt to evaluate the cell_value as a list or single URI\n",
    "        uris = ast.literal_eval(cell_value)\n",
    "        # Handle list of URIs\n",
    "        if isinstance(uris, list):\n",
    "            abbreviated_uris = [get_qname(nsm, uri) for uri in uris]\n",
    "            return str(abbreviated_uris)\n",
    "        else:\n",
    "            # Handle a single URI\n",
    "            return get_qname(nsm, uris)\n",
    "    except:\n",
    "        # If the above fails (either due to eval or not being a URI), return original\n",
    "        try:\n",
    "            # Attempt to directly abbreviate assuming it's a single URI\n",
    "            return get_qname(nsm, cell_value)\n",
    "        except:\n",
    "            # If all else fails, return the cell value as is\n",
    "            return cell_value\n",
    "\n",
    "\n",
    "def abbreviate_using_namespaces(study_graph: Graph, violations_graph: Graph):\n",
    "    \"\"\"\n",
    "    Replaces URIs in a DataFrame and a list of violations with QNames if their namespaces\n",
    "    are found in two provided RDF graphs. Saves the updated DataFrame and the list back to disk.\n",
    "\n",
    "    Parameters:\n",
    "    study_graph (Graph): the first RDF graph to use for namespaces\n",
    "    violations_graph (Graph): the second RDF graph to use for namespaces\n",
    "    \"\"\"\n",
    "\n",
    "    # load violation_list\n",
    "    with open(VIOLATION_LIST_FILE, 'r') as f:\n",
    "        violation_list = json.load(f)\n",
    "\n",
    "    # read study_df tabularized data\n",
    "    study_df = pd.read_csv(STUDY_CSV_FILE, index_col=0)\n",
    "\n",
    "    # Create a new graph to combine namespaces\n",
    "    combined_graph = Graph()\n",
    "\n",
    "    # Get namespaces from both graphs and bind them to the combined graph\n",
    "    for prefix, ns_uri in study_graph.namespace_manager.namespaces():\n",
    "        combined_graph.namespace_manager.bind(prefix, ns_uri)\n",
    "\n",
    "    for prefix, ns_uri in violations_graph.namespace_manager.namespaces():\n",
    "        combined_graph.namespace_manager.bind(prefix, ns_uri)\n",
    "\n",
    "    # Now use the NamespaceManager of the combined graph\n",
    "    nsm_combined = NamespaceManager(combined_graph)\n",
    "\n",
    "    # Change column names\n",
    "    for col in study_df.columns:\n",
    "        try:\n",
    "            study_df.rename(columns={col: get_qname(nsm_combined, col)}, inplace=True)\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "     # Change cell values to handle arrays in string form\n",
    "    for col in study_df.columns:\n",
    "        for idx in study_df.index:\n",
    "            cell_value = study_df.loc[idx, col]\n",
    "            # Check if the cell contains a string that needs abbreviation\n",
    "            if isinstance(cell_value, str):\n",
    "                study_df.loc[idx, col] = abbreviate_cell_value(nsm_combined, cell_value)\n",
    "\n",
    "    # Change indices and cell values\n",
    "    for col in study_df.columns:\n",
    "        for idx in study_df.index:\n",
    "            try:\n",
    "                study_df.loc[idx, col] = get_qname(nsm_combined, study_df.loc[idx, col])\n",
    "            except:\n",
    "                pass\n",
    "\n",
    "    # change index\n",
    "    for idx in study_df.index:\n",
    "        try:\n",
    "            study_df.rename(index={idx: get_qname(nsm_combined, str(idx))}, inplace=True)\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "    # change violation_list\n",
    "    for i in range(len(violation_list)):\n",
    "        try:\n",
    "            violation_list[i] = get_qname(nsm_combined, violation_list[i])\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "    # store tabular graph as csv\n",
    "    study_df.to_csv(STUDY_CSV_FILE)    \n",
    "\n",
    "    # store violation_list as json\n",
    "    with open(VIOLATION_LIST_FILE, 'w') as f:\n",
    "        json.dump(violation_list, f)\n",
    "\n",
    "abbreviate_using_namespaces(study_g, violations_g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(VIOLATION_LIST_FILE, 'r') as f:\n",
    "    violation_list = json.load(f)\n",
    "\n",
    "print('unique violations ('+str(len(violation_list))+'):\\n')\n",
    "print('\\n'.join([str(v) for v in violation_list]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "study_df = pd.read_csv(STUDY_CSV_FILE, index_col=0)\n",
    "study_df.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compute Violation Embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "import umap\n",
    "import pandas as pd\n",
    "import json\n",
    "import numpy as np\n",
    "\n",
    "from scipy.sparse import coo_matrix\n",
    "\n",
    "def create_embedding():\n",
    "    # read study.csv\n",
    "    study_df = pd.read_csv(STUDY_CSV_FILE, index_col=0)\n",
    "    \n",
    "    # read violation_list.json\n",
    "    with open(VIOLATION_LIST_FILE, 'r') as f:\n",
    "        violation_columns = json.load(f)\n",
    "    \n",
    "    view_df = study_df[violation_columns]  # view of study dataframe with only violation columns\n",
    "\n",
    "    print(\"Creating UMAP embedding... (this may take some time)\")\n",
    "    reducer = umap.UMAP(n_neighbors=int(np.sqrt(len(view_df))), min_dist=0.1, n_components=2, random_state=0, verbose=True)\n",
    "    result = reducer.fit_transform(view_df)  # 2D projection\n",
    "\n",
    "    # Check if the result is a tuple and unpack accordingly\n",
    "    if isinstance(result, tuple):\n",
    "        embedding, *_ = result\n",
    "    else:\n",
    "        embedding = result\n",
    "\n",
    "    # Convert to dense numpy array if it's a sparse matrix\n",
    "    if isinstance(embedding, coo_matrix):\n",
    "        embedding = embedding.toarray()\n",
    "\n",
    "    # Assuming that the embedding is now a NumPy array, based on the docstring\n",
    "    study_df[\"x\"] = embedding[:, 0]\n",
    "    study_df[\"y\"] = embedding[:, 1]\n",
    "\n",
    "    study_df.to_csv(STUDY_CSV_FILE)\n",
    "    print(\"UMAP embedding created and saved to\", STUDY_CSV_FILE)\n",
    "\n",
    "\n",
    "def plot_embedding():\n",
    "    study_df = pd.read_csv(STUDY_CSV_FILE, index_col=0)\n",
    "\n",
    "    plt.scatter(study_df[\"x\"], study_df[\"y\"])\n",
    "    plt.show()\n",
    "\n",
    "create_embedding()\n",
    "plot_embedding()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Replace URIs with Label Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "from rdflib import URIRef\n",
    "\n",
    "# list of label predicates in ascending order of priority\n",
    "label_predicates = [\n",
    "    # URIRef('http://data.boehringer.com/ontology/omics/hasValue'),\n",
    "    # URIRef('http://data.boehringer.com/ontology/omics/name'),\n",
    "    URIRef('http://www.w3.org/2004/02/skos/core#prefLabel')\n",
    "]\n",
    "\n",
    "# read tabularized graph .csv as a dataframe, and violations list .json as a list\n",
    "study_df = pd.read_csv(STUDY_CSV_FILE, index_col=0)\n",
    "\n",
    "with open(VIOLATION_LIST_FILE, 'r') as f:\n",
    "    violation_columns = json.load(f)\n",
    "\n",
    "# Create a new graph to combine namespaces\n",
    "combined_graph = Graph()\n",
    "\n",
    "# Get namespaces from both graphs and bind them to the combined graph\n",
    "for prefix, ns_uri in study_g.namespace_manager.namespaces():\n",
    "    combined_graph.namespace_manager.bind(prefix, ns_uri)\n",
    "\n",
    "for prefix, ns_uri in violations_g.namespace_manager.namespaces():\n",
    "    combined_graph.namespace_manager.bind(prefix, ns_uri)\n",
    "\n",
    "# Now use the NamespaceManager of the combined graph\n",
    "nsm_combined = NamespaceManager(combined_graph)\n",
    "\n",
    "# create a dictionary of {s: o} pairs for translating source shapes to their labels\n",
    "label_dict = {}\n",
    "for label_predicate in label_predicates:\n",
    "    temp_dict = {str(s): str(o) for s, p, o in study_g.triples((None, label_predicate, None))}\n",
    "    # replace the temp_dict keys with their corresponding QNames\n",
    "    temp_dict = {get_qname(nsm_combined, k): v for k, v in temp_dict.items()}\n",
    "    # Update label_dict with temp_dict, overwriting existing keys\n",
    "    label_dict.update(temp_dict)\n",
    "\n",
    "# replace all column names, indices, and cell values in the dataframe with their corresponding labels\n",
    "study_df.columns = [label_dict.get(col, col) for col in study_df.columns]\n",
    "study_df.index = pd.Index([label_dict.get(idx, idx) for idx in study_df.index])\n",
    "for col in study_df.columns:\n",
    "    study_df[col] = study_df[col].apply(lambda x: label_dict.get(x, x))\n",
    "\n",
    "# replace all dictionary keys and values with their corresponding labels\n",
    "violation_columns = [label_dict.get(x, x) for x in violation_columns]\n",
    "\n",
    "# store dataframe back to .csv and violations list back to .json\n",
    "study_df.to_csv(STUDY_CSV_FILE)\n",
    "with open(VIOLATION_LIST_FILE, 'w') as f:\n",
    "    json.dump(violation_columns, f)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create Union of Ontology and Violation Report Exemplars"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rdflib import URIRef, RDFS\n",
    "\n",
    "def replace_uris_with_labels(data, label_dict, data_type=\"graph\"):\n",
    "    if data_type == \"graph\":\n",
    "        # Iterate through each triple\n",
    "        for s, p, o in list(data.triples((None, None, None))):\n",
    "            new_s = URIRef(label_dict.get(str(s), s))\n",
    "            new_o = URIRef(label_dict.get(str(o), o))\n",
    "            \n",
    "            if new_s != s or new_o != o:\n",
    "                data.remove((s, p, o))\n",
    "                data.add((new_s, p, new_o))\n",
    "                \n",
    "    elif data_type == \"dict\":\n",
    "        new_data = {}\n",
    "        for key, values in data.items():\n",
    "            new_key = label_dict.get(key, key)\n",
    "            \n",
    "            # If the values are in set form\n",
    "            if isinstance(values, set):\n",
    "                new_values = {label_dict.get(val, val) for val in values}\n",
    "                \n",
    "            # If the values are in dictionary form with concatenated string keys\n",
    "            elif isinstance(values, dict):\n",
    "                new_values = {}\n",
    "                for po_str, count in values.items():\n",
    "                    if '__' in po_str:\n",
    "                        p, o = po_str.split('__')\n",
    "                        new_p = label_dict.get(p, p)\n",
    "                        new_o = label_dict.get(o, o)\n",
    "                        new_po_str = f\"{new_p}__{new_o}\"\n",
    "                        new_values[new_po_str] = count\n",
    "                    else:\n",
    "                        new_po_str = label_dict.get(po_str, po_str)\n",
    "                        new_values[new_po_str] = count\n",
    "            \n",
    "            new_data[new_key] = new_values\n",
    "            \n",
    "        return new_data\n",
    "    \n",
    "    return data\n",
    "\n",
    "\n",
    "\n",
    "def add_labels_to_graph(graph, label_dict):\n",
    "    for uri, label in label_dict.items():\n",
    "        graph.add((URIRef(uri), URIRef('http://www.w3.org/2004/02/skos/core#prefLabel'), Literal(label)))\n",
    "    return graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from rdflib import Graph\n",
    "import json\n",
    "from routers.utils import get_violation_report_exemplars\n",
    "\n",
    "ontology_union_violation_exemplars_g, edge_count_dict, focus_node_exemplar_dict, exemplar_focus_node_dict, violation_exemplar_dict = get_violation_report_exemplars(ontology_g, violations_g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "c = 0\n",
    "for k,v in violation_exemplar_dict.items():\n",
    "    if c == 10:\n",
    "        break\n",
    "    print(k, v)\n",
    "    c += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Also replace in joined .ttl of ontology and violations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For the graph\n",
    "ontology_union_violation_exemplars_g = add_labels_to_graph(ontology_union_violation_exemplars_g, label_dict)\n",
    "\n",
    "# For the dictionaries\n",
    "edge_count_dict = replace_uris_with_labels(edge_count_dict, label_dict, data_type=\"dict\")\n",
    "focus_node_exemplar_dict = replace_uris_with_labels(focus_node_exemplar_dict, label_dict, data_type=\"dict\")\n",
    "exemplar_focus_node_dict = replace_uris_with_labels(exemplar_focus_node_dict, label_dict, data_type=\"dict\")\n",
    "# if necessary add code for replacing uris with labels in violation_exemplar_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Store"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ontology_union_violation_exemplars_g.serialize(destination=OMICS_MODEL_UNION_VIOLATION_EXEMPLAR_TTL_PATH, format=\"ttl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from routers.utils import save_nested_counts_dict_json, save_lists_dict\n",
    "\n",
    "save_nested_counts_dict_json(edge_count_dict, EXEMPLAR_EDGE_COUNT_DICT_PATH)\n",
    "save_lists_dict(focus_node_exemplar_dict, FOCUS_NODE_EXEMPLAR_DICT_PATH)\n",
    "save_lists_dict(exemplar_focus_node_dict, EXEMPLAR_FOCUS_NODE_DICT_PATH)\n",
    "save_nested_counts_dict_json(violation_exemplar_dict, VIOLATION_EXEMPLAR_DICT_PATH)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py38",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
